

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>celloracle.network package &mdash; celloracle 0.3.5 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../_static/jquery.js"></script>
        <script type="text/javascript" src="../_static/underscore.js"></script>
        <script type="text/javascript" src="../_static/doctools.js"></script>
        <script type="text/javascript" src="../_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> celloracle
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../index.html">Welcome to celloracle’s documentation!</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../installation/index.html">Installation</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/index.html">Tutorial</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="index.html">API</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../changelog/index.html">Changelog</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../license/index.html">License</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../citation/index.html">Authors and citations</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">celloracle</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
      <li>celloracle.network package</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/modules/celloracle.network.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="module-celloracle.network">
<span id="celloracle-network-package"></span><h1>celloracle.network package<a class="headerlink" href="#module-celloracle.network" title="Permalink to this headline">¶</a></h1>
<p>The <a class="reference internal" href="#module-celloracle.network" title="celloracle.network"><code class="xref py py-mod docutils literal notranslate"><span class="pre">network</span></code></a> module implements GRN inference.</p>
<dl class="class">
<dt id="celloracle.network.Net">
<em class="property">class </em><code class="sig-prename descclassname">celloracle.network.</code><code class="sig-name descname">Net</code><span class="sig-paren">(</span><em class="sig-param">gene_expression_matrix</em>, <em class="sig-param">gem_standerdized=None</em>, <em class="sig-param">TFinfo_matrix=None</em>, <em class="sig-param">cellstate=None</em>, <em class="sig-param">TFinfo_dic=None</em>, <em class="sig-param">annotation=None</em>, <em class="sig-param">verbose=True</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network.Net" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Net is a custom class for inferring sample-specific GRN from scRNA-seq data.
This class is used inside the Oracle class for GRN inference.
This class requires two types of information below.</p>
<ol class="arabic simple">
<li><p>Single-cell RNA-seq data:
The Net class needs processed scRNA-seq data.
Gene and cell filtering, quality check, normalization, log-transformation (but not scaling and centering) have to be done before starting the GRN calculation with this class.
You can also use any arbitrary metadata (i.e., mRNA count, cell-cycle phase) for GRN input.</p></li>
<li><p>Potential regulatory connection (or base GRN):
This method uses the list of potential regulatory TFs as input.
This information can be calculated from ATAC-seq data using the motif-analysis module.
If sample-specific ATAC-seq data is not available,
you can use general TF-binding info derived from public ATAC-seq dataset of various tissue/cell type.</p></li>
</ol>
<dl class="attribute">
<dt id="celloracle.network.Net.linkList">
<code class="sig-name descname">linkList</code><a class="headerlink" href="#celloracle.network.Net.linkList" title="Permalink to this definition">¶</a></dt>
<dd><p>The results of the GRN inference.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.network.Net.all_genes">
<code class="sig-name descname">all_genes</code><a class="headerlink" href="#celloracle.network.Net.all_genes" title="Permalink to this definition">¶</a></dt>
<dd><p>An array of all genes that exist in the input gene expression matrix</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>numpy.array</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.network.Net.embedding_name">
<code class="sig-name descname">embedding_name</code><a class="headerlink" href="#celloracle.network.Net.embedding_name" title="Permalink to this definition">¶</a></dt>
<dd><p>The key name name in adata.obsm containing dimensional reduction coordinates</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.network.Net.annotation">
<code class="sig-name descname">annotation</code><a class="headerlink" href="#celloracle.network.Net.annotation" title="Permalink to this definition">¶</a></dt>
<dd><p>Annotation. you can add custom annotation.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>dictionary</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.network.Net.coefs_dict">
<code class="sig-name descname">coefs_dict</code><a class="headerlink" href="#celloracle.network.Net.coefs_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Coefs of linear regression.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>dictionary</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.network.Net.stats_dict">
<code class="sig-name descname">stats_dict</code><a class="headerlink" href="#celloracle.network.Net.stats_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Statistic values about coefs.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>dictionary</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.network.Net.fitted_genes">
<code class="sig-name descname">fitted_genes</code><a class="headerlink" href="#celloracle.network.Net.fitted_genes" title="Permalink to this definition">¶</a></dt>
<dd><p>List of genes where the regression model was successfully calculated.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>list of str</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.network.Net.failed_genes">
<code class="sig-name descname">failed_genes</code><a class="headerlink" href="#celloracle.network.Net.failed_genes" title="Permalink to this definition">¶</a></dt>
<dd><p>List of genes that were not assigned coefs</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>list of str</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.network.Net.cellstate">
<code class="sig-name descname">cellstate</code><a class="headerlink" href="#celloracle.network.Net.cellstate" title="Permalink to this definition">¶</a></dt>
<dd><p>A metadata for GRN input</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.network.Net.TFinfo">
<code class="sig-name descname">TFinfo</code><a class="headerlink" href="#celloracle.network.Net.TFinfo" title="Permalink to this definition">¶</a></dt>
<dd><p>Information about potential regulatory TFs.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.network.Net.gem">
<code class="sig-name descname">gem</code><a class="headerlink" href="#celloracle.network.Net.gem" title="Permalink to this definition">¶</a></dt>
<dd><p>Merged matrix made with gene_expression_matrix and cellstate matrix.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.network.Net.gem_standerdized">
<code class="sig-name descname">gem_standerdized</code><a class="headerlink" href="#celloracle.network.Net.gem_standerdized" title="Permalink to this definition">¶</a></dt>
<dd><p>Almost the same as gem, but the gene_expression_matrix was standardized.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.network.Net.library_last_update_date">
<code class="sig-name descname">library_last_update_date</code><a class="headerlink" href="#celloracle.network.Net.library_last_update_date" title="Permalink to this definition">¶</a></dt>
<dd><p>Last update date of this code. This info is for code development. It can be deprecated in the future</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="celloracle.network.Net.object_initiation_date">
<code class="sig-name descname">object_initiation_date</code><a class="headerlink" href="#celloracle.network.Net.object_initiation_date" title="Permalink to this definition">¶</a></dt>
<dd><p>The date when this object was made.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="celloracle.network.Net.addAnnotation">
<code class="sig-name descname">addAnnotation</code><span class="sig-paren">(</span><em class="sig-param">annotation_dictionary</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network.Net.addAnnotation" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a new annotation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>annotation_dictionary</strong> (<em>dictionary</em>) – e.g. {“sample_name”: “NIH 3T3 cell”}</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="celloracle.network.Net.addTFinfo_dictionary">
<code class="sig-name descname">addTFinfo_dictionary</code><span class="sig-paren">(</span><em class="sig-param">TFdict</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network.Net.addTFinfo_dictionary" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a new TF info to pre-exiting TFdict.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>TFdict</strong> (<em>dictionary</em>) – python dictionary of TF info.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="celloracle.network.Net.addTFinfo_matrix">
<code class="sig-name descname">addTFinfo_matrix</code><span class="sig-paren">(</span><em class="sig-param">TFinfo_matrix</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network.Net.addTFinfo_matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Load TF info dataframe.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>TFinfo</strong> (<em>pandas.DataFrame</em>) – information about potential regulatory TFs.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="celloracle.network.Net.copy">
<code class="sig-name descname">copy</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network.Net.copy" title="Permalink to this definition">¶</a></dt>
<dd><p>Deepcopy itself</p>
</dd></dl>

<dl class="method">
<dt id="celloracle.network.Net.fit_All_genes">
<code class="sig-name descname">fit_All_genes</code><span class="sig-paren">(</span><em class="sig-param">bagging_number=200</em>, <em class="sig-param">scaling=True</em>, <em class="sig-param">model_method='bagging_ridge'</em>, <em class="sig-param">command_line_mode=False</em>, <em class="sig-param">log=None</em>, <em class="sig-param">alpha=1</em>, <em class="sig-param">verbose=True</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network.Net.fit_All_genes" title="Permalink to this definition">¶</a></dt>
<dd><p>Make ML models for all genes.
The calculation will be performed in parallel using scikit-learn bagging function.
You can select a modeling method (bagging_ridge or bayesian_ridge).  This calculation usually takes a long time.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bagging_number</strong> (<em>int</em>) – The number of estimators for bagging.</p></li>
<li><p><strong>scaling</strong> (<em>bool</em>) – Whether or not to scale regulatory gene expression values.</p></li>
<li><p><strong>model_method</strong> (<em>str</em>) – ML model name. Please select either “bagging_ridge” or “bayesian_ridge”</p></li>
<li><p><strong>command_line_mode</strong> (<em>bool</em>) – Please select False if the calculation is performed on jupyter notebook.</p></li>
<li><p><strong>log</strong> (<em>logging object</em>) – log object to output log</p></li>
<li><p><strong>alpha</strong> (<em>int</em>) – Strength of regularization.</p></li>
<li><p><strong>verbose</strong> (<em>bool</em>) – Whether or not to show a progress bar.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="celloracle.network.Net.fit_All_genes_parallel">
<code class="sig-name descname">fit_All_genes_parallel</code><span class="sig-paren">(</span><em class="sig-param">bagging_number=200</em>, <em class="sig-param">scaling=True</em>, <em class="sig-param">log=None</em>, <em class="sig-param">verbose=10</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network.Net.fit_All_genes_parallel" title="Permalink to this definition">¶</a></dt>
<dd><p>IMPORTANT: this function being debugged and is currently unavailable.</p>
<p>Make ML models for all genes.
The calculation will be performed in parallel using joblib parallel module.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bagging_number</strong> (<em>int</em>) – The number of estimators for bagging.</p></li>
<li><p><strong>scaling</strong> (<em>bool</em>) – Whether or not to scale regulatory gene expression values.</p></li>
<li><p><strong>log</strong> (<em>logging object</em>) – log object to output log</p></li>
<li><p><strong>verbose</strong> (<em>int</em>) – verbose for joblib parallel</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="celloracle.network.Net.fit_genes">
<code class="sig-name descname">fit_genes</code><span class="sig-paren">(</span><em class="sig-param">target_genes</em>, <em class="sig-param">bagging_number=200</em>, <em class="sig-param">scaling=True</em>, <em class="sig-param">model_method='bagging_ridge'</em>, <em class="sig-param">save_coefs=False</em>, <em class="sig-param">command_line_mode=False</em>, <em class="sig-param">log=None</em>, <em class="sig-param">alpha=1</em>, <em class="sig-param">verbose=True</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network.Net.fit_genes" title="Permalink to this definition">¶</a></dt>
<dd><p>Make ML models for genes of interest.
This calculation will be performed in parallel using scikit-learn’s bagging function.
You can select a modeling method; Please chose either bagging_ridge or bayesian_ridge.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>target_genes</strong> (<em>list of str</em>) – gene list</p></li>
<li><p><strong>bagging_number</strong> (<em>int</em>) – The number of estimators for bagging.</p></li>
<li><p><strong>scaling</strong> (<em>bool</em>) – Whether or not to scale regulatory gene expression values.</p></li>
<li><p><strong>model_method</strong> (<em>str</em>) – ML model name. Please select either “bagging_ridge” or “bayesian_ridge”</p></li>
<li><p><strong>save_coefs</strong> (<em>bool</em>) – Whether or not to store details of coef values in bagging model.</p></li>
<li><p><strong>command_line_mode</strong> (<em>bool</em>) – Please select False if the calculation is performed on jupyter notebook.</p></li>
<li><p><strong>log</strong> (<em>logging object</em>) – log object to output log</p></li>
<li><p><strong>alpha</strong> (<em>int</em>) – Strength of regularization.</p></li>
<li><p><strong>verbose</strong> (<em>bool</em>) – Whether or not to show a progress bar.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="celloracle.network.Net.plotCoefs">
<code class="sig-name descname">plotCoefs</code><span class="sig-paren">(</span><em class="sig-param">target_gene</em>, <em class="sig-param">sort=True</em>, <em class="sig-param">threshold_p=None</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network.Net.plotCoefs" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot the distribution of Coef values (network edge weights).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>target_gene</strong> (<em>str</em>) – gene name</p></li>
<li><p><strong>sort</strong> (<em>bool</em>) – Whether or not to sort genes by its strength</p></li>
<li><p><strong>bagging_number</strong> (<em>int</em>) – The number of estimators for bagging.</p></li>
<li><p><strong>threshold_p</strong> (<em>float</em>) – the threshold for p-values. TFs will be filtered based on the p-value.
if None, no filtering is applied.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="celloracle.network.Net.to_hdf5">
<code class="sig-name descname">to_hdf5</code><span class="sig-paren">(</span><em class="sig-param">file_path</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network.Net.to_hdf5" title="Permalink to this definition">¶</a></dt>
<dd><p>Save object as hdf5.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>file_path</strong> (<em>str</em>) – file path to save file. Filename needs to end with ‘.celloracle.net’</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="celloracle.network.Net.updateLinkList">
<code class="sig-name descname">updateLinkList</code><span class="sig-paren">(</span><em class="sig-param">verbose=True</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network.Net.updateLinkList" title="Permalink to this definition">¶</a></dt>
<dd><p>Update LinkList.
LinkList is a data frame that store information about inferred GRNs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>verbose</strong> (<em>bool</em>) – Whether or not to show a progress bar</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="celloracle.network.Net.updateTFinfo_dictionary">
<code class="sig-name descname">updateTFinfo_dictionary</code><span class="sig-paren">(</span><em class="sig-param">TFdict</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network.Net.updateTFinfo_dictionary" title="Permalink to this definition">¶</a></dt>
<dd><p>Update TF info matrix</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>TFdict</strong> (<em>dictionary</em>) – A python dictionary in which a key is Target gene, value are potential regulatory genes for the target gene.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="celloracle.network.load_net_from_patquets">
<code class="sig-prename descclassname">celloracle.network.</code><code class="sig-name descname">load_net_from_patquets</code><span class="sig-paren">(</span><em class="sig-param">folder_path</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network.load_net_from_patquets" title="Permalink to this definition">¶</a></dt>
<dd><p>Load a Net object that was saved with “save_as_compressed” function.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>folder_path</strong> (<em>str</em>) – Folder path</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="celloracle.network.getDF_TGxTF">
<code class="sig-prename descclassname">celloracle.network.</code><code class="sig-name descname">getDF_TGxTF</code><span class="sig-paren">(</span><em class="sig-param">net_object</em>, <em class="sig-param">value_of_interest</em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network.getDF_TGxTF" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract inferred GRN information and return as a pandas.DataFrame.
The results was converted as Target gene x TF.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>net_object</strong> (<a class="reference internal" href="celloracle.html#celloracle.Net" title="celloracle.Net"><em>Net</em></a>) – Net object. GRN calculation have to be done in this object.</p></li>
<li><p><strong>value_of_interest</strong> (<em>str</em>) – Kind of information to extract.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020, Samantha Morris Lab

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>