

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Custom class in celloracle &mdash; celloracle 0.6.9 documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="celloracle.motif_analysis module" href="celloracle.motif_analysis.html" />
    <link rel="prev" title="API" href="index.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home" alt="Documentation Home"> celloracle
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../index.html">Welcome to celloracle’s documentation!</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../installation/index.html">Installation</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/index.html">Tutorial</a></li>
</ul>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="index.html">API</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="index.html#command-line-api">Command Line API</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="index.html#python-api">Python API</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="#">Custom class in celloracle</a></li>
<li class="toctree-l3"><a class="reference internal" href="#modules-for-atac-seq-analysis">Modules for ATAC-seq analysis</a><ul>
<li class="toctree-l4"><a class="reference internal" href="celloracle.motif_analysis.html">celloracle.motif_analysis module</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#modules-for-network-analysis">Modules for Network analysis</a><ul>
<li class="toctree-l4"><a class="reference internal" href="celloracle.network_analysis.html">celloracle.network_analysis module</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#other-modules">Other modules</a><ul>
<li class="toctree-l4"><a class="reference internal" href="celloracle.go_analysis.html">celloracle.go_analysis module</a></li>
<li class="toctree-l4"><a class="reference internal" href="celloracle.utility.html">celloracle.utility module</a></li>
<li class="toctree-l4"><a class="reference internal" href="celloracle.data.html">celloracle.data module</a></li>
<li class="toctree-l4"><a class="reference internal" href="celloracle.data_conversion.html">celloracle.data_conversion module</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../changelog/index.html">Changelog</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../license/index.html">License</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../citation/index.html">Authors and citations</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">celloracle</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="index.html">API</a> &raquo;</li>
        
      <li>Custom class in celloracle</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/modules/celloracle.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="custom-class-in-celloracle">
<h1>Custom class in celloracle<a class="headerlink" href="#custom-class-in-celloracle" title="Permalink to this headline">¶</a></h1>
<p>We define some custom classes in celloracle.</p>
<span class="target" id="module-celloracle"></span><dl class="py class">
<dt id="celloracle.Links">
<em class="property">class </em><code class="sig-prename descclassname">celloracle.</code><code class="sig-name descname">Links</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">name</span></em>, <em class="sig-param"><span class="n">links_dict</span><span class="o">=</span><span class="default_value">{}</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Links" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>This is a class for the processing and visualization of GRNs.
Links object stores cluster-specific GRNs and metadata.
Please use “get_links” function in Oracle object to generate Links object.</p>
<dl class="py attribute">
<dt id="celloracle.Links.links_dict">
<code class="sig-name descname">links_dict</code><a class="headerlink" href="#celloracle.Links.links_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Dictionary that store unprocessed network data.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>dictionary</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="celloracle.Links.filtered_links">
<code class="sig-name descname">filtered_links</code><a class="headerlink" href="#celloracle.Links.filtered_links" title="Permalink to this definition">¶</a></dt>
<dd><p>Dictionary that store filtered network data.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>dictionary</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="celloracle.Links.merged_score">
<code class="sig-name descname">merged_score</code><a class="headerlink" href="#celloracle.Links.merged_score" title="Permalink to this definition">¶</a></dt>
<dd><p>Network scores.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>pandas.dataframe</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="celloracle.Links.cluster">
<code class="sig-name descname">cluster</code><a class="headerlink" href="#celloracle.Links.cluster" title="Permalink to this definition">¶</a></dt>
<dd><p>List of cluster name.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>list of str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="celloracle.Links.name">
<code class="sig-name descname">name</code><a class="headerlink" href="#celloracle.Links.name" title="Permalink to this definition">¶</a></dt>
<dd><p>Name of clustering unit.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="celloracle.Links.palette">
<code class="sig-name descname">palette</code><a class="headerlink" href="#celloracle.Links.palette" title="Permalink to this definition">¶</a></dt>
<dd><p>DataFrame that store color information.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>pandas.dataframe</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Links.filter_links">
<code class="sig-name descname">filter_links</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">p</span><span class="o">=</span><span class="default_value">0.001</span></em>, <em class="sig-param"><span class="n">weight</span><span class="o">=</span><span class="default_value">'coef_abs'</span></em>, <em class="sig-param"><span class="n">thread_number</span><span class="o">=</span><span class="default_value">10000</span></em>, <em class="sig-param"><span class="n">genelist_source</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">genelist_target</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Links.filter_links" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter network edges.
In most cases, inferred GRN has non-significant random edges.
We have to remove these edges before analyzing the network structure.
You can do the filtering in any of the following ways.</p>
<ol class="arabic simple">
<li><p>Filter based on the p-value of the network edge.
Please enter p-value for thresholding.</p></li>
<li><p>Filter based on network edge number.
If you set the number, network edges will be filtered based on the order of a network score. The top n-th network edges with network weight will remain, and the other edges will be removed.
The network data has several types of network weight, so you have to select which network weight do you want to use.</p></li>
<li><p>Filter based on an arbitrary gene list. You can set a gene list for source nodes or target nodes.</p></li>
</ol>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>p</strong> (<em>float</em>) – threshold for p-value of the network edge.</p></li>
<li><p><strong>weight</strong> (<em>str</em>) – Please select network weight name for the filtering</p></li>
<li><p><strong>genelist_source</strong> (<em>list of str</em>) – gene list to remain in regulatory gene nodes. Default is None.</p></li>
<li><p><strong>genelist_target</strong> (<em>list of str</em>) – gene list to remain in target gene nodes. Default is None.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Links.get_network_entropy">
<code class="sig-name descname">get_network_entropy</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">value</span><span class="o">=</span><span class="default_value">'coef_abs'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Links.get_network_entropy" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate network entropy scores.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>value</strong> (<em>str</em>) – Default is “coef_abs”.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Links.get_score">
<code class="sig-name descname">get_score</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">test_mode</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Links.get_score" title="Permalink to this definition">¶</a></dt>
<dd><p>Get several network sores using R libraries.
Make sure all dependent R libraries are installed in your environment before running this function.
You can check the installation for the R libraries by running test_installation() in network_analysis module.</p>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Links.plot_cartography_scatter_per_cluster">
<code class="sig-name descname">plot_cartography_scatter_per_cluster</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">gois</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">clusters</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">scatter</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">kde</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">auto_gene_annot</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">percentile</span><span class="o">=</span><span class="default_value">98</span></em>, <em class="sig-param"><span class="n">args_dot</span><span class="o">=</span><span class="default_value">{'n_levels': 105}</span></em>, <em class="sig-param"><span class="n">args_line</span><span class="o">=</span><span class="default_value">{'c': 'gray'}</span></em>, <em class="sig-param"><span class="n">args_annot</span><span class="o">=</span><span class="default_value">{}</span></em>, <em class="sig-param"><span class="n">save</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Links.plot_cartography_scatter_per_cluster" title="Permalink to this definition">¶</a></dt>
<dd><p>Make a gene network cartography plot.
Please read the original paper describing gene network cartography for more information.
<a class="reference external" href="https://www.nature.com/articles/nature03288">https://www.nature.com/articles/nature03288</a></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>links</strong> (<a class="reference internal" href="#celloracle.Links" title="celloracle.Links"><em>Links</em></a>) – See network_analysis.Links class for detail.</p></li>
<li><p><strong>gois</strong> (<em>list of srt</em>) – List of gene name to highlight.</p></li>
<li><p><strong>clusters</strong> (<em>list of str</em>) – List of cluster name to analyze. If None, all clusters in Links object will be analyzed.</p></li>
<li><p><strong>scatter</strong> (<em>bool</em>) – Whether to make a scatter plot.</p></li>
<li><p><strong>auto_gene_annot</strong> (<em>bool</em>) – Whether to pick up genes to make an annotation.</p></li>
<li><p><strong>percentile</strong> (<em>float</em>) – Genes with a network score above the percentile will be shown with annotation. Default is 98.</p></li>
<li><p><strong>args_dot</strong> (<em>dictionary</em>) – Arguments for scatter plot.</p></li>
<li><p><strong>args_line</strong> (<em>dictionary</em>) – Arguments for lines in cartography plot.</p></li>
<li><p><strong>args_annot</strong> (<em>dictionary</em>) – Arguments for annotation in plots.</p></li>
<li><p><strong>save</strong> (<em>str</em>) – Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if [save=None]. Default is None.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Links.plot_cartography_term">
<code class="sig-name descname">plot_cartography_term</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">goi</span></em>, <em class="sig-param"><span class="n">save</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">plt_show</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Links.plot_cartography_term" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot the gene network cartography term like a heatmap.
Please read the original paper of gene network cartography for the principle of gene network cartography.
<a class="reference external" href="https://www.nature.com/articles/nature03288">https://www.nature.com/articles/nature03288</a></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>links</strong> (<a class="reference internal" href="#celloracle.Links" title="celloracle.Links"><em>Links</em></a>) – See network_analysis.Links class for detail.</p></li>
<li><p><strong>gois</strong> (<em>list of srt</em>) – List of gene name to highlight.</p></li>
<li><p><strong>save</strong> (<em>str</em>) – Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if [save=None]. Default is None.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Links.plot_degree_distributions">
<code class="sig-name descname">plot_degree_distributions</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">plot_model</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">save</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Links.plot_degree_distributions" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot the network degree distributions (the number of edge per gene).
The network degree will be visualized in both linear scale and log scale.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>links</strong> (<a class="reference internal" href="#celloracle.Links" title="celloracle.Links"><em>Links</em></a>) – See network_analysis.Links class for detail.</p></li>
<li><p><strong>plot_model</strong> (<em>bool</em>) – Whether to plot linear approximation line.</p></li>
<li><p><strong>save</strong> (<em>str</em>) – Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if [save=None]. Default is None.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Links.plot_network_entropy_distributions">
<code class="sig-name descname">plot_network_entropy_distributions</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">update_network_entropy</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">save</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Links.plot_network_entropy_distributions" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot the distribution for network entropy.
See the CellOracle paper for more detail.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>links</strong> (<em>Links object</em>) – See network_analysis.Links class for detail.</p></li>
<li><p><strong>values</strong> (<em>list of str</em>) – The list of score to visualize. If it is None, all network score (listed above) will be used.</p></li>
<li><p><strong>update_network_entropy</strong> (<em>bool</em>) – Whether to recalculate network entropy.</p></li>
<li><p><strong>save</strong> (<em>str</em>) – Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if [save=None]. Default is None.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Links.plot_score_comparison_2D">
<code class="sig-name descname">plot_score_comparison_2D</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">value</span></em>, <em class="sig-param"><span class="n">cluster1</span></em>, <em class="sig-param"><span class="n">cluster2</span></em>, <em class="sig-param"><span class="n">percentile</span><span class="o">=</span><span class="default_value">99</span></em>, <em class="sig-param"><span class="n">annot_shifts</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">save</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">plt_show</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">interactive</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Links.plot_score_comparison_2D" title="Permalink to this definition">¶</a></dt>
<dd><p>Make a scatter plot that compares specific network scores in two groups.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>links</strong> (<a class="reference internal" href="#celloracle.Links" title="celloracle.Links"><em>Links</em></a>) – See network_analysis.Links class for detail.</p></li>
<li><p><strong>value</strong> (<em>srt</em>) – The network score type.</p></li>
<li><p><strong>cluster1</strong> (<em>str</em>) – Cluster name. Network scores in cluster1 will be visualized in the x-axis.</p></li>
<li><p><strong>cluster2</strong> (<em>str</em>) – Cluster name. Network scores in cluster2 will be visualized in the y-axis.</p></li>
<li><p><strong>percentile</strong> (<em>float</em>) – Genes with a network score above the percentile will be shown with annotation. Default is 99.</p></li>
<li><p><strong>annot_shifts</strong> (<em>(</em><em>float</em><em>, </em><em>float</em><em>)</em>) – Annotation visualization setting.</p></li>
<li><p><strong>save</strong> (<em>str</em>) – Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if [save=None]. Default is None.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Links.plot_score_discributions">
<code class="sig-name descname">plot_score_discributions</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">values</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">method</span><span class="o">=</span><span class="default_value">'boxplot'</span></em>, <em class="sig-param"><span class="n">save</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Links.plot_score_discributions" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot the distribution of network scores.
An individual data point is a network edge (gene).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>links</strong> (<a class="reference internal" href="#celloracle.Links" title="celloracle.Links"><em>Links</em></a>) – See Links class for details.</p></li>
<li><p><strong>values</strong> (<em>list of str</em>) – The list of score to visualize. If it is None, all of the network score will be used.</p></li>
<li><p><strong>method</strong> (<em>str</em>) – Plotting method. Select either “boxplot” or “barplot”.</p></li>
<li><p><strong>save</strong> (<em>str</em>) – Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if [save=None]. Default is None.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Links.plot_score_per_cluster">
<code class="sig-name descname">plot_score_per_cluster</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">goi</span></em>, <em class="sig-param"><span class="n">save</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">plt_show</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Links.plot_score_per_cluster" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot network score for a gene.
This function visualizes the network score for a specific gene between clusters to get an insight into the dynamics of the gene.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>links</strong> (<a class="reference internal" href="#celloracle.Links" title="celloracle.Links"><em>Links</em></a>) – See network_analysis.Links class for detail.</p></li>
<li><p><strong>goi</strong> (<em>srt</em>) – Gene name.</p></li>
<li><p><strong>save</strong> (<em>str</em>) – Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if [save=None]. Default is None.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Links.plot_scores_as_rank">
<code class="sig-name descname">plot_scores_as_rank</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cluster</span></em>, <em class="sig-param"><span class="n">n_gene</span><span class="o">=</span><span class="default_value">50</span></em>, <em class="sig-param"><span class="n">save</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Links.plot_scores_as_rank" title="Permalink to this definition">¶</a></dt>
<dd><p>Pick up top n-th genes wich high-network scores and make plots.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>links</strong> (<a class="reference internal" href="#celloracle.Links" title="celloracle.Links"><em>Links</em></a>) – See network_analysis.Links class for detail.</p></li>
<li><p><strong>cluster</strong> (<em>str</em>) – Cluster name to analyze.</p></li>
<li><p><strong>n_gene</strong> (<em>int</em>) – Number of genes to plot. Default is 50.</p></li>
<li><p><strong>save</strong> (<em>str</em>) – Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if [save=None]. Default is None.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Links.to_hdf5">
<code class="sig-name descname">to_hdf5</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">file_path</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Links.to_hdf5" title="Permalink to this definition">¶</a></dt>
<dd><p>Save object as hdf5.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>file_path</strong> (<em>str</em>) – file path to save file. Filename needs to end with ‘.celloracle.links’</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="celloracle.Net">
<em class="property">class </em><code class="sig-prename descclassname">celloracle.</code><code class="sig-name descname">Net</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">gene_expression_matrix</span></em>, <em class="sig-param"><span class="n">gem_standerdized</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">TFinfo_matrix</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">cellstate</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">TFinfo_dic</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">annotation</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">verbose</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Net" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Net is a custom class for inferring sample-specific GRN from scRNA-seq data.
This class is used inside the Oracle class for GRN inference.
This class requires two types of information below.</p>
<ol class="arabic simple">
<li><p>Single-cell RNA-seq data:
The Net class needs processed scRNA-seq data.
Gene and cell filtering, quality check, normalization, log-transformation (but not scaling and centering) have to be done before starting the GRN calculation with this class.
You can also use any arbitrary metadata (i.e., mRNA count, cell-cycle phase) for GRN input.</p></li>
<li><p>Potential regulatory connection (or base GRN):
This method uses the list of potential regulatory TFs as input.
This information can be calculated from ATAC-seq data using the motif-analysis module.
If sample-specific ATAC-seq data is not available,
you can use general TF-binding info derived from public ATAC-seq dataset of various tissue/cell type.</p></li>
</ol>
<dl class="py attribute">
<dt id="celloracle.Net.linkList">
<code class="sig-name descname">linkList</code><a class="headerlink" href="#celloracle.Net.linkList" title="Permalink to this definition">¶</a></dt>
<dd><p>The results of the GRN inference.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="celloracle.Net.all_genes">
<code class="sig-name descname">all_genes</code><a class="headerlink" href="#celloracle.Net.all_genes" title="Permalink to this definition">¶</a></dt>
<dd><p>An array of all genes that exist in the input gene expression matrix</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>numpy.array</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="celloracle.Net.embedding_name">
<code class="sig-name descname">embedding_name</code><a class="headerlink" href="#celloracle.Net.embedding_name" title="Permalink to this definition">¶</a></dt>
<dd><p>The key name name in adata.obsm containing dimensional reduction coordinates</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="celloracle.Net.annotation">
<code class="sig-name descname">annotation</code><a class="headerlink" href="#celloracle.Net.annotation" title="Permalink to this definition">¶</a></dt>
<dd><p>Annotation. you can add custom annotation.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>dictionary</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="celloracle.Net.coefs_dict">
<code class="sig-name descname">coefs_dict</code><a class="headerlink" href="#celloracle.Net.coefs_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Coefs of linear regression.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>dictionary</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="celloracle.Net.stats_dict">
<code class="sig-name descname">stats_dict</code><a class="headerlink" href="#celloracle.Net.stats_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Statistic values about coefs.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>dictionary</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="celloracle.Net.fitted_genes">
<code class="sig-name descname">fitted_genes</code><a class="headerlink" href="#celloracle.Net.fitted_genes" title="Permalink to this definition">¶</a></dt>
<dd><p>List of genes where the regression model was successfully calculated.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>list of str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="celloracle.Net.failed_genes">
<code class="sig-name descname">failed_genes</code><a class="headerlink" href="#celloracle.Net.failed_genes" title="Permalink to this definition">¶</a></dt>
<dd><p>List of genes that were not assigned coefs</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>list of str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="celloracle.Net.cellstate">
<code class="sig-name descname">cellstate</code><a class="headerlink" href="#celloracle.Net.cellstate" title="Permalink to this definition">¶</a></dt>
<dd><p>A metadata for GRN input</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="celloracle.Net.TFinfo">
<code class="sig-name descname">TFinfo</code><a class="headerlink" href="#celloracle.Net.TFinfo" title="Permalink to this definition">¶</a></dt>
<dd><p>Information about potential regulatory TFs.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="celloracle.Net.gem">
<code class="sig-name descname">gem</code><a class="headerlink" href="#celloracle.Net.gem" title="Permalink to this definition">¶</a></dt>
<dd><p>Merged matrix made with gene_expression_matrix and cellstate matrix.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="celloracle.Net.gem_standerdized">
<code class="sig-name descname">gem_standerdized</code><a class="headerlink" href="#celloracle.Net.gem_standerdized" title="Permalink to this definition">¶</a></dt>
<dd><p>Almost the same as gem, but the gene_expression_matrix was standardized.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="celloracle.Net.library_last_update_date">
<code class="sig-name descname">library_last_update_date</code><a class="headerlink" href="#celloracle.Net.library_last_update_date" title="Permalink to this definition">¶</a></dt>
<dd><p>Last update date of this code. This info is for code development. It can be deprecated in the future</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="celloracle.Net.object_initiation_date">
<code class="sig-name descname">object_initiation_date</code><a class="headerlink" href="#celloracle.Net.object_initiation_date" title="Permalink to this definition">¶</a></dt>
<dd><p>The date when this object was made.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Net.addAnnotation">
<code class="sig-name descname">addAnnotation</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">annotation_dictionary</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Net.addAnnotation" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a new annotation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>annotation_dictionary</strong> (<em>dictionary</em>) – e.g. {“sample_name”: “NIH 3T3 cell”}</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Net.addTFinfo_dictionary">
<code class="sig-name descname">addTFinfo_dictionary</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">TFdict</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Net.addTFinfo_dictionary" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a new TF info to pre-exiting TFdict.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>TFdict</strong> (<em>dictionary</em>) – python dictionary of TF info.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Net.addTFinfo_matrix">
<code class="sig-name descname">addTFinfo_matrix</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">TFinfo_matrix</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Net.addTFinfo_matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Load TF info dataframe.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>TFinfo</strong> (<em>pandas.DataFrame</em>) – information about potential regulatory TFs.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Net.copy">
<code class="sig-name descname">copy</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Net.copy" title="Permalink to this definition">¶</a></dt>
<dd><p>Deepcopy itself</p>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Net.fit_All_genes">
<code class="sig-name descname">fit_All_genes</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">bagging_number</span><span class="o">=</span><span class="default_value">200</span></em>, <em class="sig-param"><span class="n">scaling</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">model_method</span><span class="o">=</span><span class="default_value">'bagging_ridge'</span></em>, <em class="sig-param"><span class="n">command_line_mode</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">log</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">alpha</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">verbose</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Net.fit_All_genes" title="Permalink to this definition">¶</a></dt>
<dd><p>Make ML models for all genes.
The calculation will be performed in parallel using scikit-learn bagging function.
You can select a modeling method (bagging_ridge or bayesian_ridge).  This calculation usually takes a long time.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bagging_number</strong> (<em>int</em>) – The number of estimators for bagging.</p></li>
<li><p><strong>scaling</strong> (<em>bool</em>) – Whether or not to scale regulatory gene expression values.</p></li>
<li><p><strong>model_method</strong> (<em>str</em>) – ML model name. Please select either “bagging_ridge” or “bayesian_ridge”</p></li>
<li><p><strong>command_line_mode</strong> (<em>bool</em>) – Please select False if the calculation is performed on jupyter notebook.</p></li>
<li><p><strong>log</strong> (<em>logging object</em>) – log object to output log</p></li>
<li><p><strong>alpha</strong> (<em>int</em>) – Strength of regularization.</p></li>
<li><p><strong>verbose</strong> (<em>bool</em>) – Whether or not to show a progress bar.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Net.fit_All_genes_parallel">
<code class="sig-name descname">fit_All_genes_parallel</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">bagging_number</span><span class="o">=</span><span class="default_value">200</span></em>, <em class="sig-param"><span class="n">scaling</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">log</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">verbose</span><span class="o">=</span><span class="default_value">10</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Net.fit_All_genes_parallel" title="Permalink to this definition">¶</a></dt>
<dd><p>IMPORTANT: this function being debugged and is currently unavailable.</p>
<p>Make ML models for all genes.
The calculation will be performed in parallel using joblib parallel module.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bagging_number</strong> (<em>int</em>) – The number of estimators for bagging.</p></li>
<li><p><strong>scaling</strong> (<em>bool</em>) – Whether or not to scale regulatory gene expression values.</p></li>
<li><p><strong>log</strong> (<em>logging object</em>) – log object to output log</p></li>
<li><p><strong>verbose</strong> (<em>int</em>) – verbose for joblib parallel</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Net.fit_genes">
<code class="sig-name descname">fit_genes</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target_genes</span></em>, <em class="sig-param"><span class="n">bagging_number</span><span class="o">=</span><span class="default_value">200</span></em>, <em class="sig-param"><span class="n">scaling</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">model_method</span><span class="o">=</span><span class="default_value">'bagging_ridge'</span></em>, <em class="sig-param"><span class="n">save_coefs</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">command_line_mode</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">log</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">alpha</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">verbose</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Net.fit_genes" title="Permalink to this definition">¶</a></dt>
<dd><p>Make ML models for genes of interest.
This calculation will be performed in parallel using scikit-learn’s bagging function.
You can select a modeling method; Please chose either bagging_ridge or bayesian_ridge.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>target_genes</strong> (<em>list of str</em>) – gene list</p></li>
<li><p><strong>bagging_number</strong> (<em>int</em>) – The number of estimators for bagging.</p></li>
<li><p><strong>scaling</strong> (<em>bool</em>) – Whether or not to scale regulatory gene expression values.</p></li>
<li><p><strong>model_method</strong> (<em>str</em>) – ML model name. Please select either “bagging_ridge” or “bayesian_ridge”</p></li>
<li><p><strong>save_coefs</strong> (<em>bool</em>) – Whether or not to store details of coef values in bagging model.</p></li>
<li><p><strong>command_line_mode</strong> (<em>bool</em>) – Please select False if the calculation is performed on jupyter notebook.</p></li>
<li><p><strong>log</strong> (<em>logging object</em>) – log object to output log</p></li>
<li><p><strong>alpha</strong> (<em>int</em>) – Strength of regularization.</p></li>
<li><p><strong>verbose</strong> (<em>bool</em>) – Whether or not to show a progress bar.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Net.plotCoefs">
<code class="sig-name descname">plotCoefs</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target_gene</span></em>, <em class="sig-param"><span class="n">sort</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">threshold_p</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Net.plotCoefs" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot the distribution of Coef values (network edge weights).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>target_gene</strong> (<em>str</em>) – gene name</p></li>
<li><p><strong>sort</strong> (<em>bool</em>) – Whether or not to sort genes by its strength</p></li>
<li><p><strong>bagging_number</strong> (<em>int</em>) – The number of estimators for bagging.</p></li>
<li><p><strong>threshold_p</strong> (<em>float</em>) – the threshold for p-values. TFs will be filtered based on the p-value.
if None, no filtering is applied.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Net.to_hdf5">
<code class="sig-name descname">to_hdf5</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">file_path</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Net.to_hdf5" title="Permalink to this definition">¶</a></dt>
<dd><p>Save object as hdf5.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>file_path</strong> (<em>str</em>) – file path to save file. Filename needs to end with ‘.celloracle.net’</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Net.updateLinkList">
<code class="sig-name descname">updateLinkList</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">verbose</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Net.updateLinkList" title="Permalink to this definition">¶</a></dt>
<dd><p>Update LinkList.
LinkList is a data frame that store information about inferred GRNs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>verbose</strong> (<em>bool</em>) – Whether or not to show a progress bar</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Net.updateTFinfo_dictionary">
<code class="sig-name descname">updateTFinfo_dictionary</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">TFdict</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Net.updateTFinfo_dictionary" title="Permalink to this definition">¶</a></dt>
<dd><p>Update TF info matrix</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>TFdict</strong> (<em>dictionary</em>) – A python dictionary in which a key is Target gene, value are potential regulatory genes for the target gene.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="celloracle.Oracle">
<em class="property">class </em><code class="sig-prename descclassname">celloracle.</code><code class="sig-name descname">Oracle</code><a class="headerlink" href="#celloracle.Oracle" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">celloracle.trajectory.modified_VelocytoLoom_class.modified_VelocytoLoom</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">celloracle.visualizations.oracle_object_visualization.Oracle_visualization</span></code></p>
<p>Oracle is the main class in CellOracle. Oracle object imports scRNA-seq data (anndata) and TF information to infer cluster-specific GRNs. It can predict the future gene expression patterns and cell state transitions in response to  the perturbation of TFs. Please see the CellOracle paper for details.
The code of the Oracle class was made of the three components below.</p>
<ol class="arabic simple">
<li><p>Anndata: Gene expression matrix and metadata from single-cell RNA-seq are stored in the anndata object. Processed values, such as normalized counts and simulated values, are stored as layers of anndata. Metadata (i.e., Cluster info) are saved in anndata.obs. Refer to scanpy/anndata documentation for detail.</p></li>
<li><p>Net: Net is a custom class in celloracle. Net object processes several data to infer GRN. See the Net class documentation for details.</p></li>
<li><p>VelycytoLoom: Calculation of transition probability and visualization of directed trajectory graph will be performed in the same way as velocytoloom. VelocytoLoom is class from Velocyto, a python library for RNA-velocity analysis. In celloracle, we use some functions in velocytoloom for the visualization.</p></li>
</ol>
<dl class="py attribute">
<dt id="celloracle.Oracle.adata">
<code class="sig-name descname">adata</code><a class="headerlink" href="#celloracle.Oracle.adata" title="Permalink to this definition">¶</a></dt>
<dd><p>Imported anndata object</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>anndata</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="celloracle.Oracle.cluster_column_name">
<code class="sig-name descname">cluster_column_name</code><a class="headerlink" href="#celloracle.Oracle.cluster_column_name" title="Permalink to this definition">¶</a></dt>
<dd><p>The column name in adata.obs containing cluster info</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="celloracle.Oracle.embedding_name">
<code class="sig-name descname">embedding_name</code><a class="headerlink" href="#celloracle.Oracle.embedding_name" title="Permalink to this definition">¶</a></dt>
<dd><p>The key name in adata.obsm containing dimensional reduction cordinates</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Oracle.addTFinfo_dictionary">
<code class="sig-name descname">addTFinfo_dictionary</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">TFdict</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Oracle.addTFinfo_dictionary" title="Permalink to this definition">¶</a></dt>
<dd><p>Add new TF info to pre-existing TFdict.
Values in the old TF dictionary will remain.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>TFdict</strong> (<em>dictionary</em>) – Python dictionary of TF info.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Oracle.calculate_mass_filter">
<code class="sig-name descname">calculate_mass_filter</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">min_mass</span><span class="o">=</span><span class="default_value">0.01</span></em>, <em class="sig-param"><span class="n">plot</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Oracle.calculate_mass_filter" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt id="celloracle.Oracle.calculate_p_mass">
<code class="sig-name descname">calculate_p_mass</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">smooth</span><span class="o">=</span><span class="default_value">0.8</span></em>, <em class="sig-param"><span class="n">n_grid</span><span class="o">=</span><span class="default_value">40</span></em>, <em class="sig-param"><span class="n">n_neighbors</span><span class="o">=</span><span class="default_value">200</span></em>, <em class="sig-param"><span class="n">n_jobs</span><span class="o">=</span><span class="default_value">- 1</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Oracle.calculate_p_mass" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt id="celloracle.Oracle.change_cluster_unit">
<code class="sig-name descname">change_cluster_unit</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">new_cluster_column_name</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Oracle.change_cluster_unit" title="Permalink to this definition">¶</a></dt>
<dd><p>Change clustering unit.
If you change cluster, previous GRN data and simulation data will be delated.
Please re-calculate GRNs.</p>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Oracle.copy">
<code class="sig-name descname">copy</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Oracle.copy" title="Permalink to this definition">¶</a></dt>
<dd><p>Deepcopy itself.</p>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Oracle.count_cells_in_mc_resutls">
<code class="sig-name descname">count_cells_in_mc_resutls</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cluster_use</span></em>, <em class="sig-param"><span class="n">end</span><span class="o">=</span><span class="default_value">- 1</span></em>, <em class="sig-param"><span class="n">order</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Oracle.count_cells_in_mc_resutls" title="Permalink to this definition">¶</a></dt>
<dd><p>Count the simulated cell by the cluster.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>cluster_use</strong> (<em>str</em>) – cluster information name in anndata.obs.
You can use any cluster information in anndata.obs.</p></li>
<li><p><strong>end</strong> (<em>int</em>) – The end point of Sankey-diagram. Please select a  step in the Markov simulation.
if you set [end=-1], the final step of Markov simulation will be used.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Number of cells before / after simulation</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Oracle.extract_active_gene_lists">
<code class="sig-name descname">extract_active_gene_lists</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">return_as</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">verbose</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Oracle.extract_active_gene_lists" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>return_as</strong> (<em>str</em>) – If not None, it returns dictionary or list. Chose either “indivisual_dict” or “unified_list”.</p></li>
<li><p><strong>verbose</strong> (<em>bool</em>) – Whether to show progress bar.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The format depends on the argument, “return_as”.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>dictionary or list</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Oracle.fit_GRN_for_simulation">
<code class="sig-name descname">fit_GRN_for_simulation</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">GRN_unit</span><span class="o">=</span><span class="default_value">'cluster'</span></em>, <em class="sig-param"><span class="n">alpha</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">use_cluster_specific_TFdict</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Oracle.fit_GRN_for_simulation" title="Permalink to this definition">¶</a></dt>
<dd><p>Do GRN inference.
Please see the paper of CellOracle paper for details.</p>
<p>GRN can be constructed for the entire population or each clusters.
If you want to infer cluster-specific GRN, please set [GRN_unit=”cluster”].
You can select cluster information when you import data.</p>
<p>If you set [GRN_unit=”whole”], GRN will be made using all cells.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>GRN_unit</strong> (<em>str</em>) – Select “cluster” or “whole”</p></li>
<li><p><strong>alpha</strong> (<em>float</em><em> or </em><em>int</em>) – The strength of regularization.
If you set a lower value, the sensitivity increases, and you can detect weaker network connections. However, there may be more noise.
If you select a higher value, it will reduce the chance of overfitting.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Oracle.get_cluster_specific_TFdict_from_Links">
<code class="sig-name descname">get_cluster_specific_TFdict_from_Links</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">links_object</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Oracle.get_cluster_specific_TFdict_from_Links" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract TF and its target gene information from Links object.
This function can be used to reconstruct GRNs based on pre-existing GRNs saved in Links object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>links_object</strong> (<a class="reference internal" href="#celloracle.Links" title="celloracle.Links"><em>Links</em></a>) – Please see the explanation of Links class.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Oracle.get_links">
<code class="sig-name descname">get_links</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cluster_name_for_GRN_unit</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">alpha</span><span class="o">=</span><span class="default_value">10</span></em>, <em class="sig-param"><span class="n">bagging_number</span><span class="o">=</span><span class="default_value">20</span></em>, <em class="sig-param"><span class="n">verbose_level</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">test_mode</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Oracle.get_links" title="Permalink to this definition">¶</a></dt>
<dd><p>Makes GRN for each cluster and returns results as a Links object.
Several preprocessing should be done before using this function.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>cluster_name_for_GRN_unit</strong> (<em>str</em>) – Cluster name for GRN calculation. The cluster information should be stored in Oracle.adata.obs.</p></li>
<li><p><strong>alpha</strong> (<em>float</em><em> or </em><em>int</em>) – The strength of regularization.
If you set a lower value, the sensitivity increases, and you can detect weaker network connections. However, there may be more noise.
If you select a higher value, it will reduce the chance of overfitting.</p></li>
<li><p><strong>bagging_number</strong> (<em>int</em>) – The number used in bagging calculation.</p></li>
<li><p><strong>verbose_level</strong> (<em>int</em>) – if [verbose_level&gt;1], most detailed progress information will be shown.
if [verbose_level &gt; 0], one progress bar will be shown.
if [verbose_level == 0], no progress bar will be shown.</p></li>
<li><p><strong>test_mode</strong> (<em>bool</em>) – If test_mode is True, GRN calculation will be done for only one cluster rather than all clusters.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Oracle.get_mcmc_cell_transition_table">
<code class="sig-name descname">get_mcmc_cell_transition_table</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cluster_column_name</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">end</span><span class="o">=</span><span class="default_value">- 1</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Oracle.get_mcmc_cell_transition_table" title="Permalink to this definition">¶</a></dt>
<dd><p>Return cell count in the initial state and final state after mcmc.
Cell counts are grouped by the cluster of interest.
Result will be returned as 2D matrix.</p>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Oracle.import_TF_data">
<code class="sig-name descname">import_TF_data</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">TF_info_matrix</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">TF_info_matrix_path</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">TFdict</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Oracle.import_TF_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Load data about potential-regulatory TFs.
You can import either TF_info_matrix or TFdict.
For more information on how to make these files, please see the motif analysis module within the celloracle tutorial.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>TF_info_matrix</strong> (<em>pandas.DataFrame</em>) – TF_info_matrix.</p></li>
<li><p><strong>TF_info_matrix_path</strong> (<em>str</em>) – File path for TF_info_matrix (pandas.DataFrame).</p></li>
<li><p><strong>TFdict</strong> (<em>dictionary</em>) – Python dictionary of TF info.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Oracle.import_anndata_as_normalized_count">
<code class="sig-name descname">import_anndata_as_normalized_count</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">adata</span></em>, <em class="sig-param"><span class="n">cluster_column_name</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">embedding_name</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">test_mode</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Oracle.import_anndata_as_normalized_count" title="Permalink to this definition">¶</a></dt>
<dd><p>Load scRNA-seq data. scRNA-seq data should be prepared as an anndata object.
Preprocessing (cell and gene filtering, dimensional reduction, clustering, etc.) should be done before loading data.
The method will import NORMALIZED and LOG TRANSFORMED data but NOT SCALED and NOT CENTERED data.
See the tutorial for more details on how to process scRNA-seq data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>anndata</em>) – anndata object containing scRNA-seq data.</p></li>
<li><p><strong>cluster_column_name</strong> (<em>str</em>) – the name of column containing cluster information in anndata.obs.
Clustering data should be in anndata.obs.</p></li>
<li><p><strong>embedding_name</strong> (<em>str</em>) – the key name for dimensional reduction information in anndata.obsm.
Dimensional reduction (or 2D trajectory graph) should be in anndata.obsm.</p></li>
<li><p><strong>transform</strong> (<em>str</em>) – The method for log-transformation. Chose one from “natural_log” or “log2”.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Oracle.import_anndata_as_raw_count">
<code class="sig-name descname">import_anndata_as_raw_count</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">adata</span></em>, <em class="sig-param"><span class="n">cluster_column_name</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">embedding_name</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">transform</span><span class="o">=</span><span class="default_value">'natural_log'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Oracle.import_anndata_as_raw_count" title="Permalink to this definition">¶</a></dt>
<dd><p>Load scRNA-seq data. scRNA-seq data should be prepared as an anndata object.
Preprocessing (cell and gene filtering, dimensional reduction, clustering, etc.) should be done before loading data.
The method imports RAW GENE COUNTS because unscaled and uncentered gene expression data are required for the GRN inference and simulation.
See tutorial notebook for the details about how to process scRNA-seq data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>anndata</em>) – anndata object that stores scRNA-seq data.</p></li>
<li><p><strong>cluster_column_name</strong> (<em>str</em>) – the name of column containing cluster information in anndata.obs.
Clustering data should be in anndata.obs.</p></li>
<li><p><strong>embedding_name</strong> (<em>str</em>) – the key name for dimensional reduction information in anndata.obsm.
Dimensional reduction (or 2D trajectory graph) should be in anndata.obsm.</p></li>
<li><p><strong>transform</strong> (<em>str</em>) – The method for log-transformation. Chose one from “natural_log” or “log2”.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Oracle.plot_mc_result_as_kde">
<code class="sig-name descname">plot_mc_result_as_kde</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">n_time</span></em>, <em class="sig-param"><span class="n">args</span><span class="o">=</span><span class="default_value">{}</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Oracle.plot_mc_result_as_kde" title="Permalink to this definition">¶</a></dt>
<dd><p>Pick up one timepoint in the cell state-transition simulation and plot as a kde plot.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n_time</strong> (<em>int</em>) – the number in Markov simulation</p></li>
<li><p><strong>args</strong> (<em>dictionary</em>) – An argument for seaborn.kdeplot.
See seaborn documentation for details (<a class="reference external" href="https://seaborn.pydata.org/generated/seaborn.kdeplot.html#seaborn.kdeplot">https://seaborn.pydata.org/generated/seaborn.kdeplot.html#seaborn.kdeplot</a>).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Oracle.plot_mc_result_as_trajectory">
<code class="sig-name descname">plot_mc_result_as_trajectory</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cell_name</span></em>, <em class="sig-param"><span class="n">time_range</span></em>, <em class="sig-param"><span class="n">args</span><span class="o">=</span><span class="default_value">{}</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Oracle.plot_mc_result_as_trajectory" title="Permalink to this definition">¶</a></dt>
<dd><p>Pick up several timepoints in the cell state-transition simulation and plot as a line plot.
This function can be used to visualize how cell-state changes after perturbation focusing on a specific cell.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>cell_name</strong> (<em>str</em>) – cell name. chose from adata.obs.index</p></li>
<li><p><strong>time_range</strong> (<em>list of int</em>) – the list of index in Markov simulation</p></li>
<li><p><strong>args</strong> (<em>dictionary</em>) – dictionary for the arguments for matplotlib.pyplit.plot.
See matplotlib documentation for details (<a class="reference external" href="https://matplotlib.org/api/_as_gen/matplotlib.pyplot.plot.html#matplotlib.pyplot.plot">https://matplotlib.org/api/_as_gen/matplotlib.pyplot.plot.html#matplotlib.pyplot.plot</a>).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Oracle.plot_mc_resutls_as_sankey">
<code class="sig-name descname">plot_mc_resutls_as_sankey</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cluster_use</span></em>, <em class="sig-param"><span class="n">start</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">end</span><span class="o">=</span><span class="default_value">- 1</span></em>, <em class="sig-param"><span class="n">order</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">font_size</span><span class="o">=</span><span class="default_value">10</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Oracle.plot_mc_resutls_as_sankey" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot the simulated cell state-transition as a Sankey-diagram after groping by the cluster.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>cluster_use</strong> (<em>str</em>) – cluster information name in anndata.obs.
You can use any cluster information in anndata.obs.</p></li>
<li><p><strong>start</strong> (<em>int</em>) – The starting point of Sankey-diagram. Please select a  step in the Markov simulation.</p></li>
<li><p><strong>end</strong> (<em>int</em>) – The end point of Sankey-diagram. Please select a  step in the Markov simulation.
if you set [end=-1], the final step of Markov simulation will be used.</p></li>
<li><p><strong>order</strong> (<em>list of str</em>) – The order of cluster name in the Sankey-diagram.</p></li>
<li><p><strong>font_size</strong> (<em>int</em>) – Font size for cluster name label in the Sankey diagram.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Oracle.prepare_markov_simulation">
<code class="sig-name descname">prepare_markov_simulation</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">verbose</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Oracle.prepare_markov_simulation" title="Permalink to this definition">¶</a></dt>
<dd><p>Pick up cells for Markov simulation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>verbose</strong> (<em>bool</em>) – If True, it plots selected cells.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Oracle.run_markov_chain_simulation">
<code class="sig-name descname">run_markov_chain_simulation</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">n_steps</span><span class="o">=</span><span class="default_value">500</span></em>, <em class="sig-param"><span class="n">n_duplication</span><span class="o">=</span><span class="default_value">5</span></em>, <em class="sig-param"><span class="n">seed</span><span class="o">=</span><span class="default_value">123</span></em>, <em class="sig-param"><span class="n">calculate_randomized</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Oracle.run_markov_chain_simulation" title="Permalink to this definition">¶</a></dt>
<dd><p>Do Markov simlations to predict cell transition after perturbation.
The transition probability between cells has been calculated
based on simulated gene expression values in the signal propagation process.
The cell state transition will be simulated based on the probability.
You can simulate the process multiple times to get a robust outcome.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n_steps</strong> (<em>int</em>) – steps for Markov simulation. This value is equivalent to the amount of time after perturbation.</p></li>
<li><p><strong>n_duplication</strong> (<em>int</em>) – the number for multiple calculations.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Oracle.simulate_shift">
<code class="sig-name descname">simulate_shift</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">perturb_condition</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">GRN_unit</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">n_propagation</span><span class="o">=</span><span class="default_value">3</span></em>, <em class="sig-param"><span class="n">ignore_warning</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Oracle.simulate_shift" title="Permalink to this definition">¶</a></dt>
<dd><p>Simulate signal propagation with GRNs. Please see the CellOracle paper for details.
This function simulates a gene expression pattern in the near future.
Simulated values will be stored in anndata.layers: [“simulated_count”]</p>
<p>The simulation use three types of data.
(1) GRN inference results (coef_matrix).
(2) Perturb_condition: You can set arbitrary perturbation condition.
(3) Gene expression matrix: The simulation starts from imputed gene expression data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>perturb_condition</strong> (<em>dictionary</em>) – condition for perturbation.
if you want to simulate knockout for GeneX, please set [perturb_condition={“GeneX”: 0.0}]
Although you can set any non-negative values for the gene condition, avoid setting biologically infeasible values for the perturb condition.
It is strongly recommended to check gene expression values in your data before selecting the perturb condition.</p></li>
<li><p><strong>GRN_unit</strong> (<em>str</em>) – GRN type. Please select either “whole” or “cluster”. See the documentation of “fit_GRN_for_simulation” for the detailed explanation.</p></li>
<li><p><strong>n_propagation</strong> (<em>int</em>) – Calculation will be performed iteratively to simulate signal propagation in GRN.
You can set the number of steps for this calculation.
With a higher number, the results may recapitulate signal propagation for many genes.
However, a higher number of propagation may cause more error/noise.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Oracle.suggest_mass_thresholds">
<code class="sig-name descname">suggest_mass_thresholds</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">n_suggestion</span><span class="o">=</span><span class="default_value">12</span></em>, <em class="sig-param"><span class="n">s</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">n_col</span><span class="o">=</span><span class="default_value">4</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Oracle.suggest_mass_thresholds" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt id="celloracle.Oracle.summarize_mc_results_by_cluster">
<code class="sig-name descname">summarize_mc_results_by_cluster</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cluster_use</span></em>, <em class="sig-param"><span class="n">random</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Oracle.summarize_mc_results_by_cluster" title="Permalink to this definition">¶</a></dt>
<dd><p>This function summarizes the simulated cell state-transition by groping the results into each cluster.
It returns sumarized results as a pandas.DataFrame.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>cluster_use</strong> (<em>str</em>) – cluster information name in anndata.obs.
You can use any arbitrary cluster information in anndata.obs.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Oracle.to_hdf5">
<code class="sig-name descname">to_hdf5</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">file_path</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Oracle.to_hdf5" title="Permalink to this definition">¶</a></dt>
<dd><p>Save object as hdf5.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>file_path</strong> (<em>str</em>) – file path to save file. Filename needs to end with ‘.celloracle.oracle’</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.Oracle.updateTFinfo_dictionary">
<code class="sig-name descname">updateTFinfo_dictionary</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">TFdict</span><span class="o">=</span><span class="default_value">{}</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.Oracle.updateTFinfo_dictionary" title="Permalink to this definition">¶</a></dt>
<dd><p>Update a TF dictionary.
If a key in the new TF dictionary already exists in the old TF dictionary, old values will be replaced with a new one.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>TFdict</strong> (<em>dictionary</em>) – Python dictionary of TF info.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt id="celloracle.load_hdf5">
<code class="sig-prename descclassname">celloracle.</code><code class="sig-name descname">load_hdf5</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">file_path</span></em>, <em class="sig-param"><span class="n">object_class_name</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.load_hdf5" title="Permalink to this definition">¶</a></dt>
<dd><p>Load an object of celloracle’s custom class that was saved as hdf5.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_path</strong> (<em>str</em>) – file_path.</p></li>
<li><p><strong>object_class_name</strong> (<em>str</em>) – Types of object.
If it is None, object class will be identified from the extension of file_name.
Default is None.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="modules-for-atac-seq-analysis">
<h1>Modules for ATAC-seq analysis<a class="headerlink" href="#modules-for-atac-seq-analysis" title="Permalink to this headline">¶</a></h1>
<blockquote>
<div><div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="celloracle.motif_analysis.html">celloracle.motif_analysis module</a></li>
</ul>
</div>
</div></blockquote>
</div>
<div class="section" id="modules-for-network-analysis">
<h1>Modules for Network analysis<a class="headerlink" href="#modules-for-network-analysis" title="Permalink to this headline">¶</a></h1>
<blockquote>
<div><div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="celloracle.network_analysis.html">celloracle.network_analysis module</a></li>
</ul>
</div>
</div></blockquote>
</div>
<div class="section" id="other-modules">
<h1>Other modules<a class="headerlink" href="#other-modules" title="Permalink to this headline">¶</a></h1>
<blockquote>
<div><div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="celloracle.go_analysis.html">celloracle.go_analysis module</a></li>
<li class="toctree-l1"><a class="reference internal" href="celloracle.utility.html">celloracle.utility module</a></li>
<li class="toctree-l1"><a class="reference internal" href="celloracle.data.html">celloracle.data module</a></li>
<li class="toctree-l1"><a class="reference internal" href="celloracle.data_conversion.html">celloracle.data_conversion module</a></li>
</ul>
</div>
</div></blockquote>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="celloracle.motif_analysis.html" class="btn btn-neutral float-right" title="celloracle.motif_analysis module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="index.html" class="btn btn-neutral float-left" title="API" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, Samantha Morris Lab

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>