

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>celloracle.network_analysis module &mdash; celloracle 0.4.0 documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="celloracle.go_analysis module" href="celloracle.go_analysis.html" />
    <link rel="prev" title="celloracle.motif_analysis module" href="celloracle.motif_analysis.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home" alt="Documentation Home"> celloracle
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../index.html">Welcome to celloracle’s documentation!</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../installation/index.html">Installation</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/index.html">Tutorial</a></li>
</ul>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="index.html">API</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="index.html#command-line-api">Command Line API</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="index.html#python-api">Python API</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="celloracle.html">Custom class in celloracle</a></li>
<li class="toctree-l3"><a class="reference internal" href="celloracle.html#modules-for-atac-seq-analysis">Modules for ATAC-seq analysis</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="celloracle.html#modules-for-network-analysis">Modules for Network analysis</a><ul class="current">
<li class="toctree-l4 current"><a class="current reference internal" href="#">celloracle.network_analysis module</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="celloracle.html#other-modules">Other modules</a></li>
</ul>
</li>
</ul>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../changelog/index.html">Changelog</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../license/index.html">License</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../citation/index.html">Authors and citations</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">celloracle</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="index.html">API</a> &raquo;</li>
        
          <li><a href="celloracle.html">Custom class in celloracle</a> &raquo;</li>
        
      <li>celloracle.network_analysis module</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/modules/celloracle.network_analysis.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="module-celloracle.network_analysis">
<span id="celloracle-network-analysis-module"></span><h1>celloracle.network_analysis module<a class="headerlink" href="#module-celloracle.network_analysis" title="Permalink to this headline">¶</a></h1>
<p>The <a class="reference internal" href="#module-celloracle.network_analysis" title="celloracle.network_analysis"><code class="xref py py-mod docutils literal notranslate"><span class="pre">network_analysis</span></code></a> module implements Network analysis.</p>
<dl class="py class">
<dt id="celloracle.network_analysis.Links">
<em class="property">class </em><code class="sig-prename descclassname">celloracle.network_analysis.</code><code class="sig-name descname">Links</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">name</span></em>, <em class="sig-param"><span class="n">links_dict</span><span class="o">=</span><span class="default_value">{}</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network_analysis.Links" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>This is a class for the processing and visualization of GRNs.
Links object stores cluster-specific GRNs and metadata.
Please use “get_links” function in Oracle object to generate Links object.</p>
<dl class="py attribute">
<dt id="celloracle.network_analysis.Links.links_dict">
<code class="sig-name descname">links_dict</code><a class="headerlink" href="#celloracle.network_analysis.Links.links_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Dictionary that store unprocessed network data.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>dictionary</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="celloracle.network_analysis.Links.filtered_links">
<code class="sig-name descname">filtered_links</code><a class="headerlink" href="#celloracle.network_analysis.Links.filtered_links" title="Permalink to this definition">¶</a></dt>
<dd><p>Dictionary that store filtered network data.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>dictionary</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="celloracle.network_analysis.Links.merged_score">
<code class="sig-name descname">merged_score</code><a class="headerlink" href="#celloracle.network_analysis.Links.merged_score" title="Permalink to this definition">¶</a></dt>
<dd><p>Network scores.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>pandas.dataframe</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="celloracle.network_analysis.Links.cluster">
<code class="sig-name descname">cluster</code><a class="headerlink" href="#celloracle.network_analysis.Links.cluster" title="Permalink to this definition">¶</a></dt>
<dd><p>List of cluster name.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>list of str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="celloracle.network_analysis.Links.name">
<code class="sig-name descname">name</code><a class="headerlink" href="#celloracle.network_analysis.Links.name" title="Permalink to this definition">¶</a></dt>
<dd><p>Name of clustering unit.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="celloracle.network_analysis.Links.palette">
<code class="sig-name descname">palette</code><a class="headerlink" href="#celloracle.network_analysis.Links.palette" title="Permalink to this definition">¶</a></dt>
<dd><p>DataFrame that store color information.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>pandas.dataframe</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.network_analysis.Links.filter_links">
<code class="sig-name descname">filter_links</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">p</span><span class="o">=</span><span class="default_value">0.001</span></em>, <em class="sig-param"><span class="n">weight</span><span class="o">=</span><span class="default_value">'coef_abs'</span></em>, <em class="sig-param"><span class="n">thread_number</span><span class="o">=</span><span class="default_value">10000</span></em>, <em class="sig-param"><span class="n">genelist_source</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">genelist_target</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network_analysis.Links.filter_links" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter network edges.
In most cases, inferred GRN has non-significant random edges.
We have to remove these edges before analyzing the network structure.
You can do the filtering in any of the following ways.</p>
<ol class="arabic simple">
<li><p>Filter based on the p-value of the network edge.
Please enter p-value for thresholding.</p></li>
<li><p>Filter based on network edge number.
If you set the number, network edges will be filtered based on the order of a network score. The top n-th network edges with network weight will remain, and the other edges will be removed.
The network data has several types of network weight, so you have to select which network weight do you want to use.</p></li>
<li><p>Filter based on an arbitrary gene list. You can set a gene list for source nodes or target nodes.</p></li>
</ol>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>p</strong> (<em>float</em>) – threshold for p-value of the network edge.</p></li>
<li><p><strong>weight</strong> (<em>str</em>) – Please select network weight name for the filtering</p></li>
<li><p><strong>genelist_source</strong> (<em>list of str</em>) – gene list to remain in regulatory gene nodes. Default is None.</p></li>
<li><p><strong>genelist_target</strong> (<em>list of str</em>) – gene list to remain in target gene nodes. Default is None.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.network_analysis.Links.get_network_entropy">
<code class="sig-name descname">get_network_entropy</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">value</span><span class="o">=</span><span class="default_value">'coef_abs'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network_analysis.Links.get_network_entropy" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate network entropy scores.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>value</strong> (<em>str</em>) – Default is “coef_abs”.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.network_analysis.Links.get_score">
<code class="sig-name descname">get_score</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">test_mode</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network_analysis.Links.get_score" title="Permalink to this definition">¶</a></dt>
<dd><p>Get several network sores using R libraries.
Make sure all dependent R libraries are installed in your environment before running this function.
You can check the installation for the R libraries by running test_installation() in network_analysis module.</p>
</dd></dl>

<dl class="py method">
<dt id="celloracle.network_analysis.Links.plot_cartography_scatter_per_cluster">
<code class="sig-name descname">plot_cartography_scatter_per_cluster</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">gois</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">clusters</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">scatter</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">kde</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">auto_gene_annot</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">percentile</span><span class="o">=</span><span class="default_value">98</span></em>, <em class="sig-param"><span class="n">args_dot</span><span class="o">=</span><span class="default_value">{'n_levels': 105}</span></em>, <em class="sig-param"><span class="n">args_line</span><span class="o">=</span><span class="default_value">{'c': 'gray'}</span></em>, <em class="sig-param"><span class="n">args_annot</span><span class="o">=</span><span class="default_value">{}</span></em>, <em class="sig-param"><span class="n">save</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network_analysis.Links.plot_cartography_scatter_per_cluster" title="Permalink to this definition">¶</a></dt>
<dd><p>Make a gene network cartography plot.
Please read the original paper describing gene network cartography for more information.
<a class="reference external" href="https://www.nature.com/articles/nature03288">https://www.nature.com/articles/nature03288</a></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>links</strong> (<a class="reference internal" href="celloracle.html#celloracle.Links" title="celloracle.Links"><em>Links</em></a>) – See network_analysis.Links class for detail.</p></li>
<li><p><strong>gois</strong> (<em>list of srt</em>) – List of gene name to highlight.</p></li>
<li><p><strong>clusters</strong> (<em>list of str</em>) – List of cluster name to analyze. If None, all clusters in Links object will be analyzed.</p></li>
<li><p><strong>scatter</strong> (<em>bool</em>) – Whether to make a scatter plot.</p></li>
<li><p><strong>auto_gene_annot</strong> (<em>bool</em>) – Whether to pick up genes to make an annotation.</p></li>
<li><p><strong>percentile</strong> (<em>float</em>) – Genes with a network score above the percentile will be shown with annotation. Default is 98.</p></li>
<li><p><strong>args_dot</strong> (<em>dictionary</em>) – Arguments for scatter plot.</p></li>
<li><p><strong>args_line</strong> (<em>dictionary</em>) – Arguments for lines in cartography plot.</p></li>
<li><p><strong>args_annot</strong> (<em>dictionary</em>) – Arguments for annotation in plots.</p></li>
<li><p><strong>save</strong> (<em>str</em>) – Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if [save=None]. Default is None.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.network_analysis.Links.plot_cartography_term">
<code class="sig-name descname">plot_cartography_term</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">goi</span></em>, <em class="sig-param"><span class="n">save</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network_analysis.Links.plot_cartography_term" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot the gene network cartography term like a heatmap.
Please read the original paper of gene network cartography for the principle of gene network cartography.
<a class="reference external" href="https://www.nature.com/articles/nature03288">https://www.nature.com/articles/nature03288</a></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>links</strong> (<a class="reference internal" href="celloracle.html#celloracle.Links" title="celloracle.Links"><em>Links</em></a>) – See network_analysis.Links class for detail.</p></li>
<li><p><strong>gois</strong> (<em>list of srt</em>) – List of gene name to highlight.</p></li>
<li><p><strong>save</strong> (<em>str</em>) – Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if [save=None]. Default is None.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.network_analysis.Links.plot_degree_distributions">
<code class="sig-name descname">plot_degree_distributions</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">plot_model</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">save</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network_analysis.Links.plot_degree_distributions" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot the network degree distributions (the number of edge per gene).
The network degree will be visualized in both linear scale and log scale.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>links</strong> (<a class="reference internal" href="celloracle.html#celloracle.Links" title="celloracle.Links"><em>Links</em></a>) – See network_analysis.Links class for detail.</p></li>
<li><p><strong>plot_model</strong> (<em>bool</em>) – Whether to plot linear approximation line.</p></li>
<li><p><strong>save</strong> (<em>str</em>) – Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if [save=None]. Default is None.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.network_analysis.Links.plot_network_entropy_distributions">
<code class="sig-name descname">plot_network_entropy_distributions</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">update_network_entropy</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">save</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network_analysis.Links.plot_network_entropy_distributions" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot the distribution for network entropy.
See the CellOracle paper for more detail.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>links</strong> (<em>Links object</em>) – See network_analysis.Links class for detail.</p></li>
<li><p><strong>values</strong> (<em>list of str</em>) – The list of score to visualize. If it is None, all network score (listed above) will be used.</p></li>
<li><p><strong>update_network_entropy</strong> (<em>bool</em>) – Whether to recalculate network entropy.</p></li>
<li><p><strong>save</strong> (<em>str</em>) – Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if [save=None]. Default is None.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.network_analysis.Links.plot_score_comparison_2D">
<code class="sig-name descname">plot_score_comparison_2D</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">value</span></em>, <em class="sig-param"><span class="n">cluster1</span></em>, <em class="sig-param"><span class="n">cluster2</span></em>, <em class="sig-param"><span class="n">percentile</span><span class="o">=</span><span class="default_value">99</span></em>, <em class="sig-param"><span class="n">annot_shifts</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">save</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network_analysis.Links.plot_score_comparison_2D" title="Permalink to this definition">¶</a></dt>
<dd><p>Make a scatter plot that compares specific network scores in two groups.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>links</strong> (<a class="reference internal" href="celloracle.html#celloracle.Links" title="celloracle.Links"><em>Links</em></a>) – See network_analysis.Links class for detail.</p></li>
<li><p><strong>value</strong> (<em>srt</em>) – The network score type.</p></li>
<li><p><strong>cluster1</strong> (<em>str</em>) – Cluster name. Network scores in cluster1 will be visualized in the x-axis.</p></li>
<li><p><strong>cluster2</strong> (<em>str</em>) – Cluster name. Network scores in cluster2 will be visualized in the y-axis.</p></li>
<li><p><strong>percentile</strong> (<em>float</em>) – Genes with a network score above the percentile will be shown with annotation. Default is 99.</p></li>
<li><p><strong>annot_shifts</strong> (<em>(</em><em>float</em><em>, </em><em>float</em><em>)</em>) – Annotation visualization setting.</p></li>
<li><p><strong>save</strong> (<em>str</em>) – Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if [save=None]. Default is None.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.network_analysis.Links.plot_score_discributions">
<code class="sig-name descname">plot_score_discributions</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">values</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">method</span><span class="o">=</span><span class="default_value">'boxplot'</span></em>, <em class="sig-param"><span class="n">save</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network_analysis.Links.plot_score_discributions" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot the distribution of network scores.
An individual data point is a network edge (gene).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>links</strong> (<a class="reference internal" href="celloracle.html#celloracle.Links" title="celloracle.Links"><em>Links</em></a>) – See Links class for details.</p></li>
<li><p><strong>values</strong> (<em>list of str</em>) – The list of score to visualize. If it is None, all of the network score will be used.</p></li>
<li><p><strong>method</strong> (<em>str</em>) – Plotting method. Select either “boxplot” or “barplot”.</p></li>
<li><p><strong>save</strong> (<em>str</em>) – Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if [save=None]. Default is None.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.network_analysis.Links.plot_score_per_cluster">
<code class="sig-name descname">plot_score_per_cluster</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">goi</span></em>, <em class="sig-param"><span class="n">save</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network_analysis.Links.plot_score_per_cluster" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot network score for a gene.
This function visualizes the network score for a specific gene between clusters to get an insight into the dynamics of the gene.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>links</strong> (<a class="reference internal" href="celloracle.html#celloracle.Links" title="celloracle.Links"><em>Links</em></a>) – See network_analysis.Links class for detail.</p></li>
<li><p><strong>goi</strong> (<em>srt</em>) – Gene name.</p></li>
<li><p><strong>save</strong> (<em>str</em>) – Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if [save=None]. Default is None.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.network_analysis.Links.plot_scores_as_rank">
<code class="sig-name descname">plot_scores_as_rank</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cluster</span></em>, <em class="sig-param"><span class="n">n_gene</span><span class="o">=</span><span class="default_value">50</span></em>, <em class="sig-param"><span class="n">save</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network_analysis.Links.plot_scores_as_rank" title="Permalink to this definition">¶</a></dt>
<dd><p>Pick up top n-th genes wich high-network scores and make plots.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>links</strong> (<a class="reference internal" href="celloracle.html#celloracle.Links" title="celloracle.Links"><em>Links</em></a>) – See network_analysis.Links class for detail.</p></li>
<li><p><strong>cluster</strong> (<em>str</em>) – Cluster name to analyze.</p></li>
<li><p><strong>n_gene</strong> (<em>int</em>) – Number of genes to plot. Default is 50.</p></li>
<li><p><strong>save</strong> (<em>str</em>) – Folder path to save plots. If the folder does not exist in the path, the function creates the folder.
Plots will not be saved if [save=None]. Default is None.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="celloracle.network_analysis.Links.to_hdf5">
<code class="sig-name descname">to_hdf5</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">file_path</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network_analysis.Links.to_hdf5" title="Permalink to this definition">¶</a></dt>
<dd><p>Save object as hdf5.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>file_path</strong> (<em>str</em>) – file path to save file. Filename needs to end with ‘.celloracle.links’</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt id="celloracle.network_analysis.draw_network">
<code class="sig-prename descclassname">celloracle.network_analysis.</code><code class="sig-name descname">draw_network</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">linkList</span></em>, <em class="sig-param"><span class="n">return_graph</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network_analysis.draw_network" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot network graph.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>linkList</strong> (<em>pandas.DataFrame</em>) – GRN saved as linkList.</p></li>
<li><p><strong>return_graph</strong> (<em>bool</em>) – Whether to return graph object.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Network X graph objenct.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Graph object</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="celloracle.network_analysis.get_R_path">
<code class="sig-prename descclassname">celloracle.network_analysis.</code><code class="sig-name descname">get_R_path</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network_analysis.get_R_path" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="celloracle.network_analysis.get_links">
<code class="sig-prename descclassname">celloracle.network_analysis.</code><code class="sig-name descname">get_links</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">oracle_object</span></em>, <em class="sig-param"><span class="n">cluster_name_for_GRN_unit</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">alpha</span><span class="o">=</span><span class="default_value">10</span></em>, <em class="sig-param"><span class="n">bagging_number</span><span class="o">=</span><span class="default_value">20</span></em>, <em class="sig-param"><span class="n">verbose_level</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">test_mode</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network_analysis.get_links" title="Permalink to this definition">¶</a></dt>
<dd><p>Make GRN for each cluster and returns results as a Links object.
Several preprocessing should be done before using this function.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>oracle_object</strong> (<a class="reference internal" href="celloracle.html#celloracle.Oracle" title="celloracle.Oracle"><em>Oracle</em></a>) – See Oracle module for detail.</p></li>
<li><p><strong>cluster_name_for_GRN_unit</strong> (<em>str</em>) – Cluster name for GRN calculation. The cluster information should be stored in Oracle.adata.obs.</p></li>
<li><p><strong>alpha</strong> (<em>float</em><em> or </em><em>int</em>) – The strength of regularization.
If you set a lower value, the sensitivity increases, and you can detect weaker network connections. However, there may be more noise.
If you select a higher value, it will reduce the chance of overfitting.</p></li>
<li><p><strong>bagging_number</strong> (<em>int</em>) – The number used in bagging calculation.</p></li>
<li><p><strong>verbose_level</strong> (<em>int</em>) – if [verbose_level&gt;1], most detailed progress information will be shown.
if [verbose_level &gt; 0], one progress bar will be shown.
if [verbose_level == 0], no progress bar will be shown.</p></li>
<li><p><strong>test_mode</strong> (<em>bool</em>) – If test_mode is True, GRN calculation will be done for only one cluster rather than all clusters.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="celloracle.network_analysis.linkList_to_networkgraph">
<code class="sig-prename descclassname">celloracle.network_analysis.</code><code class="sig-name descname">linkList_to_networkgraph</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">filteredlinkList</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network_analysis.linkList_to_networkgraph" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert linkList into Graph object in NetworkX.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>filteredlinkList</strong> (<em>pandas.DataFrame</em>) – GRN saved as linkList.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Network X graph objenct.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Graph object</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="celloracle.network_analysis.load_links">
<code class="sig-prename descclassname">celloracle.network_analysis.</code><code class="sig-name descname">load_links</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">file_path</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network_analysis.load_links" title="Permalink to this definition">¶</a></dt>
<dd><p>Load links object saved as a hdf5 file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>file_path</strong> (<em>str</em>) – file path.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>loaded links object.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="celloracle.html#celloracle.Links" title="celloracle.Links">Links</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="celloracle.network_analysis.set_R_path">
<code class="sig-prename descclassname">celloracle.network_analysis.</code><code class="sig-name descname">set_R_path</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">R_path</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network_analysis.set_R_path" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="celloracle.network_analysis.test_R_libraries_installation">
<code class="sig-prename descclassname">celloracle.network_analysis.</code><code class="sig-name descname">test_R_libraries_installation</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">show_all_stdout</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network_analysis.test_R_libraries_installation" title="Permalink to this definition">¶</a></dt>
<dd><p>CellOracle.network_analysis use several R libraries for network analysis.
This is a test function to check for instalation of the necessary R libraries.</p>
</dd></dl>

<dl class="py function">
<dt id="celloracle.network_analysis.transfer_scores_from_links_to_adata">
<code class="sig-prename descclassname">celloracle.network_analysis.</code><code class="sig-name descname">transfer_scores_from_links_to_adata</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">adata</span></em>, <em class="sig-param"><span class="n">links</span></em>, <em class="sig-param"><span class="n">method</span><span class="o">=</span><span class="default_value">'median'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#celloracle.network_analysis.transfer_scores_from_links_to_adata" title="Permalink to this definition">¶</a></dt>
<dd><p>Transfer the summary of network scores (median or mean) per group from Links object into adata.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>anndata</em>) – anndata</p></li>
<li><p><strong>links</strong> (<a class="reference internal" href="celloracle.html#celloracle.Links" title="celloracle.Links"><em>Links</em></a>) – Likns object</p></li>
<li><p><strong>method</strong> (<em>str</em>) – The method to summarize data.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="celloracle.go_analysis.html" class="btn btn-neutral float-right" title="celloracle.go_analysis module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="celloracle.motif_analysis.html" class="btn btn-neutral float-left" title="celloracle.motif_analysis module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, Samantha Morris Lab

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>