

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Welcome to celloracle’s documentation! &mdash; celloracle 0.1.0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script type="text/javascript" src="_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Installation" href="installation/index.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="#" class="icon icon-home"> celloracle
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Welcome to celloracle’s documentation!</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="installation/index.html">Installation</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="tutorials/index.html">Tutorial</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="modules/index.html">API</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Changelog</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="license/index.html">License</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="citation/index.html">Authors and citations</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="#">celloracle</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="#">Docs</a> &raquo;</li>
        
      <li>Welcome to celloracle’s documentation!</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/index.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="welcome-to-celloracle-s-documentation">
<h1>Welcome to celloracle’s documentation!<a class="headerlink" href="#welcome-to-celloracle-s-documentation" title="Permalink to this headline">¶</a></h1>
<p>CellOracle is a python library for the analysis of Gene Regulatory Network with single cell data.</p>
<p>Source code are available at <a class="reference external" href="https://github.com/morris-lab/CellOracle">celloracle  git hub repository</a></p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<div class="line-block">
<div class="line"><strong>For colleagues in Morris Lab</strong></div>
<div class="line"><strong>Thank you for your kind help!!</strong></div>
</div>
<div class="line-block">
<div class="line">Documentation is also available as a word and pdf file.</div>
</div>
<p><a class="reference download internal" download="" href="_downloads/2b68115d706e6ca9c7d63e468f820981/celloracle.pdf"><code class="xref download docutils literal notranslate"><span class="pre">pdf</span> <span class="pre">documentation</span></code></a></p>
<dl class="simple">
<dt><a class="reference download internal" download="" href="_downloads/be48a717cb6d826b0fff88ea52479901/documentation_microsoft_word.docx"><code class="xref download docutils literal notranslate"><span class="pre">word</span> <span class="pre">documentation</span></code></a></dt><dd><p>The word documentation does not include figures.</p>
</dd>
</dl>
<div class="line-block">
<div class="line"><strong>Could you please add comments on the file and send it back to Kenji?</strong></div>
<div class="line">I really appreciate your support!</div>
<div class="line"><br /></div>
<div class="line">Thanks,</div>
<div class="line">Kenji</div>
</div>
<blockquote>
<div><p>(This message will be deleted when we launch celloracle.)</p>
</div></blockquote>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>CellOracle is still under development. It is alpha version and functions in this package may change in the future release.</p>
</div>
</div>
<div class="section" id="contents">
<h1>Contents<a class="headerlink" href="#contents" title="Permalink to this headline">¶</a></h1>
<div class="toctree-wrapper compound">
</div>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="installation/index.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="installation/index.html#docker-image">Docker image</a></li>
<li class="toctree-l2"><a class="reference internal" href="installation/index.html#system-requirements">System Requirements</a></li>
<li class="toctree-l2"><a class="reference internal" href="installation/index.html#python-requirements">Python Requirements</a><ul>
<li class="toctree-l3"><a class="reference internal" href="installation/index.html#optional-make-a-new-environment">0. (Optional) Make a new environment</a></li>
<li class="toctree-l3"><a class="reference internal" href="installation/index.html#add-conda-channels">1. Add conda channels</a></li>
<li class="toctree-l3"><a class="reference internal" href="installation/index.html#install-velocyto">2. Install velocyto</a></li>
<li class="toctree-l3"><a class="reference internal" href="installation/index.html#install-scanpy">3. Install scanpy</a></li>
<li class="toctree-l3"><a class="reference internal" href="installation/index.html#install-gimmemotifs">4. Install gimmemotifs</a></li>
<li class="toctree-l3"><a class="reference internal" href="installation/index.html#install-other-python-libraries">5. Install other python libraries</a></li>
<li class="toctree-l3"><a class="reference internal" href="installation/index.html#install-celloracle-from-github">6. install celloracle from github</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="installation/index.html#r-requirements">R requirements</a><ul>
<li class="toctree-l3"><a class="reference internal" href="installation/index.html#id3">Seurat</a></li>
<li class="toctree-l3"><a class="reference internal" href="installation/index.html#id5">Cicero</a></li>
<li class="toctree-l3"><a class="reference internal" href="installation/index.html#id7">igraph</a></li>
<li class="toctree-l3"><a class="reference internal" href="installation/index.html#id9">linkcomm</a></li>
<li class="toctree-l3"><a class="reference internal" href="installation/index.html#id11">rnetcarto</a></li>
<li class="toctree-l3"><a class="reference internal" href="installation/index.html#check-installation">Check installation</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="tutorials/index.html">Tutorial</a><ul>
<li class="toctree-l2"><a class="reference internal" href="tutorials/atac.html">1. ATAC-seq data preprocessing</a><ul>
<li class="toctree-l3"><a class="reference internal" href="tutorials/atac.html#a-extract-tf-binding-information-from-scatac-seq-data">A. Extract TF binding information from scATAC-seq data</a><ul>
<li class="toctree-l4"><a class="reference internal" href="notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/01_atacdata_to_cicero.html">0. Import library</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/01_atacdata_to_cicero.html#1.-Prepare-data">1. Prepare data</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/01_atacdata_to_cicero.html#2.-Load-data-and-make-Cell-Data-Set-(CDS)-object">2. Load data and make Cell Data Set (CDS) object</a><ul>
<li class="toctree-l5"><a class="reference internal" href="notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/01_atacdata_to_cicero.html#2.1.-Process-data-to-make-CDS-object">2.1. Process data to make CDS object</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/01_atacdata_to_cicero.html#3.-Qauality-check-and-Filtering">3. Qauality check and Filtering</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/01_atacdata_to_cicero.html#4.-Process-cicero-CDS-object">4. Process cicero-CDS object</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/01_atacdata_to_cicero.html#5.-Run-cicero-to-get-cis-regulatory-connection-scores">5. Run cicero to get cis-regulatory connection scores</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/01_atacdata_to_cicero.html#6.-Save-results-for-next-step">6. Save results for next step</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/02_preprocess_peak_data.html">0. Import libraries</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/02_preprocess_peak_data.html#1.-Load-data-made-with-cicero">1. Load data made with cicero</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/02_preprocess_peak_data.html#2.-Make-TSS-annotation">2. Make TSS annotation</a><ul>
<li class="toctree-l5"><a class="reference internal" href="notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/02_preprocess_peak_data.html#IMPORTANT:-Please-make-sure-that-you-are-setting-correct-reference-genoms.">IMPORTANT: Please make sure that you are setting correct reference genoms.</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/02_preprocess_peak_data.html#3.-Integrate-TSS-info-and-cicero-connections">3. Integrate TSS info and cicero connections</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/02_preprocess_peak_data.html#4.-Filter-peaks">4. Filter peaks</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/02_preprocess_peak_data.html#5.-Save-data">5. Save data</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="tutorials/atac.html#b-extract-tf-binding-information-from-bulk-atac-seq-data-or-chip-seq-data">B. Extract TF binding information from bulk ATAC-seq data or Chip-seq data</a><ul>
<li class="toctree-l4"><a class="reference internal" href="notebooks/01_ATAC-seq_data_processing/option2_Bulk_ATAC-seq_data/01_preprocess_Bulk_ATAC_seq_peak_data.html">0. Import libraries</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/01_ATAC-seq_data_processing/option2_Bulk_ATAC-seq_data/01_preprocess_Bulk_ATAC_seq_peak_data.html#1.-Load-bed-file">1. Load bed file</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/01_ATAC-seq_data_processing/option2_Bulk_ATAC-seq_data/01_preprocess_Bulk_ATAC_seq_peak_data.html#2.-Make-TSS-annotation">2. Make TSS annotation</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/01_ATAC-seq_data_processing/option2_Bulk_ATAC-seq_data/01_preprocess_Bulk_ATAC_seq_peak_data.html#3.-Save-data">3. Save data</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="tutorials/motifscan.html">2. Transcription factor binding motif scan</a><ul>
<li class="toctree-l3"><a class="reference internal" href="notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_190901.html">0. Import libraries</a></li>
<li class="toctree-l3"><a class="reference internal" href="notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_190901.html#1.-Load-data">1. Load data</a></li>
<li class="toctree-l3"><a class="reference internal" href="notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_190901.html#2.-Check-data">2. Check data</a><ul>
<li class="toctree-l4"><a class="reference internal" href="notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_190901.html#2.1.-Remove-short-peaks">2.1. Remove short peaks</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_190901.html#3.-Instantiate-TFinfo-object-and-search-for-TF-binding-motifs">3. Instantiate TFinfo object and search for TF binding motifs</a></li>
<li class="toctree-l3"><a class="reference internal" href="notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_190901.html#3.1-check-reference-genome-installation">3.1 check reference genome installation</a><ul>
<li class="toctree-l4"><a class="reference internal" href="notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_190901.html#3.2.-Install-reference-genome-(if-refgenome-is-not-installed)">3.2. Install reference genome (if refgenome is not installed)</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_190901.html#4.-Scan-motifs-and-save-object">4. Scan motifs and save object</a></li>
<li class="toctree-l3"><a class="reference internal" href="notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_190901.html#5.-Filtering-motifs">5. Filtering motifs</a></li>
<li class="toctree-l3"><a class="reference internal" href="notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_190901.html#6.-Get-Final-results">6. Get Final results</a><ul>
<li class="toctree-l4"><a class="reference internal" href="notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_190901.html#6.1.-Get-resutls-as-a-dictionary">6.1. Get resutls as a dictionary</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_190901.html#6.2.-Get-results-as-a-dataframe">6.2. Get results as a dataframe</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_190901.html#7.-Save-TFinfo-as-dictionary-or-dataframe">7. Save TFinfo as dictionary or dataframe</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="tutorials/scrnaprocess.html">3. Single-cell RNA-seq data preprocessing</a><ul>
<li class="toctree-l3"><a class="reference internal" href="tutorials/scrnaprocess.html#a-scrna-seq-data-preprocessing-with-scanpy">A. scRNA-seq data preprocessing with scanpy</a><ul>
<li class="toctree-l4"><a class="reference internal" href="notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data.html">0. Import libraries</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data.html#1.-Load-data">1. Load data</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data.html#2.-Filtering">2. Filtering</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data.html#3.-Normalization">3. Normalization</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data.html#4.-Identification-of-highly-variable-genes">4. Identification of highly variable genes</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data.html#5.-Log-transformation">5. Log transformation</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data.html#6.-Dimensional-reduction">6. Dimensional reduction</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data.html#7.-Clustering">7. Clustering</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data.html#(Optional)-Re-calculate-Dimensional-reduction-graph">(Optional) Re-calculate Dimensional reduction graph</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data.html#8.-Check-data">8. Check data</a><ul>
<li class="toctree-l5"><a class="reference internal" href="notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data.html#8.1.-Visualize-marker-gene-expression">8.1. Visualize marker gene expression</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data.html#8.-Make-annotation-for-cluster">8. Make annotation for cluster</a><ul>
<li class="toctree-l5"><a class="reference internal" href="notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data.html#8.1.-Make-annotation-(1)">8.1. Make annotation (1)</a><ul>
<li class="toctree-l6"><a class="reference internal" href="notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data.html#!!-Please-change-the-dictionary-below-depending-on-the-clustering-results.-The-results-may-change-depending-on-the-execution-environment.">!! Please change the dictionary below depending on the clustering results. The results may change depending on the execution environment.</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data.html#8.2.-Make-annotation-(2)">8.2. Make annotation (2)</a><ul>
<li class="toctree-l6"><a class="reference internal" href="notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data.html#id1">!! Please change the dictionary below depending on the clustering results. The results may change depending on the execution environment.</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data.html#9.-(Option)-Subset-cells">9. (Option) Subset cells</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/03_scRNA-seq_data_preprocessing/scanpy_preprocessing_with_Paul_etal_2015_data.html#10.-Save-data">10. Save data</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="tutorials/scrnaprocess.html#b-scrna-seq-data-preprocessing-with-seurat">B. scRNA-seq data preprocessing with Seurat</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="tutorials/networkanalysis.html">4. Network analysis</a><ul>
<li class="toctree-l3"><a class="reference internal" href="notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data.html">0. Import libraries</a><ul>
<li class="toctree-l4"><a class="reference internal" href="notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data.html#0.1.-Check-installation">0.1. Check installation</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data.html#0.2.-Make-a-folder-to-save-graph">0.2. Make a folder to save graph</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data.html#1.-Load-data">1. Load data</a><ul>
<li class="toctree-l4"><a class="reference internal" href="notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data.html#1.1.-Load-processed-gene-expression-data-(anndata)">1.1. Load processed gene expression data (anndata)</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data.html#1.2.-Load-TF-data.">1.2. Load TF data.</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data.html#2.-Initiate-Oracle-object">2. Initiate Oracle object</a><ul>
<li class="toctree-l4"><a class="reference internal" href="notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data.html#2.1.-load-gene-expression-data-into-oracle-object.">2.1. load gene expression data into oracle object.</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data.html#2.2.-Load-TFinfo-into-oracle-object">2.2. Load TFinfo into oracle object</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data.html#2.3.-(Optional)-Add-TF-info-manually">2.3. (Optional) Add TF info manually</a><ul>
<li class="toctree-l5"><a class="reference internal" href="notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data.html#2.3.1.-Make-TF-info-dictionary-manually">2.3.1. Make TF info dictionary manually</a></li>
<li class="toctree-l5"><a class="reference internal" href="notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data.html#2.3.2.-Add-TF-informatio-dictionary-into-the-oracle-object">2.3.2. Add TF informatio dictionary into the oracle object</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data.html#3.-Knn-imputation">3. Knn imputation</a><ul>
<li class="toctree-l4"><a class="reference internal" href="notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data.html#3.1.-PCA">3.1. PCA</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data.html#3.2.-KNN-imputation">3.2. KNN imputation</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data.html#4.-Save-and-Load.">4. Save and Load.</a></li>
<li class="toctree-l3"><a class="reference internal" href="notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data.html#4.-GRN-calculation">4. GRN calculation</a><ul>
<li class="toctree-l4"><a class="reference internal" href="notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data.html#4.1.-Get-GRNs">4.1. Get GRNs</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data.html#4.2.-(Optional)-Export-GRNs">4.2. (Optional) Export GRNs</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data.html#4.3.-(Optional)-Change-order">4.3. (Optional) Change order</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data.html#5.-Network-preprocessing">5. Network preprocessing</a><ul>
<li class="toctree-l4"><a class="reference internal" href="notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data.html#5.1.-Filter-network-edges">5.1. Filter network edges</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data.html#5.2.-Degree-distribution">5.2. Degree distribution</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data.html#5.3.-Calculate-netowrk-score">5.3. Calculate netowrk score</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data.html#5.4.-Save">5.4. Save</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data.html#6.-Network-analysis;-Network-score-for-each-gene">6. Network analysis; Network score for each gene</a><ul>
<li class="toctree-l4"><a class="reference internal" href="notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data.html#6.1.-Network-score-in-each-cluster">6.1. Network score in each cluster</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data.html#6.2.-Network-score-comparison-between-two-clusters">6.2. Network score comparison between two clusters</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data.html#6.3.-Network-score-dynamics">6.3. Network score dynamics</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data.html#6.4.-Gene-cartography-analysis">6.4. Gene cartography analysis</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data.html#7.-Network-analysis;-network-score-distribution">7. Network analysis; network score distribution</a><ul>
<li class="toctree-l4"><a class="reference internal" href="notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data.html#7.1.-Distribution-of-network-degree">7.1. Distribution of network degree</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/04_Network_analysis/Network_analysis_with_with_Paul_etal_2015_data.html#7.2.-Distribution-of-netowrk-entolopy">7.2. Distribution of netowrk entolopy</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="tutorials/simulation.html">5. Simulation with GRNs</a><ul>
<li class="toctree-l3"><a class="reference internal" href="notebooks/05_simulation/Gata1_KO_simulation_with_with_Paul_etal_2015_data.html">0. Import libraries</a><ul>
<li class="toctree-l4"><a class="reference internal" href="notebooks/05_simulation/Gata1_KO_simulation_with_with_Paul_etal_2015_data.html#0.1.-Import-public-libraries">0.1. Import public libraries</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/05_simulation/Gata1_KO_simulation_with_with_Paul_etal_2015_data.html#0.1.-Make-a-folder-to-save-graph">0.1. Make a folder to save graph</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="notebooks/05_simulation/Gata1_KO_simulation_with_with_Paul_etal_2015_data.html#1.-Load-data">1. Load data</a><ul>
<li class="toctree-l4"><a class="reference internal" href="notebooks/05_simulation/Gata1_KO_simulation_with_with_Paul_etal_2015_data.html#1.1.-Load-processed-oracle-object">1.1. Load processed oracle object</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/05_simulation/Gata1_KO_simulation_with_with_Paul_etal_2015_data.html#1.2.-Load-inferred-GRNs">1.2. Load inferred GRNs</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="notebooks/05_simulation/Gata1_KO_simulation_with_with_Paul_etal_2015_data.html#3.-Make-predictive-models-for-simulation">3. Make predictive models for simulation</a></li>
<li class="toctree-l3"><a class="reference internal" href="notebooks/05_simulation/Gata1_KO_simulation_with_with_Paul_etal_2015_data.html#4.-in-silico-Perturbation-simulation">4. in silico Perturbation-simulation</a><ul>
<li class="toctree-l4"><a class="reference internal" href="notebooks/05_simulation/Gata1_KO_simulation_with_with_Paul_etal_2015_data.html#4.1.-Check-gene-expression-pattern.">4.1. Check gene expression pattern.</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/05_simulation/Gata1_KO_simulation_with_with_Paul_etal_2015_data.html#4.1.-calculate-future-gene-expression-after-perturbation.">4.1. calculate future gene expression after perturbation.</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/05_simulation/Gata1_KO_simulation_with_with_Paul_etal_2015_data.html#4.2.-calculate-transition-probability-between-cells">4.2. calculate transition probability between cells</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/05_simulation/Gata1_KO_simulation_with_with_Paul_etal_2015_data.html#4.3.-Visualization">4.3. Visualization</a><ul>
<li class="toctree-l5"><a class="reference internal" href="notebooks/05_simulation/Gata1_KO_simulation_with_with_Paul_etal_2015_data.html#4.3.1.-Detailed-directed-trajectory-graph">4.3.1. Detailed directed trajectory graph</a></li>
<li class="toctree-l5"><a class="reference internal" href="notebooks/05_simulation/Gata1_KO_simulation_with_with_Paul_etal_2015_data.html#4.3.2.-Grid-graph">4.3.2. Grid graph</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/05_simulation/Gata1_KO_simulation_with_with_Paul_etal_2015_data.html#4.4.-Markov-simulation-to-analyze-the-effects-of-perturbation-on-cell-fate-transition">4.4. Markov simulation to analyze the effects of perturbation on cell fate transition</a><ul>
<li class="toctree-l5"><a class="reference internal" href="notebooks/05_simulation/Gata1_KO_simulation_with_with_Paul_etal_2015_data.html#4.4.1.-Do-simulation">4.4.1. Do simulation</a></li>
<li class="toctree-l5"><a class="reference internal" href="notebooks/05_simulation/Gata1_KO_simulation_with_with_Paul_etal_2015_data.html#4.4.2.-Check-the-results-of-the-simulation-for-specific-cells">4.4.2. Check the results of the simulation for specific cells</a></li>
<li class="toctree-l5"><a class="reference internal" href="notebooks/05_simulation/Gata1_KO_simulation_with_with_Paul_etal_2015_data.html#4.4.3.-Summarize-the-results-of-simulation-by-plotting-sankey-diagram">4.4.3. Summarize the results of simulation by plotting sankey diagram</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="modules/index.html">API</a><ul>
<li class="toctree-l2"><a class="reference internal" href="modules/index.html#command-line-api">Command Line API</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules/index.html#python-api">Python API</a><ul>
<li class="toctree-l3"><a class="reference internal" href="modules/celloracle.html">Custom class in celloracle</a></li>
<li class="toctree-l3"><a class="reference internal" href="modules/celloracle.html#modules-for-atac-seq-analysis">Modules for ATAC-seq analysis</a><ul>
<li class="toctree-l4"><a class="reference internal" href="modules/celloracle.motif_analysis.html">celloracle.motif_analysis module</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="modules/celloracle.html#modules-for-network-analysis">Modules for Network analysis</a><ul>
<li class="toctree-l4"><a class="reference internal" href="modules/celloracle.network_analysis.html">celloracle.network_analysis module</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="modules/celloracle.html#other-modules">Other modules</a><ul>
<li class="toctree-l4"><a class="reference internal" href="modules/celloracle.go_analysis.html">celloracle.go_analysis module</a></li>
<li class="toctree-l4"><a class="reference internal" href="modules/celloracle.utility.html">celloracle.utility module</a></li>
<li class="toctree-l4"><a class="reference internal" href="modules/celloracle.data.html">celloracle.data module</a></li>
<li class="toctree-l4"><a class="reference internal" href="modules/celloracle.data_conversion.html">celloracle.data_conversion module</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Changelog</a></li>
</ul>
</div>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="license/index.html">License</a></li>
</ul>
</div>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="citation/index.html">Authors and citations</a><ul>
<li class="toctree-l2"><a class="reference internal" href="citation/index.html#cite-celloracle">Cite celloracle</a></li>
<li class="toctree-l2"><a class="reference internal" href="citation/index.html#celloracle-software-development">celloracle software development</a></li>
</ul>
</li>
</ul>
</div>
</div>
<div class="section" id="indices-and-tables">
<h1>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><p><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></p></li>
<li><p><a class="reference internal" href="py-modindex.html"><span class="std std-ref">Module Index</span></a></p></li>
<li><p><a class="reference internal" href="search.html"><span class="std std-ref">Search Page</span></a></p></li>
</ul>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="installation/index.html" class="btn btn-neutral float-right" title="Installation" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Samantha Morris Lab

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>