

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>5. Simulation with GRNs &mdash; celloracle 0.3.5 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../_static/jquery.js"></script>
        <script type="text/javascript" src="../_static/underscore.js"></script>
        <script type="text/javascript" src="../_static/doctools.js"></script>
        <script type="text/javascript" src="../_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="0. Import libraries" href="../notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data.html" />
    <link rel="prev" title="0. Import libraries" href="../notebooks/04_Network_analysis/Network_analysis_with_Paul_etal_2015_data.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> celloracle
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../index.html">Welcome to celloracle’s documentation!</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../installation/index.html">Installation</a></li>
</ul>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Tutorial</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="atac.html">1. ATAC-seq data preprocessing</a></li>
<li class="toctree-l2"><a class="reference internal" href="motifscan.html">2. Transcription factor binding motif scan</a></li>
<li class="toctree-l2"><a class="reference internal" href="scrnaprocess.html">3. Single-cell RNA-seq data preprocessing</a></li>
<li class="toctree-l2"><a class="reference internal" href="networkanalysis.html">4. Network analysis</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">5. Simulation with GRNs</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data.html">0. Import libraries</a></li>
<li class="toctree-l3"><a class="reference internal" href="../notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data.html#1.-Load-data">1. Load data</a></li>
<li class="toctree-l3"><a class="reference internal" href="../notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data.html#2.-Make-predictive-models-for-simulation">2. Make predictive models for simulation</a></li>
<li class="toctree-l3"><a class="reference internal" href="../notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data.html#3.-in-silico-Perturbation-simulation">3. in silico Perturbation-simulation</a></li>
<li class="toctree-l3"><a class="reference internal" href="../notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data.html#4.-Visualization">4. Visualization</a></li>
<li class="toctree-l3"><a class="reference internal" href="../notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data.html#5.-Markov-simulation-to-analyze-the-effects-of-perturbation-on-cell-fate-transition">5. Markov simulation to analyze the effects of perturbation on cell fate transition</a></li>
<li class="toctree-l3"><a class="reference internal" href="../notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data.html#5.1.-Do-Markov-simulation">5.1. Do Markov simulation</a></li>
</ul>
</li>
</ul>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../modules/index.html">API</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../changelog/index.html">Changelog</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../license/index.html">License</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../citation/index.html">Authors and citations</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">celloracle</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
          <li><a href="index.html">Tutorial</a> &raquo;</li>
        
      <li>5. Simulation with GRNs</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/tutorials/simulation.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="simulation-with-grns">
<span id="simulation"></span><h1>5. Simulation with GRNs<a class="headerlink" href="#simulation-with-grns" title="Permalink to this headline">¶</a></h1>
<p><code class="docutils literal notranslate"><span class="pre">celloracle</span></code> leverage GRNs to simulate signal propagation inside a cell.
We can estimate the effect of gene perturbation by the simulation with GRNs.</p>
<p>Additonally, we will combine the signal propagation simulation with a cell state transition simulation. The latter simulation is performed by a python library for RNA-velocity analysis, called <code class="docutils literal notranslate"><span class="pre">velocyto</span></code> .
This analysis may provide an insight into a complex system how TF controls enormous target genes to determines cell fate.</p>
<p>Python notebook</p>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="../notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data.html">0. Import libraries</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data.html#0.1.-Import-public-libraries">0.1. Import public libraries</a></li>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data.html#0.1.-Make-a-folder-to-save-graph">0.1. Make a folder to save graph</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data.html#1.-Load-data">1. Load data</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data.html#1.1.-Load-processed-oracle-object">1.1. Load processed oracle object</a></li>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data.html#1.2.-Load-inferred-GRNs">1.2. Load inferred GRNs</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data.html#2.-Make-predictive-models-for-simulation">2. Make predictive models for simulation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data.html#3.-in-silico-Perturbation-simulation">3. in silico Perturbation-simulation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data.html#3.1.-Check-gene-expression-pattern.">3.1. Check gene expression pattern.</a></li>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data.html#3.2.-calculate-future-gene-expression-after-perturbation.">3.2. calculate future gene expression after perturbation.</a></li>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data.html#3.3.-calculate-transition-probability-between-cells">3.3. calculate transition probability between cells</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data.html#4.-Visualization">4. Visualization</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data.html#4.1.-Detailed-directed-trajectory-graph">4.1. Detailed directed trajectory graph</a></li>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data.html#4.2.-Grid-graph">4.2. Grid graph</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data.html#5.-Markov-simulation-to-analyze-the-effects-of-perturbation-on-cell-fate-transition">5. Markov simulation to analyze the effects of perturbation on cell fate transition</a></li>
<li class="toctree-l1"><a class="reference internal" href="../notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data.html#5.1.-Do-Markov-simulation">5.1. Do Markov simulation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data.html#5.2.-Check-the-results-of-the-simulation-for-specific-cells">5.2. Check the results of the simulation for specific cells</a></li>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data.html#5.3.-Summarize-the-results-of-simulation-by-plotting-sankey-diagram">5.3. Summarize the results of simulation by plotting sankey diagram</a></li>
</ul>
</li>
</ul>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data.html" class="btn btn-neutral float-right" title="0. Import libraries" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="../notebooks/04_Network_analysis/Network_analysis_with_Paul_etal_2015_data.html" class="btn btn-neutral float-left" title="0. Import libraries" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020, Samantha Morris Lab

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>