

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  <script type="text/javascript">

      var _gaq = _gaq || [];
      _gaq.push(['_setAccount', 'UA-214970391-3']);
      _gaq.push(['_trackPageview']);

      (function() {
        var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
        ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
        var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
      })();
    </script>
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Overview &mdash; celloracle 0.9.1 documentation</title>
  

  
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="1. scRNA-seq data preparation" href="../../tutorials/scrnaprocess.html" />
    <link rel="prev" title="1. In silico gene perturbation with GRNs" href="../../tutorials/simulation.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../index.html" class="icon icon-home" alt="Documentation Home"> celloracle
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../index.html">Welcome to CellOracle’s documentation!</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../installation/index.html">Installation</a></li>
</ul>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="../../tutorials/index.html">Tutorial</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../../tutorials/index.html#what-the-tutorial-covers">What the tutorial covers</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../tutorials/index.html#prerequisites">Prerequisites</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../tutorials/index.html#code-and-data-availability">Code and data availability</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../tutorials/index.html#getting-started">Getting started</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../../tutorials/index.html#index">Index</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="../../tutorials/index.html#grn-model-construction-and-network-analysis">GRN model construction and network analysis</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="../../tutorials/index.html#in-silico-gene-perturbation-with-grns">In silico gene perturbation with GRNs</a><ul class="current">
<li class="toctree-l4 current"><a class="reference internal" href="../../tutorials/simulation.html">1. In silico gene perturbation with GRNs</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../tutorials/index.html#prepare-input-data">Prepare input data</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../tutorials/index.html#tf-motif-scan-for-base-grn-construction">TF motif scan for base-GRN construction</a></li>
</ul>
</li>
</ul>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../modules/index.html">API</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../changelog/index.html">Changelog</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../license/index.html">License</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../citation/index.html">Authors and citations</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../contact/index.html">Contact</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">celloracle</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../../tutorials/index.html">Tutorial</a> &raquo;</li>
        
          <li><a href="../../tutorials/simulation.html"><span class="section-number">1. </span>In silico gene perturbation with GRNs</a> &raquo;</li>
        
      <li>Overview</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../../_sources/notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data.ipynb.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container,
div.nbinput.container div.prompt,
div.nbinput.container div.input_area,
div.nbinput.container div[class*=highlight],
div.nbinput.container div[class*=highlight] pre,
div.nboutput.container,
div.nboutput.container div.prompt,
div.nboutput.container div.output_area,
div.nboutput.container div[class*=highlight],
div.nboutput.container div[class*=highlight] pre {
    background: none;
    border: none;
    padding: 0 0;
    margin: 0;
    box-shadow: none;
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    background: #f5f5f5;
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt a.copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="Overview">
<h1>Overview<a class="headerlink" href="#Overview" title="Permalink to this headline">¶</a></h1>
<p>This notebook describes how to run in silico TF perturbations with the GRN models. Please read our paper to learn more about the CellOracle algorithm.</p>
<div class="section" id="Notebook-file">
<h2>Notebook file<a class="headerlink" href="#Notebook-file" title="Permalink to this headline">¶</a></h2>
<p>Notebook file is available at CellOracle GitHub. <a class="reference external" href="https://github.com/morris-lab/CellOracle/blob/master/docs/notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data.ipynb">https://github.com/morris-lab/CellOracle/blob/master/docs/notebooks/05_simulation/Gata1_KO_simulation_with_Paul_etal_2015_data.ipynb</a></p>
</div>
<div class="section" id="Data">
<h2>Data<a class="headerlink" href="#Data" title="Permalink to this headline">¶</a></h2>
<p>In this notebook, CellOracle uses two types of input data.</p>
<ul class="simple">
<li><p><strong>Input data 1: Oracle object</strong>. Please look at the previous notebook to learn how to make an <code class="docutils literal notranslate"><span class="pre">Oracle</span></code> object from scRNA-seq. <a class="reference external" href="https://morris-lab.github.io/CellOracle.documentation/notebooks/04_Network_analysis/Network_analysis_with_Paul_etal_2015_data.html">https://morris-lab.github.io/CellOracle.documentation/notebooks/04_Network_analysis/Network_analysis_with_Paul_etal_2015_data.html</a></p></li>
</ul>
<p>In this tutorial, we will use the mouse hematopoiesis demo data (Paul et al., 2015). We can load the demo <code class="docutils literal notranslate"><span class="pre">Oracle</span></code> object using the command below.</p>
<p><code class="docutils literal notranslate"><span class="pre">oracle</span> <span class="pre">=</span> <span class="pre">co.data.load_tutorial_oracle_object()</span></code></p>
<ul class="simple">
<li><p><strong>Input data 2: Links object</strong>. The <code class="docutils literal notranslate"><span class="pre">Links</span></code> oobject stores the GRN data used in the simulations. In this tutorial, we use demo GRNs calculated from the hematopoiesis scRNA-seq data above and one of the mouse sci-ATAC-seq atlas base GRNs. We can load the demo <code class="docutils literal notranslate"><span class="pre">Links</span></code> object with this command.</p></li>
</ul>
<p><code class="docutils literal notranslate"><span class="pre">links</span> <span class="pre">=</span> <span class="pre">co.data.load_tutorial_links_object()</span></code></p>
</div>
<div class="section" id="What-you-can-do">
<h2>What you can do<a class="headerlink" href="#What-you-can-do" title="Permalink to this headline">¶</a></h2>
<p>In this notebook, we perform two analyses.</p>
<ol class="arabic simple">
<li><p><strong>in silico TF perturbation</strong> to simulate cell identity shifts. CellOracle uses the GRN model to simulate cell identity shifts in response to TF perturbations. For this analysis, you will need the GRN models from the previous notebook.</p></li>
<li><p><strong>Compare simulation vectors with developmental vectors</strong>. In order to properly interpret the simulation results, it is also important to consider the natural direction of development. First, we will calculate a pseudotime gradient vector field to recapitulate the developmental flow. Then, we will compare the CellOracle TF perturbation vector field with the developmental vector field by calculating the inner product scores. Let’s call the inner product value as <strong>perturbation score (PS)</strong>.
Please see the <strong>step 5.6</strong> for detail.</p></li>
</ol>
</div>
<div class="section" id="Custom-data-class-/-object">
<h2>Custom data class / object<a class="headerlink" href="#Custom-data-class-/-object" title="Permalink to this headline">¶</a></h2>
<p>In this notebook, CellOracle uses four custom classes, <code class="docutils literal notranslate"><span class="pre">Oracle</span></code>, <code class="docutils literal notranslate"><span class="pre">Links</span></code>, <code class="docutils literal notranslate"><span class="pre">Gradient_calculator</span></code>, and <code class="docutils literal notranslate"><span class="pre">Oracle_development_module</span></code>.</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">Oracle</span></code> is the main class in the CellOracle package. It is responsible for most of the calculations during GRN model construction and TF perturbation simulations.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Links</span></code> is a class to store GRN data.</p></li>
<li><p>The <code class="docutils literal notranslate"><span class="pre">Gradient_calculator</span></code> calculates the developmental vector field from the pseudotime results. If you do not have pseudotime data for your trajectory, please see the pseudotime notebook to calculate this information. <a class="reference external" href="https://morris-lab.github.io/CellOracle.documentation/tutorials/pseudotime.html">https://morris-lab.github.io/CellOracle.documentation/tutorials/pseudotime.html</a></p></li>
<li><p>The <code class="docutils literal notranslate"><span class="pre">Oracle_development_module</span></code> integrates the <code class="docutils literal notranslate"><span class="pre">Oracle</span></code> object data and the <code class="docutils literal notranslate"><span class="pre">Gradient_calculator</span></code> object data to analyze how TF perturbation affects on the developmental process. It also has many visualization functions.</p></li>
</ul>
</div>
</div>
<div class="section" id="0.-Import-libraries">
<h1>0. Import libraries<a class="headerlink" href="#0.-Import-libraries" title="Permalink to this headline">¶</a></h1>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">sys</span>

<span class="kn">import</span> <span class="nn">matplotlib.colors</span> <span class="k">as</span> <span class="nn">colors</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>

</pre></div>
</div>
</div>
<p>This notebook was made with celloracle version 0.9.0. <strong>Please use celloracle&gt;=0.9.0.</strong> Otherwise, you may get an error.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">celloracle</span> <span class="k">as</span> <span class="nn">co</span>
<span class="n">co</span><span class="o">.</span><span class="n">__version__</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&#39;0.9.4&#39;
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1">#plt.rcParams[&quot;font.family&quot;] = &quot;arial&quot;</span>
<span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s2">&quot;figure.figsize&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="mi">6</span><span class="p">,</span><span class="mi">6</span><span class="p">]</span>
<span class="o">%</span><span class="k">config</span> InlineBackend.figure_format = &#39;retina&#39;
<span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s2">&quot;savefig.dpi&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">600</span>

<span class="o">%</span><span class="k">matplotlib</span> inline
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Make folder to save plots</span>
<span class="n">save_folder</span> <span class="o">=</span> <span class="s2">&quot;figures&quot;</span>
<span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">save_folder</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="1.-Load-data">
<h1>1. Load data<a class="headerlink" href="#1.-Load-data" title="Permalink to this headline">¶</a></h1>
<p>Load the oracle object. See the previous notebook for information on initializing an <code class="docutils literal notranslate"><span class="pre">Oracle</span></code> object.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Load the tutorial oracle object.</span>
<span class="n">oracle</span> <span class="o">=</span> <span class="n">co</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">load_tutorial_oracle_object</span><span class="p">()</span>
<span class="n">oracle</span>

<span class="c1"># Attention!! Please use the function below when you use your data.</span>
<span class="c1"># oracle = co.load_hdf5(&quot;ORACLE OBJECT PATH&quot;)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Oracle object

Meta data
    celloracle version used for instantiation: 0.6.11
    n_cells: 2671
    n_genes: 1999
    cluster_name: louvain_annot
    dimensional_reduction_name: X_draw_graph_fa
    n_target_genes_in_TFdict: 21259 genes
    n_regulatory_in_TFdict: 1093 genes
    n_regulatory_in_both_TFdict_and_scRNA-seq: 90 genes
    n_target_genes_both_TFdict_and_scRNA-seq: 1850 genes
    k_for_knn_imputation: 66
Status
    Gene expression matrix: Ready
    BaseGRN: Ready
    PCA calculation: Done
    Knn imputation: Done
    GRN calculation for simulation: Not finished
</pre></div></div>
</div>
<p>In the previous notebook, we calculated the GRNs. Now, we will use these GRNs for the perturbation simulations. First, we will import the GRNs from the <code class="docutils literal notranslate"><span class="pre">Links</span></code> object.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Here, we load demo links object for the training purpose.</span>
<span class="n">links</span> <span class="o">=</span> <span class="n">co</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">load_tutorial_links_object</span><span class="p">()</span>

<span class="c1"># Attention!! Please use the function below when you use your data.</span>
<span class="c1"># links = co.load_hdf5(&quot;YOUR LINK OBJCT PATH&quot;)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="2.-Make-predictive-models-for-simulation">
<h1>2. Make predictive models for simulation<a class="headerlink" href="#2.-Make-predictive-models-for-simulation" title="Permalink to this headline">¶</a></h1>
<p>Here, we will need to fit the ridge regression models again. This process will take less time than the GRN inference in the previous notebook, because we are using the filtered GRN models.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">links</span><span class="o">.</span><span class="n">filter_links</span><span class="p">()</span>
<span class="n">oracle</span><span class="o">.</span><span class="n">get_cluster_specific_TFdict_from_Links</span><span class="p">(</span><span class="n">links_object</span><span class="o">=</span><span class="n">links</span><span class="p">)</span>
<span class="n">oracle</span><span class="o">.</span><span class="n">fit_GRN_for_simulation</span><span class="p">(</span><span class="n">alpha</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">use_cluster_specific_TFdict</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
fitting GRN again...
calculating GRN in Ery_0
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "df36c9b3f557404e912faf6c7b8e2b71", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1085 genes
calculating GRN in Ery_1
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "7cb1495b84df45268a7d2ceb01050e57", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1073 genes
calculating GRN in Ery_2
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "bd962454242b4b4d8bceeb94f754f262", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1070 genes
calculating GRN in Ery_3
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "9549c39b89a24291b5cc8e3254bd15fe", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1086 genes
calculating GRN in Ery_4
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "bed3126614d64411b0c854adfd58e4b5", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1087 genes
calculating GRN in Ery_5
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "0bb140efaae64f1abd237033f5a6b226", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1146 genes
calculating GRN in Ery_6
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "974f1dcd5e50453ea0b6c32cc583bcf5", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1095 genes
calculating GRN in Ery_7
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "dcda65446c2b472997077d944fd2607b", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1061 genes
calculating GRN in Ery_8
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "8ea942e672254d0b998c225851ac670d", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1109 genes
calculating GRN in Ery_9
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "384827398e24412baef1ca52381dae8a", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1111 genes
calculating GRN in GMP_0
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "2403954b58f94f7ca77f7bbc9cfaad95", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1094 genes
calculating GRN in GMP_1
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "766c36db7f1b4c76b914defe58d502ca", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1110 genes
calculating GRN in GMP_2
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "293603cc048f457f80ad8fea03c35106", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1106 genes
calculating GRN in GMPl_0
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "17340ec23afb4a3ea5b6628282be1171", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1082 genes
calculating GRN in GMPl_1
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "5ed4bd37cc65419a945d60e87c6edb45", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1089 genes
calculating GRN in Gran_0
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "a4843781e1794b459ca79096228d71d4", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1059 genes
calculating GRN in Gran_1
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "a4ffad7f290d43d99328f163dc179e5d", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1075 genes
calculating GRN in Gran_2
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "1243c140ef5e44739e0a270207a125ef", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1102 genes
calculating GRN in Gran_3
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "7015ed1a77644c2db7b8af152839d056", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1191 genes
calculating GRN in MEP_0
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "215bb5b2f1b24ce2bca0b33771794463", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1142 genes
calculating GRN in Mk_0
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "da155560398242e7b8c11542b88e5d68", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1104 genes
calculating GRN in Mo_0
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "964f98a6aba6486982cf5a9f4a6da283", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1087 genes
calculating GRN in Mo_1
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "c5e7447c4eca4a8c82fa0d1a991815e0", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1112 genes
calculating GRN in Mo_2
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "0c267fbd6c9d4fc8a3d37f34c7d5ff2f", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

genes_in_gem: 1999
models made for 1076 genes
</pre></div></div>
</div>
</div>
<div class="section" id="3.-In-silico-TF-perturbation-analysis">
<h1>3. In silico TF perturbation analysis<a class="headerlink" href="#3.-In-silico-TF-perturbation-analysis" title="Permalink to this headline">¶</a></h1>
<p>Next, we will simulate the TF perturbation effects on cell identity to investigate its potential functions and regulatory mechanisms. Please see the CellOracle paper for more details on scientific rationale.</p>
<p>In this notebook, we will simulate the knockout of the Gata1 gene in the hematopoiesis trajectory.</p>
<p>Previous studies have shown that Gata1 is one of the TFs that regulates cell fate decisions in myeloid progenitors. Additionally, Gata1 has been shown to affect erythroid cell differentiation.</p>
<p>Here, we will use CellOracle to analyze Gata1 and attempt to recapitulate the previous findings from above.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Check gene expression</span>
<span class="n">goi</span> <span class="o">=</span> <span class="s2">&quot;Gata1&quot;</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">draw_graph</span><span class="p">(</span><span class="n">oracle</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="n">goi</span><span class="p">,</span> <span class="n">oracle</span><span class="o">.</span><span class="n">cluster_column_name</span><span class="p">],</span>
                 <span class="n">layer</span><span class="o">=</span><span class="s2">&quot;imputed_count&quot;</span><span class="p">,</span> <span class="n">use_raw</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;viridis&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_17_0.png" class="no-scaled-link" src="../../_images/notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_17_0.png" style="width: 940px; height: 265px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Plot gene expression in histogram</span>
<span class="n">sc</span><span class="o">.</span><span class="n">get</span><span class="o">.</span><span class="n">obs_df</span><span class="p">(</span><span class="n">oracle</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span> <span class="n">keys</span><span class="o">=</span><span class="p">[</span><span class="n">goi</span><span class="p">],</span> <span class="n">layer</span><span class="o">=</span><span class="s2">&quot;imputed_count&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">hist</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_18_0.png" class="no-scaled-link" src="../../_images/notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_18_0.png" style="width: 381px; height: 263px;" />
</div>
</div>
<ul class="simple">
<li><p>You can use any gene expression value in the in silico perturbations, but please avoid extremely high values that are far from the natural gene expression range. The upper limit allowed is twice the maximum gene expression.</p></li>
</ul>
<p>To simulate Gata1 KO, we will set Gata1 expression to 0.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Enter perturbation conditions to simulate signal propagation after the perturbation.</span>
<span class="n">oracle</span><span class="o">.</span><span class="n">simulate_shift</span><span class="p">(</span><span class="n">perturb_condition</span><span class="o">=</span><span class="p">{</span><span class="n">goi</span><span class="p">:</span> <span class="mf">0.0</span><span class="p">},</span>
                      <span class="n">n_propagation</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<ul class="simple">
<li><p>The steps above simulated tge global future gene expression shift after perturbation. This prediction is based on iterative calculations of signal propagation calculations within the GRN. Please look at our paper for more information.</p></li>
<li><p>The next step calculates the probability of cell state transitions based on the simulation data. You can use the transition probabilities between cells to predict how cells will change after a perturbation.</p></li>
<li><p>This transition probability will be used later.</p></li>
</ul>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Get transition probability</span>
<span class="n">oracle</span><span class="o">.</span><span class="n">estimate_transition_prob</span><span class="p">(</span><span class="n">n_neighbors</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span>
                                <span class="n">knn_random</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                                <span class="n">sampled_fraction</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

<span class="c1"># Calculate embedding</span>
<span class="n">oracle</span><span class="o">.</span><span class="n">calculate_embedding_shift</span><span class="p">(</span><span class="n">sigma_corr</span><span class="o">=</span><span class="mf">0.05</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="4.-Visualization">
<h1>4. Visualization<a class="headerlink" href="#4.-Visualization" title="Permalink to this headline">¶</a></h1>
<div class="section" id="Caution:-It-is-very-important-to-find-the-optimal-scale-parameter.">
<h2>Caution: It is very important to find the optimal <code class="docutils literal notranslate"><span class="pre">scale</span></code> parameter.<a class="headerlink" href="#Caution:-It-is-very-important-to-find-the-optimal-scale-parameter." title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>We will need to adjust the <code class="docutils literal notranslate"><span class="pre">scale</span></code> parameter. Please seek to find the optimal <code class="docutils literal notranslate"><span class="pre">scale</span></code> parameter for the data based on your data.</p></li>
<li><p>If the vectors are not visible, you can try a smaller <code class="docutils literal notranslate"><span class="pre">scale</span></code> parameter to magnify the vector length. However, if you see large vectors in the randomized results (right panel), it means that the scale parameter is too small.</p></li>
</ul>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span>  <span class="n">figsize</span><span class="o">=</span><span class="p">[</span><span class="mi">13</span><span class="p">,</span> <span class="mi">6</span><span class="p">])</span>

<span class="n">scale</span> <span class="o">=</span> <span class="mi">25</span>
<span class="c1"># Show quiver plot</span>
<span class="n">oracle</span><span class="o">.</span><span class="n">plot_quiver</span><span class="p">(</span><span class="n">scale</span><span class="o">=</span><span class="n">scale</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Simulated cell identity shift vector: </span><span class="si">{</span><span class="n">goi</span><span class="si">}</span><span class="s2"> KO&quot;</span><span class="p">)</span>

<span class="c1"># Show quiver plot that was calculated with randomized graph.</span>
<span class="n">oracle</span><span class="o">.</span><span class="n">plot_quiver_random</span><span class="p">(</span><span class="n">scale</span><span class="o">=</span><span class="n">scale</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Randomized simulation vector&quot;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_25_0.png" class="no-scaled-link" src="../../_images/notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_25_0.png" style="width: 739px; height: 356px;" />
</div>
</div>
<div class="section" id="4.2.-Vector-field-graph">
<h3>4.2. Vector field graph<a class="headerlink" href="#4.2.-Vector-field-graph" title="Permalink to this headline">¶</a></h3>
<p>We will visualize the simulation results as a vector field on a digitized grid. Single cell transition vectors are grouped by grid point.</p>
</div>
</div>
<div class="section" id="4.2.1-Find-parameters-for-n_grid-and-min_mass">
<h2>4.2.1 Find parameters for n_grid and min_mass<a class="headerlink" href="#4.2.1-Find-parameters-for-n_grid-and-min_mass" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">n_grid</span></code>: Number of grid points.</p>
<p><code class="docutils literal notranslate"><span class="pre">min_mass</span></code>: Threshold value for the cell density. The appropriate values for these parameters depends on the data. Please find appropriate values using the helper functions below.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># n_grid = 40 is a good starting value.</span>
<span class="n">n_grid</span> <span class="o">=</span> <span class="mi">40</span>
<span class="n">oracle</span><span class="o">.</span><span class="n">calculate_p_mass</span><span class="p">(</span><span class="n">smooth</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span> <span class="n">n_grid</span><span class="o">=</span><span class="n">n_grid</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Please run <code class="docutils literal notranslate"><span class="pre">oracle.suggest_mass_thresholds()</span></code> to display a range of <code class="docutils literal notranslate"><span class="pre">min_mass</span></code> parameter values and choose a value to fit the data.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Search for best min_mass.</span>
<span class="n">oracle</span><span class="o">.</span><span class="n">suggest_mass_thresholds</span><span class="p">(</span><span class="n">n_suggestion</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_29_0.png" class="no-scaled-link" src="../../_images/notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_29_0.png" style="width: 1130px; height: 845px;" />
</div>
</div>
<p>According to the results, the optimal <code class="docutils literal notranslate"><span class="pre">min_mass</span></code> is around 0.011.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">min_mass</span> <span class="o">=</span> <span class="mf">0.01</span>
<span class="n">oracle</span><span class="o">.</span><span class="n">calculate_mass_filter</span><span class="p">(</span><span class="n">min_mass</span><span class="o">=</span><span class="n">min_mass</span><span class="p">,</span> <span class="n">plot</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_31_0.png" class="no-scaled-link" src="../../_images/notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_31_0.png" style="width: 293px; height: 301px;" />
</div>
</div>
</div>
<div class="section" id="4.2.2-Plot-vector-fields">
<h2>4.2.2 Plot vector fields<a class="headerlink" href="#4.2.2-Plot-vector-fields" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Again, we need to adjust the <code class="docutils literal notranslate"><span class="pre">scale</span></code> parameter. Please seek to find the optimal <code class="docutils literal notranslate"><span class="pre">scale</span></code> parameter that provides good visualization.</p></li>
<li><p>If you don’t see any vector, you can try the smaller <code class="docutils literal notranslate"><span class="pre">scale</span></code> parameter to magnify the vector length. However, if you see large vectors in the randomized results (right panel), it means the scale parameter is too small.</p></li>
</ul>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span>  <span class="n">figsize</span><span class="o">=</span><span class="p">[</span><span class="mi">13</span><span class="p">,</span> <span class="mi">6</span><span class="p">])</span>

<span class="n">scale_simulation</span> <span class="o">=</span> <span class="mf">0.5</span>
<span class="c1"># Show quiver plot</span>
<span class="n">oracle</span><span class="o">.</span><span class="n">plot_simulation_flow_on_grid</span><span class="p">(</span><span class="n">scale</span><span class="o">=</span><span class="n">scale_simulation</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Simulated cell identity shift vector: </span><span class="si">{</span><span class="n">goi</span><span class="si">}</span><span class="s2"> KO&quot;</span><span class="p">)</span>

<span class="c1"># Show quiver plot that was calculated with randomized graph.</span>
<span class="n">oracle</span><span class="o">.</span><span class="n">plot_simulation_flow_random_on_grid</span><span class="p">(</span><span class="n">scale</span><span class="o">=</span><span class="n">scale_simulation</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Randomized simulation vector&quot;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_33_0.png" class="no-scaled-link" src="../../_images/notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_33_0.png" style="width: 739px; height: 356px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Plot vector field with cell cluster</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">[</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">])</span>

<span class="n">oracle</span><span class="o">.</span><span class="n">plot_cluster_whole</span><span class="p">(</span><span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="n">oracle</span><span class="o">.</span><span class="n">plot_simulation_flow_on_grid</span><span class="p">(</span><span class="n">scale</span><span class="o">=</span><span class="n">scale_simulation</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span> <span class="n">show_background</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_34_0.png" class="no-scaled-link" src="../../_images/notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_34_0.png" style="width: 460px; height: 449px;" />
</div>
</div>
</div>
</div>
<div class="section" id="5.-Compare-simulation-vector-with-development-vectors">
<h1>5. Compare simulation vector with development vectors<a class="headerlink" href="#5.-Compare-simulation-vector-with-development-vectors" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><p>As shown above, we can simulate how TF perturbations affect cell identity and visualize the results as a vector field map.</p></li>
<li><p>To interpret the results, it is necessary to take into account the direction of natural differentiation. We will compare the simulated perturbation vectors with the developmental gradient vectors. By comparing them, we can intuitively understand how TFs impact cell fate determination during development. This perspective is also important for estimating the experimental perturbation results.</p></li>
<li><p>Here, we will calculate the development vector field using <strong>pseudotime gradient</strong> as follows.</p></li>
</ul>
<ol class="arabic simple">
<li><p>Transfer the <strong>pseudotime data</strong> into an n x n digitized grid.</p></li>
<li><p>Calculate the 2D gradient of pseudotime to get vector field.</p></li>
<li><p>Compare the in silico TF perturbation vector field with the development vector field by calculating inner product between these two vectors.</p></li>
</ol>
<ul class="simple">
<li><p>Note: Other methods may be used to represent a continuous scRNA-seq trajectory flow. For example, RNA velocity analysis is a good way to estimate the direction of cell differentiation. Choose the method that best suits your data.</p></li>
</ul>
<p>In the analysis below, we will use <strong>pseudotime</strong> data. Pseudotime data is included in the demo data. <strong>If you are analyzing your own scRNA-seq data, please calculate pseudotime before starting this analysis.</strong></p>
<p>Refer to the following link below to see an example about calculating pseudotime for scRNA-seq data. <a class="reference external" href="https://morris-lab.github.io/CellOracle.documentation/tutorials/pseudotime.html">https://morris-lab.github.io/CellOracle.documentation/tutorials/pseudotime.html</a></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Visualize pseudotime</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">[</span><span class="mi">6</span><span class="p">,</span><span class="mi">6</span><span class="p">])</span>

<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span><span class="n">adata</span><span class="o">=</span><span class="n">oracle</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span> <span class="n">basis</span><span class="o">=</span><span class="n">oracle</span><span class="o">.</span><span class="n">embedding_name</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;rainbow&quot;</span><span class="p">,</span>
                <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Pseudotime&quot;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_38_0.png" class="no-scaled-link" src="../../_images/notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_38_0.png" style="width: 369px; height: 369px;" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">celloracle.applications</span> <span class="kn">import</span> <span class="n">Gradient_calculator</span>

<span class="c1"># Instantiate Gradient calculator object</span>
<span class="n">gradient</span> <span class="o">=</span> <span class="n">Gradient_calculator</span><span class="p">(</span><span class="n">oracle_object</span><span class="o">=</span><span class="n">oracle</span><span class="p">,</span> <span class="n">pseudotime_key</span><span class="o">=</span><span class="s2">&quot;Pseudotime&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>We need to set <code class="docutils literal notranslate"><span class="pre">n_grid</span></code> and <code class="docutils literal notranslate"><span class="pre">min_mass</span></code> for the pseudotime grid point calculation too. <code class="docutils literal notranslate"><span class="pre">n_grid</span></code>: Number of grid point.</p>
<p>We already know approproate values for them. Please set the same values as step 4.2.1 above.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">gradient</span><span class="o">.</span><span class="n">calculate_p_mass</span><span class="p">(</span><span class="n">smooth</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span> <span class="n">n_grid</span><span class="o">=</span><span class="n">n_grid</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
<span class="n">gradient</span><span class="o">.</span><span class="n">calculate_mass_filter</span><span class="p">(</span><span class="n">min_mass</span><span class="o">=</span><span class="n">min_mass</span><span class="p">,</span> <span class="n">plot</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_42_0.png" class="no-scaled-link" src="../../_images/notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_42_0.png" style="width: 293px; height: 301px;" />
</div>
</div>
<p>Next, we convert the pseudotime data into grid points. For this calculation we can chose one of two methods.</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">knn</span></code>: K-Nearesr Neighbor regression. You will need to set number of neighbors. Please adjust <code class="docutils literal notranslate"><span class="pre">n_knn</span></code> for best results.This will depend on the population size and density of your scRNA-seq data.</p></li>
</ul>
<p><code class="docutils literal notranslate"><span class="pre">gradient.transfer_data_into_grid(args={&quot;method&quot;:</span> <span class="pre">&quot;knn&quot;,</span> <span class="pre">&quot;n_knn&quot;:50})</span></code></p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">polynomial</span></code>: Polynomial regression using x-axis and y-axis of dimensional reduction space.</p></li>
</ul>
<p>In general, this method will be more robust. Please use this method if knn method does not work. <code class="docutils literal notranslate"><span class="pre">n_poly</span></code> is the number of degrees for the polynomial regression model. Please try to find appropriate<code class="docutils literal notranslate"><span class="pre">n_poly</span></code> searching for best results.</p>
<p><code class="docutils literal notranslate"><span class="pre">gradient.transfer_data_into_grid(args={&quot;method&quot;:</span> <span class="pre">&quot;polynomial&quot;,</span> <span class="pre">&quot;n_poly&quot;:3})</span></code></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">gradient</span><span class="o">.</span><span class="n">transfer_data_into_grid</span><span class="p">(</span><span class="n">args</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;method&quot;</span><span class="p">:</span> <span class="s2">&quot;polynomial&quot;</span><span class="p">,</span> <span class="s2">&quot;n_poly&quot;</span><span class="p">:</span><span class="mi">3</span><span class="p">},</span> <span class="n">plot</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/k/anaconda3/envs/pandas1/lib/python3.6/site-packages/sklearn/linear_model/_ridge.py:148: LinAlgWarning:

Ill-conditioned matrix (rcond=1.11032e-17): result may not be accurate.

</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_45_1.png" class="no-scaled-link" src="../../_images/notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_45_1.png" style="width: 572px; height: 301px;" />
</div>
</div>
<p>Calculate the 2D vector map to represents the pseudotime gradient. After the gradient calculation, the length of the vector will be normalized automatically.</p>
<p>Please adjust the <code class="docutils literal notranslate"><span class="pre">scale</span></code> parameter to adjust vector length visualization.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Calculate graddient</span>
<span class="n">gradient</span><span class="o">.</span><span class="n">calculate_gradient</span><span class="p">()</span>

<span class="c1"># Show results</span>
<span class="n">scale_dev</span> <span class="o">=</span> <span class="mi">40</span>
<span class="n">gradient</span><span class="o">.</span><span class="n">visualize_results</span><span class="p">(</span><span class="n">scale</span><span class="o">=</span><span class="n">scale_dev</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_47_0.png" class="no-scaled-link" src="../../_images/notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_47_0.png" style="width: 1409px; height: 315px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Visualize results</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">[</span><span class="mi">6</span><span class="p">,</span> <span class="mi">6</span><span class="p">])</span>
<span class="n">gradient</span><span class="o">.</span><span class="n">plot_dev_flow_on_grid</span><span class="p">(</span><span class="n">scale</span><span class="o">=</span><span class="n">scale_dev</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_48_0.png" class="no-scaled-link" src="../../_images/notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_48_0.png" style="width: 349px; height: 340px;" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Save gradient object if you want.</span>
<span class="c1">#gradient.to_hdf5(&quot;../data/Paul_etal.celloracle.gradient&quot;)</span>
</pre></div>
</div>
</div>
<p>We will use the inner product calculations to quantitatively compare the 2D developmental flow and perturbation simulation vectors. &gt; If you are not familiar with Inner product / Dot product, please see <a class="reference external" href="https://en.wikipedia.org/wiki/Dot_product">https://en.wikipedia.org/wiki/Dot_product</a></p>
<ul class="simple">
<li><p><strong>The inner product represents the similarity between two vectors.</strong></p></li>
<li><p>Using the inner product, we compare the 2D vector field of perturbation simulation and development flow.</p></li>
<li><p>The inner product will be a positive value when two vectors are pointing in the same direction.</p></li>
<li><p>Conversely, the inner product will be a negative value when two vectors are pointing in the opposing directions.</p></li>
</ul>
<p><img alt="d8d0045adb2b457e926b5b3c2f9c5de7" src="https://raw.githubusercontent.com/morris-lab/CellOracle/master/docs/demo_data/innerproduct_explantion_1.png" /></p>
<ul class="simple">
<li><p>The length of the vectors also affects the magnitude of inner product value.</p></li>
</ul>
<p>In summary, we quantitatively compare the directionality and size of vectors between perturbation simulation and natural differentiation using inner product, and we define the score as <strong>perturbation score (PS)</strong>.</p>
<ul class="simple">
<li><p><strong>A negative PS</strong> means that the TF perturbation would <strong>block differentiation</strong>.</p></li>
<li><p><strong>A positive PS</strong> means that the TF perturbation would <strong>promote differentiation</strong>.</p></li>
</ul>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">celloracle.applications</span> <span class="kn">import</span> <span class="n">Oracle_development_module</span>

<span class="c1"># Make Oracle_development_module to compare two vector field</span>
<span class="n">dev</span> <span class="o">=</span> <span class="n">Oracle_development_module</span><span class="p">()</span>

<span class="c1"># Load development flow</span>
<span class="n">dev</span><span class="o">.</span><span class="n">load_differentiation_reference_data</span><span class="p">(</span><span class="n">gradient_object</span><span class="o">=</span><span class="n">gradient</span><span class="p">)</span>

<span class="c1"># Load simulation result</span>
<span class="n">dev</span><span class="o">.</span><span class="n">load_perturb_simulation_data</span><span class="p">(</span><span class="n">oracle_object</span><span class="o">=</span><span class="n">oracle</span><span class="p">)</span>


<span class="c1"># Calculate inner produc scores</span>
<span class="n">dev</span><span class="o">.</span><span class="n">calculate_inner_product</span><span class="p">()</span>
<span class="n">dev</span><span class="o">.</span><span class="n">calculate_digitized_ip</span><span class="p">(</span><span class="n">n_bins</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Let&#39;s visualize the results</span>
<span class="n">dev</span><span class="o">.</span><span class="n">visualize_development_module_layout_0</span><span class="p">(</span><span class="n">s</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
                                          <span class="n">scale_for_simulation</span><span class="o">=</span><span class="n">scale_simulation</span><span class="p">,</span>
                                          <span class="n">s_grid</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span>
                                          <span class="n">scale_for_pseudotime</span><span class="o">=</span><span class="n">scale_dev</span><span class="p">,</span>
                                          <span class="n">vm</span><span class="o">=</span><span class="mf">0.02</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_53_0.png" class="no-scaled-link" src="../../_images/notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_53_0.png" style="width: 1130px; height: 794px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Show perturbation scores</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">[</span><span class="mi">6</span><span class="p">,</span> <span class="mi">6</span><span class="p">])</span>
<span class="n">dev</span><span class="o">.</span><span class="n">plot_inner_product_on_grid</span><span class="p">(</span><span class="n">vm</span><span class="o">=</span><span class="mf">0.02</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_54_0.png" class="no-scaled-link" src="../../_images/notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_54_0.png" style="width: 349px; height: 340px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Show perturbation scores with perturbation simulation vector field</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">[</span><span class="mi">6</span><span class="p">,</span> <span class="mi">6</span><span class="p">])</span>
<span class="n">dev</span><span class="o">.</span><span class="n">plot_inner_product_on_grid</span><span class="p">(</span><span class="n">vm</span><span class="o">=</span><span class="mf">0.02</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
<span class="n">dev</span><span class="o">.</span><span class="n">plot_simulation_flow_on_grid</span><span class="p">(</span><span class="n">scale</span><span class="o">=</span><span class="n">scale_simulation</span><span class="p">,</span> <span class="n">show_background</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_55_0.png" class="no-scaled-link" src="../../_images/notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_55_0.png" style="width: 349px; height: 340px;" />
</div>
</div>
</div>
<div class="section" id="6.-Focus-on-a-single-development-lineage-to-interpret-the-results-in-detail">
<h1>6. Focus on a single development lineage to interpret the results in detail<a class="headerlink" href="#6.-Focus-on-a-single-development-lineage-to-interpret-the-results-in-detail" title="Permalink to this headline">¶</a></h1>
<p>So far, we have used the <code class="docutils literal notranslate"><span class="pre">Oracle_development_module</span></code> to analyze the whole cell population. However, the <code class="docutils literal notranslate"><span class="pre">Oracle_development_module</span></code> can also analyze specific subpopulations.</p>
<p>In this example, we will analyze the MEP lineage.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Get cell index list for the cells of interest</span>
<span class="n">clusters</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Ery_0&#39;</span><span class="p">,</span> <span class="s1">&#39;Ery_1&#39;</span><span class="p">,</span> <span class="s1">&#39;Ery_2&#39;</span><span class="p">,</span> <span class="s1">&#39;Ery_3&#39;</span><span class="p">,</span> <span class="s1">&#39;Ery_4&#39;</span><span class="p">,</span> <span class="s1">&#39;Ery_5&#39;</span><span class="p">,</span> <span class="s1">&#39;Ery_6&#39;</span><span class="p">,</span>
            <span class="s1">&#39;Ery_7&#39;</span><span class="p">,</span> <span class="s1">&#39;Ery_8&#39;</span><span class="p">,</span> <span class="s1">&#39;Ery_9&#39;</span><span class="p">,</span> <span class="s1">&#39;MEP_0&#39;</span><span class="p">,</span> <span class="s1">&#39;Mk_0&#39;</span><span class="p">]</span>
<span class="n">cell_idx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">oracle</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;louvain_annot&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">clusters</span><span class="p">))[</span><span class="mi">0</span><span class="p">]</span>

<span class="c1"># Check index</span>
<span class="nb">print</span><span class="p">(</span><span class="n">cell_idx</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[   0    2    4 ... 2666 2668 2670]
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">dev</span> <span class="o">=</span> <span class="n">Oracle_development_module</span><span class="p">()</span>

<span class="c1"># Load development flow</span>
<span class="n">dev</span><span class="o">.</span><span class="n">load_differentiation_reference_data</span><span class="p">(</span><span class="n">gradient_object</span><span class="o">=</span><span class="n">gradient</span><span class="p">)</span>

<span class="c1"># Load simulation result</span>
<span class="n">dev</span><span class="o">.</span><span class="n">load_perturb_simulation_data</span><span class="p">(</span><span class="n">oracle_object</span><span class="o">=</span><span class="n">oracle</span><span class="p">,</span>
                                 <span class="n">cell_idx_use</span><span class="o">=</span><span class="n">cell_idx</span><span class="p">,</span> <span class="c1"># Enter cell id list</span>
                                 <span class="n">name</span><span class="o">=</span><span class="s2">&quot;Lineage_MEP&quot;</span> <span class="c1"># Name of this cell group. You can enter arbitrary name.</span>
                                 <span class="p">)</span>

<span class="c1"># Calculation</span>
<span class="n">dev</span><span class="o">.</span><span class="n">calculate_inner_product</span><span class="p">()</span>
<span class="n">dev</span><span class="o">.</span><span class="n">calculate_digitized_ip</span><span class="p">(</span><span class="n">n_bins</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Let&#39;s visualize the results</span>
<span class="n">dev</span><span class="o">.</span><span class="n">visualize_development_module_layout_0</span><span class="p">(</span><span class="n">s</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
                                          <span class="n">scale_for_simulation</span><span class="o">=</span><span class="n">scale_simulation</span><span class="p">,</span>
                                          <span class="n">s_grid</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span>
                                          <span class="n">scale_for_pseudotime</span><span class="o">=</span><span class="n">scale_dev</span><span class="p">,</span>
                                          <span class="n">vm</span><span class="o">=</span><span class="mf">0.03</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_59_0.png" class="no-scaled-link" src="../../_images/notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_59_0.png" style="width: 1130px; height: 794px;" />
</div>
</div>
<p><strong>Attention: We have changed the default color scheme for the perturbation score visualization at CellOracle = 0.9.0.</strong> If you would like to use the previous blue and red color scheme, please run the following command to change the color scheme.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">celloracle.visualizations.config</span> <span class="kn">import</span> <span class="n">CONFIG</span>
<span class="n">CONFIG</span><span class="p">[</span><span class="s2">&quot;cmap_ps&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;coolwarm&quot;</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">dev</span><span class="o">.</span><span class="n">visualize_development_module_layout_0</span><span class="p">(</span><span class="n">s</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
                                          <span class="n">scale_for_simulation</span><span class="o">=</span><span class="n">scale_simulation</span><span class="p">,</span>
                                          <span class="n">s_grid</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span>
                                          <span class="n">scale_for_pseudotime</span><span class="o">=</span><span class="n">scale_dev</span><span class="p">,</span>
                                          <span class="n">vm</span><span class="o">=</span><span class="mf">0.03</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_62_0.png" class="no-scaled-link" src="../../_images/notebooks_05_simulation_Gata1_KO_simulation_with_Paul_etal_2015_data_62_0.png" style="width: 1130px; height: 794px;" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span>
</pre></div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../../tutorials/scrnaprocess.html" class="btn btn-neutral float-right" title="1. scRNA-seq data preparation" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="../../tutorials/simulation.html" class="btn btn-neutral float-left" title="1. In silico gene perturbation with GRNs" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, Samantha Morris Lab

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>