

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  <script type="text/javascript">

      var _gaq = _gaq || [];
      _gaq.push(['_setAccount', 'UA-214970391-3']);
      _gaq.push(['_trackPageview']);

      (function() {
        var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
        ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
        var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
      })();
    </script>
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Overview &mdash; celloracle 0.9.1 documentation</title>
  

  
  <link rel="stylesheet" href="../../../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../../../_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../../../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../../../../" src="../../../../../_static/documentation_options.js"></script>
        <script src="../../../../../_static/jquery.js"></script>
        <script src="../../../../../_static/underscore.js"></script>
        <script src="../../../../../_static/doctools.js"></script>
        <script src="../../../../../_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="../../../../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../../../../index.html" class="icon icon-home" alt="Documentation Home"> celloracle
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../../index.html">Welcome to CellOracleâ€™s documentation!</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../../installation/index.html">Installation</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../../tutorials/index.html">Tutorial</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../../modules/index.html">API</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../../changelog/index.html">Changelog</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../../license/index.html">License</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../../citation/index.html">Authors and citations</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../../contact/index.html">Contact</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../../../index.html">celloracle</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../../../index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Overview</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../../../../../_sources/notebooks/01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/misc/Make tss reference datafile from Ensembl database/1_make_tss_referenece_from_ensemble_gff3_file.ipynb.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container,
div.nbinput.container div.prompt,
div.nbinput.container div.input_area,
div.nbinput.container div[class*=highlight],
div.nbinput.container div[class*=highlight] pre,
div.nboutput.container,
div.nboutput.container div.prompt,
div.nboutput.container div.output_area,
div.nboutput.container div[class*=highlight],
div.nboutput.container div[class*=highlight] pre {
    background: none;
    border: none;
    padding: 0 0;
    margin: 0;
    box-shadow: none;
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    background: #f5f5f5;
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt a.copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="Overview">
<h1>Overview<a class="headerlink" href="#Overview" title="Permalink to this headline">Â¶</a></h1>
<p>In this notebook, we will make TSS annotation data that are used for the scATAC-seq peak annotation.</p>
<ul class="simple">
<li><p>First, we download gene annotation gff3 file from Ensembl database.</p></li>
<li><p>Second, we convert gff3 file into bed file. During this process, the TSS information is extracted.</p></li>
</ul>
</div>
<div class="section" id="!!-Caution!!">
<h1>!! Caution!!<a class="headerlink" href="#!!-Caution!!" title="Permalink to this headline">Â¶</a></h1>
<div class="section" id="1)-This-is-NOT-part-of-CellOracle-tutorial.">
<h2>1) This is NOT part of CellOracle tutorial.<a class="headerlink" href="#1)-This-is-NOT-part-of-CellOracle-tutorial." title="Permalink to this headline">Â¶</a></h2>
<ul class="simple">
<li><p>This notebook includes unusual usage of CellOracle.</p></li>
<li><p>The analysis might require expertise of python and DNA sequence analysis, but this notebook does not aim to explain them all, and please use this notebook by your responsibility.</p></li>
</ul>
</div>
<div class="section" id="2)-This-notebook-was-tested-with-Ensembl-Guinea-Pig-data,-but-we-do-not-guarantee-the-function-works-with-other-species-or-other-database.">
<h2>2) This notebook was tested with Ensembl Guinea Pig data, but we do not guarantee the function works with other species or other database.<a class="headerlink" href="#2)-This-notebook-was-tested-with-Ensembl-Guinea-Pig-data,-but-we-do-not-guarantee-the-function-works-with-other-species-or-other-database." title="Permalink to this headline">Â¶</a></h2>
<ul class="simple">
<li><p>Please let us know using git hub issue if you have problem with this notebook.</p></li>
<li><p>We can construct TSS annotation data and add them to CellOracle package. Please just let us know if you have a request for new reference genome.</p></li>
</ul>
</div>
</div>
<div class="section" id="0.-Import-libraries">
<h1>0. Import libraries<a class="headerlink" href="#0.-Import-libraries" title="Permalink to this headline">Â¶</a></h1>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="kn">import</span> <span class="nn">os</span><span class="o">,</span> <span class="nn">sys</span>
<span class="kn">from</span> <span class="nn">tqdm.notebook</span> <span class="kn">import</span>  <span class="n">tqdm</span>

<span class="kn">from</span> <span class="nn">pybedtools</span> <span class="kn">import</span> <span class="n">BedTool</span>
<span class="kn">import</span> <span class="nn">genomepy</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">celloracle</span> <span class="k">as</span> <span class="nn">co</span>
<span class="kn">from</span> <span class="nn">celloracle</span> <span class="kn">import</span> <span class="n">motif_analysis</span> <span class="k">as</span> <span class="n">ma</span>

<span class="n">co</span><span class="o">.</span><span class="n">__version__</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&#39;0.8.4&#39;
</pre></div></div>
</div>
</div>
<div class="section" id="1.-Define-custom-functions-to-process-gene-annotation-data.">
<h1>1. Define custom functions to process gene annotation data.<a class="headerlink" href="#1.-Define-custom-functions-to-process-gene-annotation-data." title="Permalink to this headline">Â¶</a></h1>
<p>Extract TSS information from gff3 file and get a bed file.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">def</span> <span class="nf">parse_ens</span><span class="p">(</span><span class="n">x</span><span class="p">):</span>
    <span class="n">dic</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">if</span> <span class="s2">&quot;;&quot;</span> <span class="ow">in</span> <span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;;&quot;</span><span class="p">):</span>
            <span class="n">key</span><span class="p">,</span> <span class="n">val</span> <span class="o">=</span> <span class="n">i</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;=&quot;</span><span class="p">)</span>
            <span class="n">dic</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">val</span>
    <span class="k">return</span> <span class="n">dic</span>

<span class="k">def</span> <span class="nf">get_tss_and_promoter_candidate_locus</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">n_downstream</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span> <span class="n">n_upstream</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span> <span class="n">clip_negative</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="n">data</span><span class="p">[</span><span class="s2">&quot;TSS&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;start&quot;</span><span class="p">]</span>

    <span class="n">mRNA_in_reversed_strand</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span><span class="p">]</span>
    <span class="n">data</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">mRNA_in_reversed_strand</span><span class="p">,</span> <span class="s2">&quot;TSS&quot;</span><span class="p">]</span> <span class="o">=</span> \
        <span class="n">data</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">mRNA_in_reversed_strand</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">]</span>

    <span class="n">data</span><span class="p">[</span><span class="s2">&quot;promTSS_left&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;TSS&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">n_upstream</span>
    <span class="n">data</span><span class="p">[</span><span class="s2">&quot;promTSS_right&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;TSS&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">n_downstream</span>

    <span class="n">data</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">mRNA_in_reversed_strand</span><span class="p">,</span> <span class="s2">&quot;promTSS_left&quot;</span><span class="p">]</span> <span class="o">=</span> \
        <span class="n">data</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">mRNA_in_reversed_strand</span><span class="p">,</span> <span class="s2">&quot;TSS&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">n_downstream</span>
    <span class="n">data</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">mRNA_in_reversed_strand</span><span class="p">,</span> <span class="s2">&quot;promTSS_right&quot;</span><span class="p">]</span> <span class="o">=</span> \
        <span class="n">data</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">mRNA_in_reversed_strand</span><span class="p">,</span> <span class="s2">&quot;TSS&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">n_upstream</span>

    <span class="k">if</span> <span class="n">clip_negative</span><span class="p">:</span>
        <span class="n">data</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">data</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">data</span><span class="o">.</span><span class="n">promTSS_left</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">],</span> <span class="s2">&quot;promTSS_left&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="k">return</span> <span class="n">data</span>

<span class="k">def</span> <span class="nf">load_and_process_ensembl_gff3_file</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">n_downstream</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">n_upstream</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">transcript_filtering</span><span class="o">=</span><span class="s2">&quot;auto&quot;</span><span class="p">,</span> <span class="n">clip_negative</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="c1"># Load gff file. Comments rows are skipped.</span>
    <span class="n">lines</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">l</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">readlines</span><span class="p">()):</span>
            <span class="k">if</span> <span class="n">l</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;#&quot;</span><span class="p">):</span>
                <span class="k">pass</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">lines</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">l</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">))</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">lines</span><span class="p">)</span>


    <span class="c1"># Data format adjustment 1</span>
    <span class="n">df</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;seqname&quot;</span><span class="p">,</span> <span class="s2">&quot;source&quot;</span><span class="p">,</span> <span class="s2">&quot;feature&quot;</span><span class="p">,</span> <span class="s2">&quot;start&quot;</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">,</span> <span class="s2">&quot;score&quot;</span><span class="p">,</span>
                  <span class="s2">&quot;strand&quot;</span><span class="p">,</span> <span class="s2">&quot;frame&quot;</span><span class="p">,</span> <span class="s2">&quot;attribute&quot;</span><span class="p">]</span>

    <span class="n">df</span><span class="p">[</span><span class="s2">&quot;start&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;start&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;int&quot;</span><span class="p">)</span>
    <span class="n">df</span><span class="p">[</span><span class="s2">&quot;end&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;end&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;int&quot;</span><span class="p">)</span>

    <span class="c1"># Data format adjustment 2</span>
    <span class="c1">## The attribute column includes detailed information. Let&#39;s extract information and store them as new columns.</span>
    <span class="n">annot</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">([</span><span class="n">parse_ens</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;attribute&quot;</span><span class="p">])])</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">df</span><span class="p">,</span> <span class="n">annot</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># Data format adjustment 2</span>
    <span class="n">df</span><span class="p">[</span><span class="s2">&quot;Parent_feature&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">Parent</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="s2">&quot;na:na&quot;</span><span class="p">)]</span>
    <span class="n">df</span><span class="p">[</span><span class="s2">&quot;Parent_id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">Parent</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="s2">&quot;na:na&quot;</span><span class="p">)]</span>


    <span class="c1"># Data format adjustment 3</span>
    <span class="n">df</span><span class="p">[</span><span class="s2">&quot;gene_id_unified&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;gene_id&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
    <span class="n">rows_non_gene</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;gene_id&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isna</span><span class="p">()</span>
    <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">df</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">rows_non_gene</span><span class="p">],</span> <span class="s2">&quot;gene_id_unified&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">rows_non_gene</span><span class="p">][</span><span class="s2">&quot;Parent_id&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>

    <span class="c1">#return df</span>
    <span class="c1"># Split data into gene entry and transcript entry.</span>
    <span class="n">df_gene</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="o">~</span><span class="n">rows_non_gene</span><span class="p">]</span>
    <span class="n">df_gene</span> <span class="o">=</span> <span class="n">df_gene</span><span class="p">[[</span><span class="s2">&quot;Name&quot;</span><span class="p">,</span> <span class="s2">&quot;gene_id&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;Name&quot;</span><span class="p">:</span> <span class="s2">&quot;gene_symbol&quot;</span><span class="p">,</span> <span class="s2">&quot;gene_id&quot;</span><span class="p">:</span> <span class="s2">&quot;gene_id_unified&quot;</span><span class="p">})</span>
    <span class="n">df_transcript</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="o">.</span><span class="n">Parent_feature</span> <span class="o">==</span> <span class="s2">&quot;gene&quot;</span><span class="p">]</span>
    <span class="n">df_transcript</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">df_transcript</span><span class="p">,</span> <span class="n">df_gene</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;gene_id_unified&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">)</span>

    <span class="c1"># We only use basic transcript, major mRNA isoform.</span>

    <span class="k">if</span> <span class="n">transcript_filtering</span> <span class="o">==</span> <span class="s2">&quot;auto&quot;</span><span class="p">:</span>
        <span class="k">if</span> <span class="s2">&quot;tag&quot;</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span> <span class="c1"># We only use basic transcript, major mRNA isoform.</span>
            <span class="n">df_transcript</span> <span class="o">=</span> <span class="n">df_transcript</span><span class="p">[</span><span class="n">df_transcript</span><span class="o">.</span><span class="n">tag</span> <span class="o">==</span> <span class="s2">&quot;basic&quot;</span><span class="p">]</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Found transcriptome tag information. Only &#39;basic&#39; transcripts are used.&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;No transcriptome tag information found. Transfript filtering step is skipped. All transcripts are used.&quot;</span><span class="p">)</span>

    <span class="k">elif</span> <span class="n">transcript_filtering</span> <span class="o">==</span> <span class="s2">&quot;basic&quot;</span><span class="p">:</span>
        <span class="k">if</span> <span class="s2">&quot;tag&quot;</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span> <span class="c1"># We only use basic transcript, major mRNA isoform.</span>
            <span class="n">df_transcript</span> <span class="o">=</span> <span class="n">df_transcript</span><span class="p">[</span><span class="n">df_transcript</span><span class="o">.</span><span class="n">tag</span> <span class="o">==</span> <span class="s2">&quot;basic&quot;</span><span class="p">]</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Found transcriptome tag information. Only &#39;basic&#39; transcripts are used.&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Could not perform basic transcript filtering because the gff3 file not contain &#39;tag&#39; information. </span><span class="se">\n\</span>
<span class="s2">                  Please set transcript_filtering=&#39;auto&#39;.</span><span class="se">\n\</span>
<span class="s2">                  With this mode, transcript filtering is not performed and all transcripts are used.&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;transcript_filtering, </span><span class="si">{</span><span class="n">transcript_filtering</span><span class="si">}</span><span class="s2"> is not implemented&quot;</span><span class="p">)</span>


    <span class="c1"># Remove transcripts that are not annotated gene name.</span>
    <span class="n">df_transcript</span> <span class="o">=</span> <span class="n">df_transcript</span><span class="p">[</span><span class="o">~</span><span class="n">df_transcript</span><span class="o">.</span><span class="n">gene_symbol</span><span class="o">.</span><span class="n">isna</span><span class="p">()]</span>


    <span class="c1"># Add PromoterTSS location.</span>
    <span class="n">df_transcript</span> <span class="o">=</span> <span class="n">get_tss_and_promoter_candidate_locus</span><span class="p">(</span><span class="n">df_transcript</span><span class="p">,</span>
                                         <span class="n">n_downstream</span><span class="o">=</span><span class="n">n_downstream</span><span class="p">,</span> <span class="n">n_upstream</span><span class="o">=</span><span class="n">n_upstream</span><span class="p">,</span> <span class="n">clip_negative</span><span class="o">=</span><span class="n">clip_negative</span><span class="p">)</span>

    <span class="c1"># Wrap up necessary information.</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">df_transcript</span><span class="p">[[</span><span class="s2">&quot;seqname&quot;</span><span class="p">,</span> <span class="s2">&quot;promTSS_left&quot;</span><span class="p">,</span> <span class="s2">&quot;promTSS_right&quot;</span><span class="p">,</span>
                        <span class="s2">&quot;gene_symbol&quot;</span><span class="p">,</span> <span class="s2">&quot;score&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">]]</span>

    <span class="k">return</span> <span class="n">result</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="2.-Install-reference-genome-first.">
<h1>2. Install reference genome first.<a class="headerlink" href="#2.-Install-reference-genome-first." title="Permalink to this headline">Â¶</a></h1>
<p>We use genomepy to get genomic DNA sequence. The first step is to install reference genome data.</p>
<p>We will use the genomepy function. <code class="docutils literal notranslate"><span class="pre">genomepy.install_genome()</span></code></p>
<p>We need (1) referenoce genome name and (2) provider.</p>
<p>Please see genomepyâ€™s documentation for more information. <a class="reference external" href="https://pypi.org/project/genomepy/">https://pypi.org/project/genomepy/</a></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Search for reference genome name and provider</span>
<span class="o">!</span>genomepy search <span class="s2">&quot;Cavia porcellus&quot;</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
<span class="ansi-bold">name         provider    accession          species            tax_id    other_info                                                  </span>
Cavpor3.0    Ensembl     GCA_000151735.1    Cavia porcellus    10141     2016-12-ensembl/2017-07
cavPor3      UCSC        na                 Cavia porcellus    10141     Feb. 2008 (Broad/cavPor3)
Cavpor3.0    NCBI        GCA_000151735.1    Cavia porcellus    10141     The Genome Sequencing Platform, The Genome Assembly Team
<span class="ansi-green-fg"> ^</span>
<span class="ansi-green-fg"> Use name for </span><span class="ansi-cyan-fg">genomepy install</span>

</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Install reference genome. You can skip this step if you already installed reference genome.</span>
<span class="n">ref_genome</span> <span class="o">=</span> <span class="s2">&quot;Cavpor3.0&quot;</span>
<span class="n">provider</span> <span class="o">=</span> <span class="s2">&quot;Ensembl&quot;</span>
<span class="n">genomepy</span><span class="o">.</span><span class="n">install_genome</span><span class="p">(</span><span class="n">ref_genome</span><span class="p">,</span> <span class="n">provider</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Check referenoce genome installation status</span>
<span class="n">genome_installation</span> <span class="o">=</span> <span class="n">ma</span><span class="o">.</span><span class="n">is_genome_installed</span><span class="p">(</span><span class="n">ref_genome</span><span class="o">=</span><span class="n">ref_genome</span><span class="p">)</span>
<span class="n">genome_installation</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
True
</pre></div></div>
</div>
</div>
<div class="section" id="3.-Download-genome-annotation-file;-gff3-file,-from-Ensemble-server.">
<h1>3. Download genome annotation file; gff3 file, from Ensemble server.<a class="headerlink" href="#3.-Download-genome-annotation-file;-gff3-file,-from-Ensemble-server." title="Permalink to this headline">Â¶</a></h1>
<p><a class="reference external" href="https://useast.ensembl.org/Cavia_porcellus/Info/Index?db=core">https://useast.ensembl.org/Cavia_porcellus/Info/Index?db=core</a></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="o">!</span>wget http://ftp.ensembl.org/pub/release-105/gff3/cavia_porcellus/Cavia_porcellus.Cavpor3.0.105.gff3.gz
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
--2021-12-29 17:30:50--  http://ftp.ensembl.org/pub/release-105/gff3/cavia_porcellus/Cavia_porcellus.Cavpor3.0.105.gff3.gz
Resolving ftp.ensembl.org (ftp.ensembl.org)... 193.62.197.76
Connecting to ftp.ensembl.org (ftp.ensembl.org)|193.62.197.76|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 12447679 (12M) [application/x-gzip]
Saving to: â€˜Cavia_porcellus.Cavpor3.0.105.gff3.gzâ€™

Cavia_porcellus.Cav 100%[===================&gt;]  11.87M   649KB/s    in 19s

2021-12-29 17:31:09 (644 KB/s) - â€˜Cavia_porcellus.Cavpor3.0.105.gff3.gzâ€™ saved [12447679/12447679]

</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="o">!</span>gunzip Cavia_porcellus.Cavpor3.0.105.gff3.gz
</pre></div>
</div>
</div>
</div>
<div class="section" id="4.-Process-data-to-get-TSS-file.">
<h1>4. Process data to get TSS file.<a class="headerlink" href="#4.-Process-data-to-get-TSS-file." title="Permalink to this headline">Â¶</a></h1>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Load and process gff3 file.</span>

<span class="n">file</span> <span class="o">=</span> <span class="s2">&quot;Cavia_porcellus.Cavpor3.0.105.gff3&quot;</span>
<span class="n">result</span> <span class="o">=</span> <span class="n">load_and_process_ensembl_gff3_file</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">n_downstream</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span>
                                            <span class="n">n_upstream</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span>
                                            <span class="n">transcript_filtering</span><span class="o">=</span><span class="s2">&quot;auto&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "466285df899546ae9bd04c54fb4faf8d", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Check result</span>
<span class="n">result</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>seqname</th>
      <th>promTSS_left</th>
      <th>promTSS_right</th>
      <th>gene_symbol</th>
      <th>score</th>
      <th>strand</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>4</th>
      <td>DS562855.1</td>
      <td>546865</td>
      <td>547965</td>
      <td>TMEM74</td>
      <td>.</td>
      <td>+</td>
    </tr>
    <tr>
      <th>6</th>
      <td>DS562855.1</td>
      <td>867637</td>
      <td>868737</td>
      <td>EMC2</td>
      <td>.</td>
      <td>-</td>
    </tr>
    <tr>
      <th>7</th>
      <td>DS562855.1</td>
      <td>849313</td>
      <td>850413</td>
      <td>EMC2</td>
      <td>.</td>
      <td>-</td>
    </tr>
    <tr>
      <th>8</th>
      <td>DS562855.1</td>
      <td>933093</td>
      <td>934193</td>
      <td>5S_rRNA</td>
      <td>.</td>
      <td>+</td>
    </tr>
    <tr>
      <th>10</th>
      <td>DS562855.1</td>
      <td>1132436</td>
      <td>1133536</td>
      <td>RSPO2</td>
      <td>.</td>
      <td>+</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>34835</th>
      <td>MT</td>
      <td>8879</td>
      <td>9979</td>
      <td>ND4L</td>
      <td>.</td>
      <td>+</td>
    </tr>
    <tr>
      <th>34836</th>
      <td>MT</td>
      <td>9169</td>
      <td>10269</td>
      <td>ND4</td>
      <td>.</td>
      <td>+</td>
    </tr>
    <tr>
      <th>34840</th>
      <td>MT</td>
      <td>10748</td>
      <td>11848</td>
      <td>ND5</td>
      <td>.</td>
      <td>+</td>
    </tr>
    <tr>
      <th>34841</th>
      <td>MT</td>
      <td>13986</td>
      <td>15086</td>
      <td>ND6</td>
      <td>.</td>
      <td>-</td>
    </tr>
    <tr>
      <th>34843</th>
      <td>MT</td>
      <td>13164</td>
      <td>14264</td>
      <td>CYTB</td>
      <td>.</td>
      <td>+</td>
    </tr>
  </tbody>
</table>
<p>24875 rows Ã— 6 columns</p>
</div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Save as bed file</span>
<span class="n">result</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ref_genome</span><span class="si">}</span><span class="s2">_tss_info.bed&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Test">
<h1>Test<a class="headerlink" href="#Test" title="Permalink to this headline">Â¶</a></h1>
<p>Try to load DNA sequence using genomepy</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Load file</span>
<span class="n">tss_file</span> <span class="o">=</span> <span class="n">BedTool</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ref_genome</span><span class="si">}</span><span class="s2">_tss_info.bed&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">to_dataframe</span><span class="p">()</span>
<span class="n">tss_file</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>chrom</th>
      <th>start</th>
      <th>end</th>
      <th>name</th>
      <th>score</th>
      <th>strand</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>DS562855.1</td>
      <td>546865</td>
      <td>547965</td>
      <td>TMEM74</td>
      <td>.</td>
      <td>+</td>
    </tr>
    <tr>
      <th>1</th>
      <td>DS562855.1</td>
      <td>867637</td>
      <td>868737</td>
      <td>EMC2</td>
      <td>.</td>
      <td>-</td>
    </tr>
    <tr>
      <th>2</th>
      <td>DS562855.1</td>
      <td>849313</td>
      <td>850413</td>
      <td>EMC2</td>
      <td>.</td>
      <td>-</td>
    </tr>
    <tr>
      <th>3</th>
      <td>DS562855.1</td>
      <td>933093</td>
      <td>934193</td>
      <td>5S_rRNA</td>
      <td>.</td>
      <td>+</td>
    </tr>
    <tr>
      <th>4</th>
      <td>DS562855.1</td>
      <td>1132436</td>
      <td>1133536</td>
      <td>RSPO2</td>
      <td>.</td>
      <td>+</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">tss_file</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(24875, 6)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Get DNA sequence</span>

<span class="n">peak_ids</span> <span class="o">=</span> <span class="n">tss_file</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="n">tss_file</span><span class="p">[</span><span class="s2">&quot;start&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;str&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="n">tss_file</span><span class="p">[</span><span class="s2">&quot;end&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;str&quot;</span><span class="p">)</span>
<span class="n">peak_ids</span> <span class="o">=</span> <span class="n">peak_ids</span><span class="o">.</span><span class="n">to_list</span><span class="p">()</span>

<span class="n">fa</span> <span class="o">=</span> <span class="n">ma</span><span class="o">.</span><span class="n">peak2fasta</span><span class="p">(</span><span class="n">peak_ids</span><span class="p">,</span> <span class="n">ref_genome</span><span class="o">=</span><span class="n">ref_genome</span><span class="p">)</span>
<span class="n">fa</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
24875 sequences
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Show 3 sequences</span>
<span class="n">n</span> <span class="o">=</span> <span class="mi">3</span>

<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="n">v</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">fa</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">v</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">i</span> <span class="o">&gt;=</span> <span class="n">n</span> <span class="o">-</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">break</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
DS562855.1_546866_547965
 TTTGACTGGCTAAAAGAAAAACATGGGAAGAGTCAGTTGTGgtaatcccagcactcacaagtttgaggcaggaggattacctcaacttcgaatccagcctgggatacacagtgcattcaaggccaggctgaactacatagtgagaccctatctcaaaaaacctgggagagaaagagagagggcaagacTAAAAAAGAGGAGAAGGAAACAGGAAAAAAGACTTTGAAGGGAACAGGTTGCAAAGTCCCCGGAACCATGAGCTTATGACCTCTGTCCAACATAAAGGAAAATCAATGCAATAGTTAGTATAGGAAAGACAAACAACGAAAACAAGAAACGGGAGCATTTATTGAAGCTACTCACAAGCAGCTGAgtaccagcaatcgaactcatgccctagtgcttgtcaggcaggcgcttatgccgctgattaaatccccagcccCCTCATACTGTTTTGACTCATTATCATTCCTGCATCCCACAGATATTGCACACTGCACTAAGCAGAGTAACTTGGTGGAAGGACAACTTCCTCCTCTGAGGTCACCTTCCCTCATCTATGGGGAAAGCAATGGCTTTGGAAGCAAATTCCTCGATGCCATTTTAGAAGTGGCCGTTCATTGACActgtacctttgcttcctcctctgaaaaatgaggattaataaaatcaatctattttgatgacttcttaactatgaagacaaaacatgcaaattacctggcaagttgttgtaactaaatacacactcCTTCCAGCCAGTCCCTCCCTCATTCTGTGACCACCCTTGAAGTTGCAGGACCCCAAGGCAAGAGAGAAGTCATGGCTTTCCCCAGTGTAGGTAGGATCTACAGTGCAAGAGGACGTAATCCTGCCCCCAGGAGCCAGGCGGGATGCTCGGCGCCCTAGACTACAACTCCCAGGATGCTCGGCGGGAGGGGCCTAAccccgccctcaggccccgccccccgtgcccagccacgcccggctcccggcctcgcccacccgCGGGCCCTGTCGCTGCCTGGTGAGTGCGGAACGTGCTCGGTTGCGGTGCTGCGCGGGCGCCTGGCGGAAGGTGAGTGCAAGAGCCGGCG

DS562855.1_867638_868737
 TGGGGCGGCGTGATCACGTGGCGTGAAGTGGGCAGGGCCTGTGACGAAATCGCTTCTTATTGGCCAAGTGGGCGGGGTCAGTGTCCACCGAGCACGTGCGCAGAGCTAAGCAGAGAGGATGGGAAAGGGGACGCAGCCGCGGGGCGGGGCTGCTGGAGCCCGGCTGACGGCGTTTGAAGGTGGAGGCGGCCCGGAGACCCGGCCCGCTGTAGGGGCGGGGCCTACGGTACTCCGCCCTGCGTCCGCCCCGTCGCTGTCCGCTTCCGGCACCAGTCGGTGCCGCGGGGTCCTTACCTGGCTAGCGGTCTTCTGCCGCTCCCAGTGATCTGCGAGGTCCGAGTCTACCGAAGCCGCCGTCTTGCTGGGTCTGGCGGAGTGCAGCGGCCGGCTCGGCCGAGGACCTCTTCGGCGTCGCCCCGGGCGTCTGCGCCTCCACCTCGCCGCACCACCAGCCTAGGACCCCAAAGCTGAGCTGCTGCGGACGGAGGGACGGGCCACAGCGCGGACTCAGCGCTTGGAGGCTGCAGCTTGTCCTCCACCCAGGTCTGCGGGAGCTACGGGGCCTGCCCGCGGCCCGCCTGACTCGGCCTCCTGCTCCATCTGTACAAGCTCCCGGGGTCCTCAGGACACACCGGGGTCAGCTGAGACCCGCGTCCTCTGCGTGGAACAATCCGGACATACAGGCTTACACCCTGCCCAGCCTCTCAGAAGCGTTGACACAGAGAGAGAAGGAGGGGGCAGGGAAGGAAAAAACGGGCTCAGTTGTGTCACGCCAGGCGCGTGAACACACCTACTGCACACATACTAATAAAATTTTAAGGGAGGAAAAAAGAGGCAAGTATTAGTGCTATCTGCCACAGGAAAGTCTTGCGAAAATAAAAGACTAGTGAATTTTTAGACTTAAGAGTCTGAATAAGTTTCATTCTCTCGAGACTAAACACAACTTTTTAAAGGAACATCAAATTCAAAACTTTAAAAATGCTCTGTAAAGAAACCCACACACTCTCCCAAGGCATCCGAGCCTGCGAGCCTGGGGTTCAGTGAGACTCAGGTTGGATCCGGGTCCCGTCCTTAAGAACTGGTGACTTTCTACAATTAGC

DS562855.1_849314_850413
 CCCCAGCCAACATCATCACTGCCCTGAAAAGCAGTGCAGGGCTCAGTCTGCAACAGCCCCTGAACCCATGGAAATGGTAGTTCTCCACTGTCTTTGCTGTTTTGGGAGGATATTGGCTATTTTTAGAGTTTGATAAAAGCTGGGGATTGTTTTCCCAGAATCCAGTAATATGGACAATTATATTCAGATATGTCAAAGTGTTCCTAAAGTTCACTCACTAATCCAAGAACATATAATTTTTATAATCAGAATATCACATATAGGTGATCCACACTAAAGGGATCCATCACAAACTTCCTTCATGACAGTGCTATGCAAAAAGAGTAAAAGAATTAAAGGAAAACTTTTTTATTATATTGAAACAAATAAGGAATAATATGAGTAATGTAAAATAAGATCAAGAACTCTGACAACATAAGTGAGCTATCACAGAACAAGGAATCATGCAAGTTTTAAGAACAGCACTAGACTAAAGCCATCTCCCACAAGGCACATGTTTGCAGTtgtctcggttactttcctctcattgctgggacaaaatacacaacactcaaaatgaaggaaagaaaggtttatttcttacagtttttggtggtttcagtccataaccaactggctccaaggcagtggggcacggcactggggcagcacttcacaaagcagcagccagagcaaagggggagtgaacaagccttttctcttggcttacactgtacacaaatcaccatctaccttcagggcaggtgtagcattcacaaaatctccttaccctcctttgtacaggatgaggatatttacaaagtaaaaacccaggctccctgctcaagtctgtcagcgactcaatttatcaagttgacacaatttatcaaccataacaAtccaccccttgtcaacttggcaaccagacatatttctttaaatcaGCCTCCAGATACAAAGGCATGCTCAAGCTTAATGTTTTGCAACTTGTACattataacacagcttaagaattataacaattcttaagaaACAGCATTAAATCTATCAATATTatattatattagaaacataacacttgtccttggacccccaaagactcatacacatccctttatgca

</pre></div></div>
</div>
<p>Looks good</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Remove gff3 file.</span>
<span class="o">!</span>rm ./*gff3
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span>
</pre></div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, Samantha Morris Lab

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>