

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>celloracle.motif_analysis.process_bed_file &mdash; celloracle 0.3.0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../../../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../../../_static/jquery.js"></script>
        <script type="text/javascript" src="../../../_static/underscore.js"></script>
        <script type="text/javascript" src="../../../_static/doctools.js"></script>
        <script type="text/javascript" src="../../../_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../../index.html" class="icon icon-home"> celloracle
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../index.html">Welcome to celloracleâ€™s documentation!</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../installation/index.html">Installation</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/index.html">Tutorial</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/index.html">API</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../changelog.html">Changelog</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../license/index.html">License</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../citation/index.html">Authors and citations</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">celloracle</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../../index.html">Module code</a> &raquo;</li>
        
      <li>celloracle.motif_analysis.process_bed_file</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for celloracle.motif_analysis.process_bed_file</h1><div class="highlight"><pre>
<span></span><span class="c1"># -*- coding: utf-8 -*-</span>
<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">This file contains custom functions for the analysis of ATAC-seq data.</span>
<span class="sd">Genomic activity information (peak of ATAC-seq) will be extracted first.</span>
<span class="sd">Then the peak DNA sequence will be subjected to TF motif scan.</span>
<span class="sd">Finally we will get list of TFs that potentially binds to a specific gene.</span>

<span class="sd">Codes were written by Kenji Kamimoto.</span>


<span class="sd">&#39;&#39;&#39;</span>

<span class="c1">###########################</span>
<span class="c1">### 0. Import libralies ###</span>
<span class="c1">###########################</span>


<span class="c1"># 0.1. libraries for fundamental data science and data processing</span>

<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="kn">import</span> <span class="nn">sys</span><span class="o">,</span> <span class="nn">os</span>

<span class="kn">from</span> <span class="nn">tqdm</span> <span class="k">import</span> <span class="n">tqdm_notebook</span> <span class="k">as</span> <span class="n">tqdm</span>

<span class="c1"># 0.2. libraries for DNA and genome data wrangling and Motif analysis</span>
<span class="kn">from</span> <span class="nn">genomepy</span> <span class="k">import</span> <span class="n">Genome</span>

<span class="c1">#from gimmemotifs.motif import Motif</span>
<span class="kn">from</span> <span class="nn">gimmemotifs.scanner</span> <span class="k">import</span> <span class="n">Scanner</span>
<span class="kn">from</span> <span class="nn">gimmemotifs.fasta</span> <span class="k">import</span> <span class="n">Fasta</span>

<span class="kn">from</span> <span class="nn">pybedtools</span> <span class="k">import</span> <span class="n">BedTool</span>


<span class="c1">####</span>
<span class="c1">### bed f</span>
<span class="k">def</span> <span class="nf">peakstr_to_int</span><span class="p">(</span><span class="n">peak_str</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Take peak name as input and return splitted strs.</span>

<span class="sd">    Args:</span>
<span class="sd">        peak_str (str): peak name.</span>

<span class="sd">    Returns:</span>
<span class="sd">        tuple: splitted peak name.</span>

<span class="sd">    Examples:</span>
<span class="sd">       &gt;&gt;&gt; peakstr_to_int(&quot;chr1_111111_222222&quot;)</span>
<span class="sd">       &quot;chr1&quot;, &quot;111111&quot;, &quot;222222&quot;</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">sp</span> <span class="o">=</span> <span class="n">peak_str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">sp</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">sp</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">sp</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>

<span class="k">def</span> <span class="nf">list_peakstr_to_df</span><span class="p">(</span><span class="n">x</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Convert list of peaks(str) into data frame.</span>

<span class="sd">    Args:</span>
<span class="sd">       x (list of str): list of peak names</span>

<span class="sd">    Returns:</span>
<span class="sd">       pandas.dataframe: peak info as DataFrame</span>

<span class="sd">    Examples:</span>
<span class="sd">       &gt;&gt;&gt; x = [&#39;chr1_3094484_3095479&#39;, &#39;chr1_3113499_3113979&#39;, &#39;chr1_3119478_3121690&#39;]</span>
<span class="sd">       &gt;&gt;&gt; list_peakstr_to_df(x)</span>
<span class="sd">                   chr	start	end</span>
<span class="sd">            0	chr1	3094484	3095479</span>
<span class="sd">            1	chr1	3113499	3113979</span>
<span class="sd">            2	chr1	3119478	3121690</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">peakstr_to_int</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">x</span><span class="p">])</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span> <span class="s2">&quot;start&quot;</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">])</span>
    <span class="n">df</span><span class="p">[</span><span class="s2">&quot;start&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;start&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">int</span><span class="p">)</span>
    <span class="n">df</span><span class="p">[</span><span class="s2">&quot;end&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;end&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">int</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">df</span>

<span class="k">def</span> <span class="nf">df_to_list_peakstr</span><span class="p">(</span><span class="n">x</span><span class="p">):</span>

    <span class="n">x</span> <span class="o">=</span> <span class="n">x</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;chrom&quot;</span><span class="p">:</span><span class="s2">&quot;chr&quot;</span><span class="p">})</span>

    <span class="n">peak_str</span> <span class="o">=</span> <span class="n">x</span><span class="o">.</span><span class="n">chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="n">x</span><span class="o">.</span><span class="n">start</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="n">x</span><span class="o">.</span><span class="n">end</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
    <span class="n">peak_str</span> <span class="o">=</span> <span class="n">peak_str</span><span class="o">.</span><span class="n">values</span>

    <span class="k">return</span> <span class="n">peak_str</span>

<span class="c1">####</span>
<span class="c1">###</span>



<span class="k">def</span> <span class="nf">peak_M1</span><span class="p">(</span><span class="n">peak_id</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Take a peak_id (index of bed file) as input,</span>
<span class="sd">    then subtract 1 from Start position.</span>

<span class="sd">    Args:</span>
<span class="sd">        peak_id (str): Index of bed file. It should be made of &quot;chromosome name&quot;, &quot;start position&quot;, &quot;end position&quot;</span>
<span class="sd">            e.g. &quot;chr11_123445555_123445577&quot;</span>
<span class="sd">    Returns:</span>
<span class="sd">        str: Processed peak_id.</span>

<span class="sd">    Examples:</span>
<span class="sd">        &gt;&gt;&gt; a = &quot;chr11_123445555_123445577&quot;</span>
<span class="sd">        &gt;&gt;&gt; peak_M1(a)</span>
<span class="sd">        &quot;chr11_123445554_123445577&quot;</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">i</span> <span class="o">=</span> <span class="n">peak_id</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="n">i</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>


<div class="viewcode-block" id="peak2fasta"><a class="viewcode-back" href="../../../modules/celloracle.motif_analysis.html#celloracle.motif_analysis.peak2fasta">[docs]</a><span class="k">def</span> <span class="nf">peak2fasta</span><span class="p">(</span><span class="n">peak_ids</span><span class="p">,</span> <span class="n">ref_genome</span><span class="p">):</span>

    <span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Convert peak_id into fasta object.</span>

<span class="sd">    Args:</span>
<span class="sd">        peak_id (str or list of str): Peak_id.  e.g. &quot;chr5_0930303_9499409&quot;</span>
<span class="sd">            or it can be a list of peak_id.  e.g. [&quot;chr5_0930303_9499409&quot;, &quot;chr11_123445555_123445577&quot;]</span>

<span class="sd">        ref_genome (str): Reference genome name.   e.g. &quot;mm9&quot;, &quot;mm10&quot;, &quot;hg19&quot; etc</span>

<span class="sd">    Returns:</span>
<span class="sd">        gimmemotifs fasta object: DNA sequence in fasta format</span>

<span class="sd">    &#39;&#39;&#39;</span>
    <span class="n">genome_data</span> <span class="o">=</span> <span class="n">Genome</span><span class="p">(</span><span class="n">ref_genome</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">peak2seq</span><span class="p">(</span><span class="n">peak_id</span><span class="p">):</span>
        <span class="n">chromosome_name</span> <span class="o">=</span> <span class="n">peak_id</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">locus</span> <span class="o">=</span> <span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">peak_id</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]),</span><span class="nb">int</span><span class="p">(</span><span class="n">peak_id</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">2</span><span class="p">]))</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">genome_data</span><span class="p">[</span><span class="n">chromosome_name</span><span class="p">][</span><span class="n">locus</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">locus</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span>
        <span class="n">name</span> <span class="o">=</span> <span class="n">f</span><span class="s2">&quot;</span><span class="si">{tmp.name}</span><span class="s2">_</span><span class="si">{tmp.start}</span><span class="s2">_</span><span class="si">{tmp.end}</span><span class="s2">&quot;</span>
        <span class="n">seq</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">seq</span>
        <span class="k">return</span> <span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">seq</span><span class="p">)</span>


    <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">peak_ids</span><span class="p">)</span> <span class="ow">is</span> <span class="nb">str</span><span class="p">:</span>
        <span class="n">peak_ids</span> <span class="o">=</span> <span class="p">[</span><span class="n">peak_ids</span><span class="p">]</span>

    <span class="n">fasta</span> <span class="o">=</span> <span class="n">Fasta</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">peak_id</span> <span class="ow">in</span> <span class="n">peak_ids</span><span class="p">:</span>
        <span class="n">name</span><span class="p">,</span> <span class="n">seq</span> <span class="o">=</span> <span class="n">peak2seq</span><span class="p">(</span><span class="n">peak_id</span><span class="p">)</span>
        <span class="n">fasta</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">seq</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">fasta</span></div>



<div class="viewcode-block" id="read_bed"><a class="viewcode-back" href="../../../modules/celloracle.motif_analysis.html#celloracle.motif_analysis.read_bed">[docs]</a><span class="k">def</span> <span class="nf">read_bed</span><span class="p">(</span><span class="n">bed_path</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Load bed file and return as dataframe.</span>

<span class="sd">    Args:</span>
<span class="sd">        bed_path (str): File path.</span>

<span class="sd">    Returns:</span>
<span class="sd">        pandas.dataframe: bed file in dataframe.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">tt</span> <span class="o">=</span> <span class="n">BedTool</span><span class="p">(</span><span class="n">bed_path</span><span class="p">)</span><span class="o">.</span><span class="n">to_dataframe</span><span class="p">()</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">tt</span><span class="p">[</span><span class="s2">&quot;seqname&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">tt</span><span class="o">.</span><span class="n">chrom</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="n">tt</span><span class="o">.</span><span class="n">start</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;int&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;str&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="n">tt</span><span class="o">.</span><span class="n">end</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;int&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;str&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">tt</span></div>
</pre></div>

           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Samantha Morris Lab

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>