

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>celloracle.motif_analysis.tss_annotation &mdash; celloracle 0.3.0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../../../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../../../_static/jquery.js"></script>
        <script type="text/javascript" src="../../../_static/underscore.js"></script>
        <script type="text/javascript" src="../../../_static/doctools.js"></script>
        <script type="text/javascript" src="../../../_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../../index.html" class="icon icon-home"> celloracle
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../index.html">Welcome to celloracle’s documentation!</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../installation/index.html">Installation</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/index.html">Tutorial</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/index.html">API</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../changelog.html">Changelog</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../license/index.html">License</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../citation/index.html">Authors and citations</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">celloracle</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../../index.html">Module code</a> &raquo;</li>
        
      <li>celloracle.motif_analysis.tss_annotation</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for celloracle.motif_analysis.tss_annotation</h1><div class="highlight"><pre>
<span></span><span class="c1"># -*- coding: utf-8 -*-</span>
<span class="sd">&#39;&#39;&#39;</span>
<span class="sd">This file contains custom functions for the analysis of ATAC-seq data.</span>
<span class="sd">Genomic activity information (peak of ATAC-seq) will be extracted first.</span>
<span class="sd">Then the peak DNA sequence will be subjected to TF motif scan.</span>
<span class="sd">Finally we will get list of TFs that potentially binds to a specific gene.</span>

<span class="sd">Codes were written by Kenji Kamimoto.</span>


<span class="sd">&#39;&#39;&#39;</span>

<span class="c1">###########################</span>
<span class="c1">### 0. Import libralies ###</span>
<span class="c1">###########################</span>


<span class="c1"># 0.1. libraries for fundamental data science and data processing</span>

<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="c1">#import matplotlib.pyplot as plt</span>
<span class="c1">#import seaborn as sns</span>

<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">pickle</span>
<span class="kn">import</span> <span class="nn">glob</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">from</span> <span class="nn">copy</span> <span class="k">import</span> <span class="n">deepcopy</span>
<span class="kn">from</span> <span class="nn">datetime</span> <span class="k">import</span> <span class="n">datetime</span>

<span class="kn">from</span> <span class="nn">tqdm</span> <span class="k">import</span> <span class="n">tqdm_notebook</span> <span class="k">as</span> <span class="n">tqdm</span>

<span class="kn">from</span> <span class="nn">pybedtools</span> <span class="k">import</span> <span class="n">BedTool</span>

<span class="kn">from</span> <span class="nn">..motif_analysis</span> <span class="k">import</span> <span class="n">__path__</span> <span class="k">as</span> <span class="n">parent_path</span>
<span class="kn">from</span> <span class="nn">.process_bed_file</span> <span class="k">import</span> <span class="n">list_peakstr_to_df</span>

<span class="k">def</span> <span class="nf">_load_tss_ref_data</span><span class="p">(</span><span class="n">ref_genome</span><span class="p">):</span>

    <span class="sd">&quot;&quot;&quot;</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">parent_path</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="s2">&quot;tss_ref_data&quot;</span><span class="p">,</span> <span class="n">f</span><span class="s2">&quot;</span><span class="si">{ref_genome}</span><span class="s2">_tss_info.bed&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">BedTool</span><span class="p">(</span><span class="n">fn</span><span class="o">=</span><span class="n">path</span><span class="p">)</span>




<div class="viewcode-block" id="get_tss_info"><a class="viewcode-back" href="../../../modules/celloracle.motif_analysis.html#celloracle.motif_analysis.get_tss_info">[docs]</a><span class="k">def</span> <span class="nf">get_tss_info</span><span class="p">(</span><span class="n">peak_str_list</span><span class="p">,</span> <span class="n">ref_genome</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get annotation about Transcription Starting Site (TSS).</span>

<span class="sd">    Args:</span>
<span class="sd">        peak_str_list (list of str): list of peak_id. e.g., [“chr5_0930303_9499409”, “chr11_123445555_123445577”]</span>
<span class="sd">        ref_genome (str): reference genome name.</span>
<span class="sd">        verbose (bool): verbosity.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">ref</span> <span class="o">=</span> <span class="n">_load_tss_ref_data</span><span class="p">(</span><span class="n">ref_genome</span><span class="o">=</span><span class="n">ref_genome</span><span class="p">)</span>

    <span class="n">queue</span> <span class="o">=</span> <span class="n">list_peakstr_to_df</span><span class="p">(</span><span class="n">peak_str_list</span><span class="p">)</span>
    <span class="n">queue</span> <span class="o">=</span> <span class="n">BedTool</span><span class="o">.</span><span class="n">from_dataframe</span><span class="p">(</span><span class="n">queue</span><span class="p">)</span>

    <span class="n">annotated</span> <span class="o">=</span> <span class="n">annotate_tss</span><span class="p">(</span><span class="n">tss_ref_bed</span><span class="o">=</span><span class="n">ref</span><span class="p">,</span> <span class="n">queue_bed</span><span class="o">=</span><span class="n">queue</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">annotated</span></div>

<span class="k">def</span> <span class="nf">annotate_tss</span><span class="p">(</span><span class="n">tss_ref_bed</span><span class="p">,</span> <span class="n">queue_bed</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Search for peaks that exist in TSS.</span>
<span class="sd">    If a peak overlap TSS peak, that peak will be annotated with gene name of the TSS.</span>
<span class="sd">    If a peak does not overlap any TSS peak, the peak will be removed and will not be contained in returned bed file.</span>

<span class="sd">    Args:</span>
<span class="sd">        tss_fer_bed (Bedtool): bedtool object of tss data.</span>
<span class="sd">            This bed file should be a bed file with 5 columns: [&quot;chrom&quot;, &quot;start&quot;, &quot;end&quot;, &quot;name&quot;, &quot;score&quot;, &quot;strand&quot;].</span>
<span class="sd">            Gene name of TSS should be stored in &quot;name&quot; column.</span>

<span class="sd">        queue_bed (Bedtool): bedtool object.</span>
<span class="sd">            This bed file should be a bed file with 3 columns: [&quot;chrom&quot;, &quot;start&quot;, &quot;end&quot;].</span>

<span class="sd">    Returns:</span>
<span class="sd">        dataframe: dataframe of peaks.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># check data structure</span>
    <span class="k">if</span> <span class="p">(</span><span class="n">tss_ref_bed</span><span class="o">.</span><span class="n">to_dataframe</span><span class="p">()</span><span class="o">.</span><span class="n">columns</span> <span class="o">!=</span> <span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">,</span> <span class="s2">&quot;start&quot;</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">,</span> <span class="s2">&quot;name&quot;</span><span class="p">,</span> <span class="s2">&quot;score&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">all</span><span class="p">():</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;tss annotation error&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="p">(</span><span class="n">queue_bed</span><span class="o">.</span><span class="n">to_dataframe</span><span class="p">()</span><span class="o">.</span><span class="n">columns</span> <span class="o">!=</span> <span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">,</span> <span class="s2">&quot;start&quot;</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">all</span><span class="p">():</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;queue bed file error&quot;</span><span class="p">)</span>

    <span class="c1"># intersect</span>
    <span class="n">intersected</span> <span class="o">=</span> <span class="n">tss_ref_bed</span><span class="o">.</span><span class="n">intersect</span><span class="p">(</span><span class="n">queue_bed</span><span class="p">,</span> <span class="n">wb</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">intersected</span> <span class="o">=</span> <span class="n">intersected</span><span class="o">.</span><span class="n">to_dataframe</span><span class="p">()</span>

    <span class="c1"># change name</span>
    <span class="n">intersected</span> <span class="o">=</span> <span class="n">intersected</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;start&quot;</span><span class="p">:</span> <span class="s2">&quot;start_intersected&quot;</span><span class="p">,</span>
                                          <span class="s2">&quot;end&quot;</span><span class="p">:</span> <span class="s2">&quot;end_intersected&quot;</span><span class="p">,</span>
                                          <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;gene_short_name&quot;</span><span class="p">})</span>
    <span class="n">intersected</span> <span class="o">=</span> <span class="n">intersected</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;thickStart&quot;</span><span class="p">:</span> <span class="s2">&quot;chr&quot;</span><span class="p">,</span> <span class="s2">&quot;thickEnd&quot;</span><span class="p">:</span> <span class="s2">&quot;start&quot;</span><span class="p">,</span>
                                          <span class="s2">&quot;itemRgb&quot;</span><span class="p">:</span> <span class="s2">&quot;end&quot;</span><span class="p">})</span>

    <span class="c1"># select data</span>
    <span class="n">intersected</span> <span class="o">=</span> <span class="n">intersected</span><span class="p">[[</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span> <span class="s2">&quot;start&quot;</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">,</span> <span class="s2">&quot;gene_short_name&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">]]</span>

    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;que bed peaks: {len(queue_bed.to_dataframe())}&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;tss peaks in que: {len(intersected)}&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">intersected</span>
</pre></div>

           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Samantha Morris Lab

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>