

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>0. Overview: How to use a different motifs for celloracle motif scan. &mdash; celloracle 0.5.1 documentation</title>
  

  
  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
        <script src="../../../_static/jquery.js"></script>
        <script src="../../../_static/underscore.js"></script>
        <script src="../../../_static/doctools.js"></script>
        <script src="../../../_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="0. Overview: How to Make custom motifs for celloracle motif scan." href="02_How_to_prepare_custom_motif_data.html" />
    <link rel="prev" title="0. Import libraries" href="../02_atac_peaks_to_TFinfo_with_celloracle_20200801.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../../index.html" class="icon icon-home" alt="Documentation Home"> celloracle
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../index.html">Welcome to celloracle’s documentation!</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../installation/index.html">Installation</a></li>
</ul>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="../../../tutorials/index.html">Tutorial</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../../../tutorials/atac.html">1. ATAC-seq data preprocessing</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../../../tutorials/motifscan.html">2. Transcription factor binding motif scan</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="../../../tutorials/motifscan.html#scan-dna-sequences-searching-for-tf-binding-motifs">Scan DNA sequences searching for TF binding motifs</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="../../../tutorials/motifscan.html#optional-step-1-how-to-use-different-motif-data">[Optional step 1] How to use different motif data</a><ul class="current">
<li class="toctree-l4 current"><a class="current reference internal" href="#">0. Overview: How to use a different motifs for celloracle motif scan.</a></li>
<li class="toctree-l4"><a class="reference internal" href="#1-Import-motifs-from-gimmemotifs-dataset.">1 Import motifs from gimmemotifs dataset.</a></li>
<li class="toctree-l4"><a class="reference internal" href="#2.-Import-motifs-from-Celloracle-dataset.">2. Import motifs from Celloracle dataset.</a></li>
<li class="toctree-l4"><a class="reference internal" href="#3.-Import-motifs-from-custom-motif-dataset.">3. Import motifs from custom motif dataset.</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../tutorials/motifscan.html#optional-step-2-how-to-make-custom-motifs-for-celloracle-motif-analysis">[Optional step 2] How to Make custom motifs for celloracle motif analysis</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../tutorials/scrnaprocess.html">3. Single-cell RNA-seq data preprocessing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../tutorials/networkanalysis.html">4. Network analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../tutorials/simulation.html">5. Simulation with GRNs</a></li>
</ul>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/index.html">API</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../changelog/index.html">Changelog</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../license/index.html">License</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../citation/index.html">Authors and citations</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">celloracle</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../../../tutorials/index.html">Tutorial</a> &raquo;</li>
        
          <li><a href="../../../tutorials/motifscan.html"><span class="section-number">2. </span>Transcription factor binding motif scan</a> &raquo;</li>
        
      <li>0. Overview: How to use a different motifs for celloracle motif scan.</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../../../_sources/notebooks/02_motif_scan/motif_data_preparation/01_How_to_load_motif_data.ipynb.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container,
div.nbinput.container div.prompt,
div.nbinput.container div.input_area,
div.nbinput.container div[class*=highlight],
div.nbinput.container div[class*=highlight] pre,
div.nboutput.container,
div.nboutput.container div.prompt,
div.nboutput.container div.output_area,
div.nboutput.container div[class*=highlight],
div.nboutput.container div[class*=highlight] pre {
    background: none;
    border: none;
    padding: 0 0;
    margin: 0;
    box-shadow: none;
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    background: #f5f5f5;
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt a.copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="0.-Overview:-How-to-use-a-different-motifs-for-celloracle-motif-scan.">
<h1>0. Overview: How to use a different motifs for celloracle motif scan.<a class="headerlink" href="#0.-Overview:-How-to-use-a-different-motifs-for-celloracle-motif-scan." title="Permalink to this headline">¶</a></h1>
<p>In this notebook, we introduce how to prepare motif dataset for celloracle motif analysis. Celloracle uses list of motif object in gimmemotifs package. See gimmemotifs documentation for more details. (<a class="reference external" href="https://gimmemotifs.readthedocs.io/en/master/api">https://gimmemotifs.readthedocs.io/en/master/api</a>.html#)</p>
</div>
<div class="section" id="1-Import-motifs-from-gimmemotifs-dataset.">
<h1>1 Import motifs from gimmemotifs dataset.<a class="headerlink" href="#1-Import-motifs-from-gimmemotifs-dataset." title="Permalink to this headline">¶</a></h1>
<p>Gimmemotifs provides many motif dataset that was generated from public motif database including CisDB, ENCODE, HOMER, and JASPAR. <a class="reference external" href="https://gimmemotifs.readthedocs.io/en/master/overview.html">https://gimmemotifs.readthedocs.io/en/master/overview.html</a></p>
<div class="section" id="1.1-gimme.vertebrate.v5.0.">
<h2>1.1 gimme.vertebrate.v5.0.<a class="headerlink" href="#1.1-gimme.vertebrate.v5.0." title="Permalink to this headline">¶</a></h2>
<p>By default GimmeMotifs uses a non-redundant, clustered database of known vertebrate motifs. These motifs come from CIS-BP (<a class="reference external" href="http://cisbp.ccbr.utoronto.ca/">http://cisbp.ccbr.utoronto.ca/</a>) and other sources. This motif dataset can be easily loaded with the following command.</p>
<p>If your dataset is Mouse or Human, this one will be a good default choice.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Compare with default motifs in gimmemotifs</span>
<span class="kn">from</span> <span class="nn">gimmemotifs.motif</span> <span class="kn">import</span> <span class="n">default_motifs</span>
<span class="n">motifs</span> <span class="o">=</span>  <span class="n">default_motifs</span><span class="p">()</span>

<span class="c1"># Check first 10 motifs</span>
<span class="n">motifs</span><span class="p">[:</span><span class="mi">10</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/k/anaconda3/envs/pandas1/lib/python3.6/site-packages/gimmemotifs/plot.py:19: MatplotlibDeprecationWarning: The &#39;warn&#39; parameter of use() is deprecated since Matplotlib 3.1 and will be removed in 3.3.  If any parameter follows &#39;warn&#39;, they should be pass as keyword, not positionally.
  mpl.use(&#34;Agg&#34;, warn=False)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[GM.5.0.Sox.0001_AACAAT,
 GM.5.0.Homeodomain.0001_AGCTGTCAnnA,
 GM.5.0.Mixed.0001_snnGGsssGGs,
 GM.5.0.Nuclear_receptor.0001_TAwsTrGGTCAsTrGGTCA,
 GM.5.0.Mixed.0002_GCTAATTA,
 GM.5.0.Nuclear_receptor.0002_wnyrCTTCCGGGkC,
 GM.5.0.bHLH.0001_ACGTG,
 GM.5.0.Myb_SANT.0001_rrCCGTTAAACnGyy,
 GM.5.0.C2H2_ZF.0001_GCGkGGGCGG,
 GM.5.0.GATA.0001_TTATCTsnnnnnnnCA]
</pre></div></div>
</div>
</div>
<div class="section" id="1.2-Motifs-that-are-provided-with-gimmemotifs-package">
<h2>1.2 Motifs that are provided with gimmemotifs package<a class="headerlink" href="#1.2-Motifs-that-are-provided-with-gimmemotifs-package" title="Permalink to this headline">¶</a></h2>
<p>Many other motif databases come included with GimmeMotifs You can load them as follows.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Get folder path that stores motif data.</span>
<span class="kn">import</span> <span class="nn">os</span><span class="o">,</span> <span class="nn">glob</span>
<span class="kn">from</span> <span class="nn">gimmemotifs.motif</span> <span class="kn">import</span> <span class="n">MotifConfig</span>
<span class="n">config</span> <span class="o">=</span> <span class="n">MotifConfig</span><span class="p">()</span>
<span class="n">motif_dir</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">get_motif_dir</span><span class="p">()</span>

<span class="c1"># Get motif data names</span>
<span class="n">motifs_data_name</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">motif_dir</span><span class="p">)</span> <span class="k">if</span> <span class="n">i</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.pfm&quot;</span><span class="p">)]</span>
<span class="n">motifs_data_name</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
<span class="n">motifs_data_name</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[&#39;CIS-BP.pfm&#39;,
 &#39;ENCODE.pfm&#39;,
 &#39;HOCOMOCOv10_HUMAN.pfm&#39;,
 &#39;HOCOMOCOv10_MOUSE.pfm&#39;,
 &#39;HOCOMOCOv11_HUMAN.pfm&#39;,
 &#39;HOCOMOCOv11_MOUSE.pfm&#39;,
 &#39;HOMER.pfm&#39;,
 &#39;IMAGE.pfm&#39;,
 &#39;JASPAR2018.pfm&#39;,
 &#39;JASPAR2018_fungi.pfm&#39;,
 &#39;JASPAR2018_insects.pfm&#39;,
 &#39;JASPAR2018_nematodes.pfm&#39;,
 &#39;JASPAR2018_plants.pfm&#39;,
 &#39;JASPAR2018_urochordates.pfm&#39;,
 &#39;JASPAR2018_vertebrates.pfm&#39;,
 &#39;RSAT_insects.pfm&#39;,
 &#39;RSAT_plants.pfm&#39;,
 &#39;RSAT_vertebrates.pfm&#39;,
 &#39;SwissRegulon.pfm&#39;,
 &#39;factorbook.pfm&#39;,
 &#39;gimme.vertebrate.v5.0.pfm&#39;]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># You can load motif files with &quot;read_motifs&quot;</span>
<span class="kn">from</span> <span class="nn">gimmemotifs.motif</span> <span class="kn">import</span> <span class="n">read_motifs</span>

<span class="n">path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">motif_dir</span><span class="p">,</span> <span class="s2">&quot;JASPAR2018_plants.pfm&quot;</span><span class="p">)</span>
<span class="n">motifs</span> <span class="o">=</span> <span class="n">read_motifs</span><span class="p">(</span><span class="n">path</span><span class="p">)</span>

<span class="c1"># Check first 10 motifs</span>
<span class="n">motifs</span><span class="p">[:</span><span class="mi">10</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[MA0020.1_Dof2_AAAGCn,
 MA0021.1_Dof3_AAAGyn,
 MA0034.1_Gam1_nnyAACCGmC,
 MA0044.1_HMG-1_sTTGTnyTy,
 MA0045.1_HMG-I/Y_nwAnAAAnrnmrAmAy,
 MA0053.1_MNB1A_AAAGC,
 MA0054.1_myb.Ph3_TAACnGTTw,
 MA0064.1_PBF_AAAGy,
 MA0082.1_squamosa_mCAwAwATrGwAAn,
 MA0096.1_bZIP910_mTGACGT]
</pre></div></div>
</div>
</div>
</div>
<div class="section" id="2.-Import-motifs-from-Celloracle-dataset.">
<h1>2. Import motifs from Celloracle dataset.<a class="headerlink" href="#2.-Import-motifs-from-Celloracle-dataset." title="Permalink to this headline">¶</a></h1>
<p>Celloracle provides many motif dataset that was generated from CisDB. These motifs were divided by species. Please select motifs for your species.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[37]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">celloracle</span> <span class="kn">import</span> <span class="n">motif_analysis</span> <span class="k">as</span> <span class="n">ma</span>

<span class="c1"># Check available motifs</span>
<span class="n">ma</span><span class="o">.</span><span class="n">MOTIFS_LIST</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[37]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[&#39;CisDB_ver2_Mus_musculus.pfm&#39;,
 &#39;CisDB_ver2_Saccharomyces_cerevisiae.pfm&#39;,
 &#39;CisDB_ver2_Danio_rerio.pfm&#39;,
 &#39;CisDB_ver2_Homo_sapiens.pfm&#39;]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[39]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Load motifs from celloracle dataset.</span>
<span class="n">motifs</span> <span class="o">=</span> <span class="n">ma</span><span class="o">.</span><span class="n">load_motifs</span><span class="p">(</span><span class="s2">&quot;CisDB_ver2_Homo_sapiens.pfm&quot;</span><span class="p">)</span>

<span class="c1"># Check first 10 motifs</span>
<span class="n">motifs</span><span class="p">[:</span><span class="mi">10</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[39]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[M00056_2.00_TAATAAAT,
 M00070_2.00_nrAACAATAnn,
 M00111_2.00_nGCCynnGGs,
 M00112_2.00_CCTsrGGCnA,
 M00113_2.00_nsCCnnAGGs,
 M00114_2.00_nnGCCynnGG,
 M00115_2.00_nnATnAAAn,
 M00116_2.00_nnAATATTAnn,
 M00130_2.00_nnnGCCCCnn,
 M00142_2.00_GTrCTCmy]
</pre></div></div>
</div>
</div>
<div class="section" id="3.-Import-motifs-from-custom-motif-dataset.">
<h1>3. Import motifs from custom motif dataset.<a class="headerlink" href="#3.-Import-motifs-from-custom-motif-dataset." title="Permalink to this headline">¶</a></h1>
<p>If you want to use another motif data source, you need to make a list of motif class in gimmemotifs. The easiest way to make such object is to use “read_motifs” function in gimmemotifs.</p>
<p>This function can load motif data text file. You need to prepare two files, XXX.motif2factors.txt and XXX.pfm.</p>
<div class="section" id="3.1-XXX.motif2factors.txt">
<h2>3.1 XXX.motif2factors.txt<a class="headerlink" href="#3.1-XXX.motif2factors.txt" title="Permalink to this headline">¶</a></h2>
<p>The text file, XXX.motif2factors.txt includes TF factor annotation for each motifs. The file should be like a tsv file like below.</p>
<p>The first column should be motif name, the motif name should match with motif name in pfm file. The second column is gene symbol, the thrid column is datasource. This column is not important. The forth column is additional information for this factor. The factor is labeled with “Y” If factor information was confirmed by some evidence. Otherwise, the factor is labeled with “N”.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[52]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">path_motif2factors</span> <span class="o">=</span> <span class="n">path</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.pfm&quot;</span><span class="p">,</span> <span class="s2">&quot;.motif2factors.txt&quot;</span><span class="p">)</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">path_motif2factors</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">f</span><span class="p">):</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">j</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">i</span><span class="o">&gt;</span><span class="mi">5</span><span class="p">:</span>
            <span class="k">break</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Motif   Factor  Evidence        Curated

MA0020.1_Dof2   Dof2    SELEX   Y

MA0021.1_Dof3   Dof3    SELEX   Y

MA0034.1_Gam1   Gam1    SELEX   Y

MA0044.1_HMG-1  HMG-1   SELEX   Y

MA0045.1_HMG-I/Y        HMG-I/Y SELEX   Y

MA0053.1_MNB1A  MNB1A   SELEX   Y

</pre></div></div>
</div>
</div>
<div class="section" id="3.2-XXX.pfm">
<h2>3.2 XXX.pfm<a class="headerlink" href="#3.2-XXX.pfm" title="Permalink to this headline">¶</a></h2>
<p>The second file, XXX.pfm. should includes motif pwm information. The file shoud be like below.</p>
<p>The motif name in this pfm file should exactly match with the motif name in motif2factor.txt file.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[55]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">f</span><span class="p">):</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">j</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">i</span><span class="o">&gt;</span><span class="mi">10</span><span class="p">:</span>
            <span class="k">break</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
# JASPAR2018_plants motif database

# Retrieved from: http://jaspar.genereg.net/download/CORE/JASPAR2018_CORE_plants_non-redundant_pfms_jaspar.txt

# Date: 2018-10-17

&gt;MA0020.1_Dof2

0.9999  0.0000  0.0000  0.0000

0.9999  0.0000  0.0000  0.0000

0.9999  0.0000  0.0000  0.0000

0.0000  0.0000  0.9999  0.0000

0.1429  0.6666  0.0953  0.0953

0.3333  0.2857  0.1429  0.2381

&gt;MA0021.1_Dof3

0.9999  0.0000  0.0000  0.0000

</pre></div></div>
</div>
</div>
<div class="section" id="3.3-Load-files-as-motif-list">
<h2>3.3 Load files as motif list<a class="headerlink" href="#3.3-Load-files-as-motif-list" title="Permalink to this headline">¶</a></h2>
<p>We can load files using read_motifs function in gimmemotifs.</p>
<p>First, please prepare two files, XXX.motif2factors.txt and XXX.pfm. in the same directly. If you have theses two file in a different place, we cannot use the read_motifs function.</p>
<p>Then use file path for XXX.pfm for the argument of read_motifs function.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[58]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">gimmemotifs.motif</span> <span class="kn">import</span> <span class="n">read_motifs</span>

<span class="c1"># Check path for pfm file</span>
<span class="nb">print</span><span class="p">(</span><span class="n">path</span><span class="p">)</span>

<span class="c1"># Read motifs</span>
<span class="n">motifs</span> <span class="o">=</span> <span class="n">read_motifs</span><span class="p">(</span><span class="n">path</span><span class="p">)</span>

<span class="c1"># Check first 10 motifs</span>
<span class="n">motifs</span><span class="p">[:</span><span class="mi">10</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
/home/k/anaconda3/envs/pandas1/lib/python3.6/site-packages/data/motif_databases/JASPAR2018_plants.pfm
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[58]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[MA0020.1_Dof2_AAAGCn,
 MA0021.1_Dof3_AAAGyn,
 MA0034.1_Gam1_nnyAACCGmC,
 MA0044.1_HMG-1_sTTGTnyTy,
 MA0045.1_HMG-I/Y_nwAnAAAnrnmrAmAy,
 MA0053.1_MNB1A_AAAGC,
 MA0054.1_myb.Ph3_TAACnGTTw,
 MA0064.1_PBF_AAAGy,
 MA0082.1_squamosa_mCAwAwATrGwAAn,
 MA0096.1_bZIP910_mTGACGT]
</pre></div></div>
</div>
<p>In another notebook, we introduce how to make XXX.pfm file and XXX.motif2factors.txt file. Pleas look at that if you want to make your motif data by yourself.</p>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="02_How_to_prepare_custom_motif_data.html" class="btn btn-neutral float-right" title="0. Overview: How to Make custom motifs for celloracle motif scan." accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="../02_atac_peaks_to_TFinfo_with_celloracle_20200801.html" class="btn btn-neutral float-left" title="0. Import libraries" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, Samantha Morris Lab

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>