

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>0. Overview: How to Make custom motifs for celloracle motif scan. &mdash; celloracle 0.5.0 documentation</title>
  

  
  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
        <script src="../../../_static/jquery.js"></script>
        <script src="../../../_static/underscore.js"></script>
        <script src="../../../_static/doctools.js"></script>
        <script src="../../../_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="3. Single-cell RNA-seq data preprocessing" href="../../../tutorials/scrnaprocess.html" />
    <link rel="prev" title="0. Overview: How to use a different motifs for celloracle motif scan." href="01_How_to_load_motif_data.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../../index.html" class="icon icon-home" alt="Documentation Home"> celloracle
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../index.html">Welcome to celloracleâ€™s documentation!</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../installation/index.html">Installation</a></li>
</ul>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="../../../tutorials/index.html">Tutorial</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../../../tutorials/atac.html">1. ATAC-seq data preprocessing</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../../../tutorials/motifscan.html">2. Transcription factor binding motif scan</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="../../../tutorials/motifscan.html#scan-dna-sequences-searching-for-tf-binding-motifs">Scan DNA sequences searching for TF binding motifs</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../tutorials/motifscan.html#optional-step-1-how-to-use-different-motif-data">[Optional step 1] How to use different motif data</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="../../../tutorials/motifscan.html#optional-step-2-how-to-make-custom-motifs-for-celloracle-motif-analysis">[Optional step 2] How to Make custom motifs for celloracle motif analysis</a><ul class="current">
<li class="toctree-l4 current"><a class="current reference internal" href="#">0. Overview: How to Make custom motifs for celloracle motif scan.</a></li>
<li class="toctree-l4"><a class="reference internal" href="#1.-Download-full-dataset-for-TF-dataset-from-CisBP-dadabase">1. Download full dataset for TF dataset from CisBP dadabase</a></li>
<li class="toctree-l4"><a class="reference internal" href="#2.-Define-custom-functions">2. Define custom functions</a></li>
<li class="toctree-l4"><a class="reference internal" href="#3.-Pick-up-motif-information-for-one-species-and-save-as-gimmemotif-pfm-file-format.">3. Pick up motif information for one species and save as gimmemotif pfm file format.</a></li>
<li class="toctree-l4"><a class="reference internal" href="#4.-Check-results">4. Check results</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../tutorials/scrnaprocess.html">3. Single-cell RNA-seq data preprocessing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../tutorials/networkanalysis.html">4. Network analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../tutorials/simulation.html">5. Simulation with GRNs</a></li>
</ul>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/index.html">API</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../changelog/index.html">Changelog</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../license/index.html">License</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../citation/index.html">Authors and citations</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">celloracle</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../../../tutorials/index.html">Tutorial</a> &raquo;</li>
        
          <li><a href="../../../tutorials/motifscan.html"><span class="section-number">2. </span>Transcription factor binding motif scan</a> &raquo;</li>
        
      <li>0. Overview: How to Make custom motifs for celloracle motif scan.</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../../../_sources/notebooks/02_motif_scan/motif_data_preparation/02_How_to_prepare_custom_motif_data.ipynb.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container,
div.nbinput.container div.prompt,
div.nbinput.container div.input_area,
div.nbinput.container div[class*=highlight],
div.nbinput.container div[class*=highlight] pre,
div.nboutput.container,
div.nboutput.container div.prompt,
div.nboutput.container div.output_area,
div.nboutput.container div[class*=highlight],
div.nboutput.container div[class*=highlight] pre {
    background: none;
    border: none;
    padding: 0 0;
    margin: 0;
    box-shadow: none;
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    background: #f5f5f5;
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt a.copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="0.-Overview:-How-to-Make-custom-motifs-for-celloracle-motif-scan.">
<h1>0. Overview: How to Make custom motifs for celloracle motif scan.<a class="headerlink" href="#0.-Overview:-How-to-Make-custom-motifs-for-celloracle-motif-scan." title="Permalink to this headline">Â¶</a></h1>
<p>In this notebook, we introduce how to make motif dataset for celloracle motif analysis. Celloracle uses list of motif object in gimmemotifs package. See gimmemotifs documentation for more details. (<a class="reference external" href="https://gimmemotifs.readthedocs.io/en/master/api">https://gimmemotifs.readthedocs.io/en/master/api</a>.html#)</p>
<p>Here, we get motif data from CisBP (version2).http://cisbp.ccbr.utoronto.ca</p>
<p>We will extract motif information for a specific species and save as XXX.pfm and XXX.motif2facrors.txt file. These files can be read with read_motifs function in gimmemotifs.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>

<span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="kn">import</span> <span class="nn">sys</span><span class="o">,</span> <span class="nn">os</span>

<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="o">%</span><span class="k">matplotlib</span> inline
<span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s2">&quot;figure.figsize&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="mi">20</span><span class="p">,</span><span class="mi">10</span><span class="p">]</span>

<span class="kn">from</span> <span class="nn">gimmemotifs.motif</span> <span class="kn">import</span> <span class="n">Motif</span><span class="p">,</span><span class="n">read_motifs</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="1.-Download-full-dataset-for-TF-dataset-from-CisBP-dadabase">
<h1>1. Download full dataset for TF dataset from CisBP dadabase<a class="headerlink" href="#1.-Download-full-dataset-for-TF-dataset-from-CisBP-dadabase" title="Permalink to this headline">Â¶</a></h1>
<p>Go the URL below for the download link for the latest data. <a class="reference external" href="http://cisbp.ccbr.utoronto.ca/entireDownload.php">http://cisbp.ccbr.utoronto.ca/entireDownload.php</a></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="o">!</span> wget http://cisbp.ccbr.utoronto.ca/data/2.00/DataFiles/Bulk_downloads/EntireDataset/PWMs.zip
<span class="o">!</span> unzip PWMs.zip
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="o">!</span>wget http://cisbp.ccbr.utoronto.ca/data/2.00/DataFiles/Bulk_downloads/EntireDataset/TF_Information_all_motifs.txt.zip
<span class="o">!</span> unzip TF_Information_all_motifs.txt.zip
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
--2020-08-03 16:34:06--  http://cisbp.ccbr.utoronto.ca/data/2.00/DataFiles/Bulk_downloads/EntireDataset/TF_Information_all_motifs.txt.zip
Resolving cisbp.ccbr.utoronto.ca (cisbp.ccbr.utoronto.ca)... 142.150.52.218
Connecting to cisbp.ccbr.utoronto.ca (cisbp.ccbr.utoronto.ca)|142.150.52.218|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 115455298 (110M) [application/zip]
Saving to: â€˜TF_Information_all_motifs.txt.zipâ€™

TF_Information_all_ 100%[===================&gt;] 110.11M  11.1MB/s    in 19s

2020-08-03 16:34:25 (5.73 MB/s) - â€˜TF_Information_all_motifs.txt.zipâ€™ saved [115455298/115455298]

Archive:  TF_Information_all_motifs.txt.zip
  inflating: TF_Information_all_motifs.txt
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Load TF information as a dataframe.</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span><span class="s2">&quot;TF_Information_all_motifs.txt&quot;</span><span class="p">)</span>
<span class="n">df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/k/anaconda3/envs/pandas1/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3072: DtypeWarning: Columns (19) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>TF_ID</th>
      <th>Family_ID</th>
      <th>TSource_ID</th>
      <th>Motif_ID</th>
      <th>MSource_ID</th>
      <th>DBID</th>
      <th>TF_Name</th>
      <th>TF_Species</th>
      <th>TF_Status</th>
      <th>Family_Name</th>
      <th>...</th>
      <th>MSource_Year</th>
      <th>PMID</th>
      <th>MSource_Version</th>
      <th>SR_Model</th>
      <th>SR_NoThreshold</th>
      <th>TfSource_Name</th>
      <th>TfSource_URL</th>
      <th>TfSource_Year</th>
      <th>TfSource_Month</th>
      <th>TfSource_Day</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>T000001_2.00</td>
      <td>F001_2.00</td>
      <td>TS12_2.00</td>
      <td>.</td>
      <td>.</td>
      <td>BRADI2G60554</td>
      <td>BRADI2G60554</td>
      <td>Brachypodium_distachyon</td>
      <td>N</td>
      <td>ABF1,B3</td>
      <td>...</td>
      <td>.</td>
      <td>.</td>
      <td>.</td>
      <td>SequenceIdentity</td>
      <td>True</td>
      <td>Ensembl</td>
      <td>http://www.ensembl.org/</td>
      <td>2018</td>
      <td>Dec</td>
      <td>8</td>
    </tr>
    <tr>
      <th>1</th>
      <td>T000002_2.00</td>
      <td>F001_2.00</td>
      <td>TS12_2.00</td>
      <td>.</td>
      <td>.</td>
      <td>LPERR05G06870</td>
      <td>LPERR05G06870</td>
      <td>Leersia_perrieri</td>
      <td>N</td>
      <td>ABF1,B3</td>
      <td>...</td>
      <td>.</td>
      <td>.</td>
      <td>.</td>
      <td>SequenceIdentity</td>
      <td>True</td>
      <td>Ensembl</td>
      <td>http://www.ensembl.org/</td>
      <td>2018</td>
      <td>Dec</td>
      <td>8</td>
    </tr>
    <tr>
      <th>2</th>
      <td>T000003_2.00</td>
      <td>F002_2.00</td>
      <td>TS04_2.00</td>
      <td>.</td>
      <td>.</td>
      <td>CPAG_02544</td>
      <td>CPAG_02544</td>
      <td>Candida_parapsilosis</td>
      <td>N</td>
      <td>ABF1</td>
      <td>...</td>
      <td>.</td>
      <td>.</td>
      <td>.</td>
      <td>SequenceIdentity</td>
      <td>True</td>
      <td>Broad</td>
      <td>http://www.broadinstitute.org/</td>
      <td>2016</td>
      <td>May</td>
      <td>1</td>
    </tr>
    <tr>
      <th>3</th>
      <td>T000004_2.00</td>
      <td>F002_2.00</td>
      <td>TS04_2.00</td>
      <td>.</td>
      <td>.</td>
      <td>PTSG_00627</td>
      <td>PTSG_00627</td>
      <td>Salpingoeca_rosetta</td>
      <td>N</td>
      <td>ABF1</td>
      <td>...</td>
      <td>.</td>
      <td>.</td>
      <td>.</td>
      <td>SequenceIdentity</td>
      <td>True</td>
      <td>Broad</td>
      <td>http://www.broadinstitute.org/</td>
      <td>2016</td>
      <td>May</td>
      <td>1</td>
    </tr>
    <tr>
      <th>4</th>
      <td>T000005_2.00</td>
      <td>F002_2.00</td>
      <td>TS04_2.00</td>
      <td>.</td>
      <td>.</td>
      <td>WUBG_06707</td>
      <td>WUBG_06707</td>
      <td>Wuchereria_bancrofti</td>
      <td>N</td>
      <td>ABF1</td>
      <td>...</td>
      <td>.</td>
      <td>.</td>
      <td>.</td>
      <td>SequenceIdentity</td>
      <td>True</td>
      <td>Broad</td>
      <td>http://www.broadinstitute.org/</td>
      <td>2016</td>
      <td>May</td>
      <td>1</td>
    </tr>
  </tbody>
</table>
<p>5 rows Ã— 28 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">df</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(10879322, 28)
</pre></div></div>
</div>
</div>
<div class="section" id="2.-Define-custom-functions">
<h1>2. Define custom functions<a class="headerlink" href="#2.-Define-custom-functions" title="Permalink to this headline">Â¶</a></h1>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># All process will be done inside these function.</span>

<span class="kn">from</span> <span class="nn">datetime</span> <span class="kn">import</span> <span class="n">datetime</span>
<span class="kn">import</span> <span class="nn">glob</span>

<span class="k">def</span> <span class="nf">read_pwn_and_convert_into_list</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
    <span class="c1"># read pwn as df</span>
    <span class="n">pwm</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># convert into list of str</span>
    <span class="n">li</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">pwm</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="n">i</span> <span class="o">=</span> <span class="n">i</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span>
        <span class="n">i</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;str&quot;</span><span class="p">))</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="n">li</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">li</span>

<span class="k">def</span> <span class="nf">make_motif_file_from_cisbp_data</span><span class="p">(</span><span class="n">pwm_folder_path</span><span class="p">,</span> <span class="n">tfinfo_df</span><span class="p">,</span> <span class="n">species</span><span class="p">):</span>

    <span class="n">data_</span> <span class="o">=</span> <span class="n">tfinfo_df</span><span class="p">[</span><span class="n">tfinfo_df</span><span class="o">.</span><span class="n">TF_Species</span> <span class="o">==</span> <span class="n">species</span><span class="p">]</span>
    <span class="n">data_name</span> <span class="o">=</span> <span class="s2">&quot;CisDB_ver2_&quot;</span> <span class="o">+</span> <span class="n">species</span>

    <span class="c1">## 1. Make file: motif2factors.txt</span>

    <span class="c1"># Select information</span>
    <span class="n">df_factors</span> <span class="o">=</span> <span class="n">data_</span><span class="p">[[</span><span class="s2">&quot;Motif_ID&quot;</span><span class="p">,</span> <span class="s2">&quot;TF_Name&quot;</span><span class="p">,</span> <span class="s2">&quot;MSource_Type&quot;</span><span class="p">,</span> <span class="s2">&quot;TF_Status&quot;</span><span class="p">]]</span>
    <span class="n">df_factors</span> <span class="o">=</span> <span class="n">df_factors</span><span class="p">[</span><span class="n">df_factors</span><span class="o">.</span><span class="n">TF_Status</span> <span class="o">!=</span> <span class="s2">&quot;N&quot;</span><span class="p">]</span>

    <span class="c1"># Formatting</span>
    <span class="n">df_factors</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="s1">&#39;Motif</span><span class="se">\t</span><span class="s1">Factor</span><span class="se">\t</span><span class="s1">Evidence</span><span class="se">\t</span><span class="s1">Curated&#39;</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">df_factors</span><span class="p">[</span><span class="s2">&quot;Curated&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[{</span><span class="s2">&quot;D&quot;</span><span class="p">:</span> <span class="s2">&quot;Y&quot;</span><span class="p">,</span> <span class="s2">&quot;I&quot;</span><span class="p">:</span> <span class="s2">&quot;N&quot;</span><span class="p">}[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">df_factors</span><span class="p">[</span><span class="s2">&quot;Curated&quot;</span><span class="p">]]</span>
    <span class="n">df_factors</span> <span class="o">=</span> <span class="n">df_factors</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s2">&quot;Motif&quot;</span><span class="p">)</span>

    <span class="c1">## 2. Make file: pfm file</span>
    <span class="n">comments</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;# CIS-BP motif database (v2.0), retrieved by Celloracle</span><span class="se">\n</span><span class="s2">&quot;</span>
    <span class="n">comments</span> <span class="o">+=</span> <span class="s2">&quot;# Retrieved from: http://cisbp.ccbr.utoronto.ca/data/2.00/DataFiles/Bulk_downloads/EntireDataset/PWMs.zip</span><span class="se">\n</span><span class="s2">&quot;</span>
    <span class="n">comments</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;#Date: </span><span class="si">{</span><span class="n">datetime</span><span class="o">.</span><span class="n">now</span><span class="p">()</span><span class="o">.</span><span class="n">ctime</span><span class="p">()</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>

    <span class="c1"># Get list of motif name</span>
    <span class="n">paths_pwm</span> <span class="o">=</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">pwm_folder_path</span><span class="p">,</span> <span class="s2">&quot;*.txt&quot;</span><span class="p">))</span>
    <span class="n">paths_pwm</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
    <span class="n">motif_names</span> <span class="o">=</span> <span class="p">[</span><span class="n">path</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;/&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.txt&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">path</span> <span class="ow">in</span> <span class="n">paths_pwm</span><span class="p">]</span>
    <span class="n">motifs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">intersect1d</span><span class="p">(</span><span class="n">motif_names</span><span class="p">,</span> <span class="n">df_factors</span><span class="o">.</span><span class="n">Motif</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>

    <span class="nb">print</span><span class="p">(</span><span class="n">motifs</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>

    <span class="c1"># Intersect motif information with pwm information</span>
    <span class="n">df_factors</span> <span class="o">=</span> <span class="n">df_factors</span><span class="p">[</span><span class="n">df_factors</span><span class="o">.</span><span class="n">Motif</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">motifs</span><span class="p">)]</span>

    <span class="c1"># Load, convert, and save pwm info</span>
    <span class="n">output</span> <span class="o">=</span> <span class="n">data_name</span> <span class="o">+</span> <span class="s2">&quot;.pfm&quot;</span>

    <span class="n">motifs_non_zero</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">output</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>

        <span class="k">for</span> <span class="n">motif_name</span> <span class="ow">in</span> <span class="n">motifs</span><span class="p">:</span>

            <span class="n">path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">pwm_folder_path</span><span class="p">,</span> <span class="n">motif_name</span> <span class="o">+</span> <span class="s2">&quot;.txt&quot;</span><span class="p">)</span>
            <span class="n">pwm</span> <span class="o">=</span> <span class="n">read_pwn_and_convert_into_list</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="n">path</span><span class="p">)</span> <span class="c1"># Load and convert</span>
            <span class="k">if</span> <span class="n">pwm</span><span class="p">:</span>
                <span class="n">motifs_non_zero</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">motif_name</span><span class="p">)</span>
                <span class="n">pwm</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;&gt;</span><span class="si">{</span><span class="n">motif_name</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">pwm</span>
                <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">pwm</span><span class="p">:</span> <span class="c1"># Save pfm</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>


    <span class="c1"># Intersect motif information with pwm information</span>
    <span class="n">df_factors</span> <span class="o">=</span> <span class="n">df_factors</span><span class="p">[</span><span class="n">df_factors</span><span class="o">.</span><span class="n">Motif</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">motifs_non_zero</span><span class="p">)]</span>

    <span class="c1"># Save factor info</span>
    <span class="n">df_factors</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">data_name</span><span class="si">}</span><span class="s1">.motif2factors.txt&#39;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span><span class="n">df_factors</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">motifs_non_zero</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="3.-Pick-up-motif-information-for-one-species-and-save-as-gimmemotif-pfm-file-format.">
<h1>3. Pick up motif information for one species and save as gimmemotif pfm file format.<a class="headerlink" href="#3.-Pick-up-motif-information-for-one-species-and-save-as-gimmemotif-pfm-file-format." title="Permalink to this headline">Â¶</a></h1>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Check species in this dataset</span>
<span class="n">species_list</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">TF_Species</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
<span class="n">species_list</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>

<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">species_list</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>

</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Acanthamoeba_castellanii
Acanthamoeba_polyphaga_mimivirus
Acanthocheilonema_viteae
Acipenser_baerii
Acremonium_chrysogenum
Acropora_formosa
Acropora_millepora
Acyrthosiphon_pisum
Aedes_aegypti
Aegilops_tauschii
Agaricus_bisporus
Ailuropoda_melanoleuca
Albugo_laibachii
Alligator_sinensis
Allomyces_macrogynus
Alternaria_brassicicola
Amanita_muscaria
Amborella_trichopoda
Amphimedon_queenslandica
Anas_platyrhynchos
Ancylostoma_caninum
Ancylostoma_ceylanicum
Ancylostoma_duodenale
Angiostrongylus_cantonensis
Angiostrongylus_costaricensis
Anisakis_simplex
Anncaliia_algerae
Anolis_carolinensis
Anopheles_albimanus
Anopheles_arabiensis
Anopheles_atroparvus
Anopheles_christyi
Anopheles_coluzzii
Anopheles_culicifacies
Anopheles_darlingi
Anopheles_dirus
Anopheles_epiroticus
Anopheles_farauti
Anopheles_funestus
Anopheles_gambiae
Anopheles_maculatus
Anopheles_melas
Anopheles_merus
Anopheles_minimus
Anopheles_quadriannulatus
Anopheles_sinensis
Anopheles_stephensi
Antirrhinum_majus
Apis_mellifera
Aplysia_californica
Aptenodytes_forsteri
Aquilegia_coerulea
Arabidopsis_lyrata
Arabidopsis_thaliana
Artemia_franciscana
Arthrobotrys_oligospora
Arthroderma_benhamiae
Arthroderma_otae
Ascaris_lumbricoides
Ascaris_suum
Ashbya_gossypii
Aspergillus_carbonarius
Aspergillus_clavatus
Aspergillus_flavus
Aspergillus_fumigatus
Aspergillus_nidulans
Aspergillus_niger
Aspergillus_oryzae
Aspergillus_ruber
Aspergillus_terreus
Astyanax_mexicanus
Atta_cephalotes
Aureobasidium_melanogenum
Aureobasidium_pullulans
Aureobasidium_subglaciale
Aureococcus_anophagefferens
Avian_erythroblastosis_virus
Avian_musculoaponeurotic_fibrosarcoma_virus_AS42
Avian_myeloblastosis_virus
Avian_sarcoma_virus_17
Babesia_bovis
Batrachochytrium_dendrobatidis
Baudoinia_compniacensis
Beauveria_bassiana
Bigelowiella_natans
Biomphalaria_glabrata
Bipolaris_maydis
Bipolaris_oryzae
Bipolaris_sorokiniana
Bipolaris_victoriae
Bipolaris_zeicola
Blastomyces_dermatitidis
Blumeria_graminis
Boechera_stricta
Bombyx_mori
Bos_grunniens
Bos_taurus
Botryobasidium_botryosum
Botrytis_cinerea
Bovine_papillomavirus_type_2
Brachypodium_distachyon
Branchiostoma_floridae
Brassica_napus
Brassica_oleracea
Brassica_rapa
Brettanomyces_bruxellensis
Brugia_malayi
Brugia_pahangi
Brugia_timori
Buceros_rhinoceros
Bursaphelenchus_xylophilus
Byssochlamys_spectabilis
Caenorhabditis_brenneri
Caenorhabditis_briggsae
Caenorhabditis_elegans
Caenorhabditis_japonica
Caenorhabditis_remanei
Callithrix_jacchus
Calypte_anna
Camponotus_floridanus
Candida_albicans
Candida_dubliniensis
Candida_glabrata
Candida_guilliermondii
Candida_lusitaniae
Candida_maltosa
Candida_orthopsilosis
Candida_parapsilosis
Candida_tenuis
Candida_tropicalis
Canis_familiaris
Cannabis_sativa
Capitella_teleta
Capronia_coronata
Capronia_epimyces
Capronia_semiimmersa
Capsaspora_owczarzaki
Capsella_grandiflora
Capsella_rubella
Carica_papaya
Cavia_porcellus
Ceriporiopsis_subvermispora
Chaetomium_globosum
Chaetomium_thermophilum
Chelonia_mydas
Chlamydomonas_reinhardtii
Chlorella_NC64A
Chlorella_vulgaris
Chlorocebus_sabaeus
Choloepus_hoffmanni
Chroomonas_mesostigmatica
Cicer_arietinum
Cimex_lectularius
Ciona_intestinalis
Ciona_savignyi
Citrullus_lanatus
Citrus_clementina
Citrus_sinensis
Cladophialophora_bantiana
Cladophialophora_carrionii
Cladophialophora_immunda
Cladophialophora_psammophila
Cladophialophora_yegresii
Claviceps_purpurea
Clavispora_lusitaniae
Clonorchis_sinensis
Coccidioides_immitis
Coccidioides_posadasii
Coccomyxa_subellipsoidea_C169
Cochliobolus_heterostrophus_C5
Colletotrichum_fioriniae
Colletotrichum_gloeosporioides
Colletotrichum_graminicola
Colletotrichum_higginsianum
Colletotrichum_orbiculare
Colletotrichum_sublineola
Coniosporium_apollinis
Coprinellus_disseminatus
Coprinopsis_cinerea
Coprinopsis_scobicola
Cordyceps_militaris
Crassostrea_gigas
Cryphonectria_parasitica
Cryptococcus_gattii
Cryptococcus_neoformans
Cryptosporidium_hominis
Cryptosporidium_muris
Cryptosporidium_parvum
Cucumis_sativus
Culex_pipiens
Culex_quinquefasciatus
Cupiennius_salei
Cyanidioschyzon_merolae
Cylicostephanus_goldi
Cyphellophora_europaea
Dacryopinax_sp
Dactylellina_haptotyla
Danaus_plexippus
Danio_rerio
Daphnia_pulex
Dasypus_novemcinctus
Debaryomyces_hansenii
Dendroctonus_ponderosae
Dichomitus_squalens
Dictyocaulus_viviparus
Dictyostelium_discoideum
Dictyostelium_purpureum
Diphyllobothrium_latum
Dipodomys_ordii
Dirofilaria_immitis
Discocelis_tigrina
Dothistroma_septosporum
Dracunculus_medinensis
Drechslerella_stenobrocha
Drosophila_ananassae
Drosophila_erecta
Drosophila_grimshawi
Drosophila_melanogaster
Drosophila_mojavensis
Drosophila_persimilis
Drosophila_pseudoobscura
Drosophila_sechellia
Drosophila_simulans
Drosophila_virilis
Drosophila_willistoni
Drosophila_yakuba
Echinococcus_canadensis
Echinococcus_granulosus
Echinococcus_multilocularis
Echinops_telfairi
Echinostoma_caproni
Edhazardia_aedis
Eimeria_tenella
Elaeophora_elaphi
Eleutheria_dichotoma
Emiliania_huxleyi
Encephalitozoon_cuniculi
Encephalitozoon_hellem
Encephalitozoon_intestinalis
Encephalitozoon_romaleae
Endocarpon_pusillum
Entamoeba_dispar
Entamoeba_histolytica
Entamoeba_invadens
Enterobius_vermicularis
Enterocytozoon_bieneusi
EpsteinBarr_virus
Equus_caballus
Eremothecium_cymbalariae
Erinaceus_europaeus
Erysiphe_necator
Eucalyptus_grandis
Eutrema_salsugineum
Eutypa_lata
Exophiala_aquamarina
Exophiala_dermatitidis
Exophiala_mesophila
Exophiala_oligosperma
Exophiala_sideris
Exophiala_spinifera
Exophiala_xenobiotica
Fasciola_hepatica
Felis_catus
Fibroporia_radiculosa
Ficedula_albicollis
Fomitopsis_pinicola
Fonsecaea_pedrosoi
Fragaria_vesca
Fragilariopsis_cylindrus
Fusarium_fujikuroi
Fusarium_graminearum
Fusarium_oxysporum
Fusarium_pseudograminearum
Fusarium_solani
Fusarium_verticillioides
Gadus_morhua
Gaeumannomyces_graminis
Galerina_marginata
Gallid_herpesvirus_2
Gallus_gallus
Gasterosteus_aculeatus
Geospiza_fortis
Giardia_lamblia
Glarea_lozoyensis
Globodera_pallida
Gloeophyllum_trabeum
Glycine_max
Gongylonema_pulchrum
Gorilla_gorilla
Gossypium_raimondii
Grosmannia_clavigera
Guillardia_theta
Gymnopus_luxurians
Haemonchus_contortus
Haemonchus_placei
Halocynthia_roretzi
Halyomorpha_halys
Harpegnathos_saltator
Hebeloma_cylindrosporum
Helianthus_annuus
Heliconius_melpomene
Heligmosomoides_bakeri
Helobdella_robusta
Hemiselmis_andersenii
Heterobasidion_annosum
Heterobasidion_irregulare
Heterodontus_francisci
Heterorhabditis_bacteriophora
Histoplasma_capsulatum
Homo_sapiens
Hordeum_vulgare
Hyaloperonospora_arabidopsidis
Hydatigera_taeniaeformis
Hydnomerulius_pinastri
Hydra_magnipapillata
Hymenolepis_diminuta
Hymenolepis_microstoma
Hymenolepis_nana
Ictidomys_tridecemlineatus
Ipomoea_batatas
Ixodes_scapularis
Jaapia_argillacea
Kazachstania_africana
Kazachstania_naganishii
Kluyveromyces_delphensis
Kluyveromyces_lactis
Kluyveromyces_thermotolerans
Kluyveromyces_waltii
Komagataella_pastoris
Kuraishia_capsulata
Laccaria_amethystina
Laccaria_bicolor
Lachancea_kluyveri
Lachancea_lanzarotensis
Lachancea_thermotolerans
Latimeria_chalumnae
Leersia_perrieri
Leishmania_braziliensis
Leishmania_infantum
Leishmania_major
Leishmania_mexicana
Lepisosteus_oculatus
Leptinotarsa_decemlineata
Leptosphaeria_maculans
Lingula_unguis
Linum_usitatissimum
Litomosoides_sigmodontis
Loa_loa
Lodderomyces_elongisporus
Lottia_gigantea
Lotus_japonicus
Loxodonta_africana
Lucilia_cuprina
Lutzomyia_longipalpis
Lycopersicon_esculentum
Macaca_fascicularis
Macaca_mulatta
Macrophomina_phaseolina
Macropus_eugenii
Magnaporthe_oryzae
Magnaporthe_poae
Malassezia_globosa
Malassezia_sympodialis
Malus_domestica
Manihot_esculenta
Marmota_monax
Marssonina_brunnea
Medicago_sativa
Medicago_truncatula
Megaselia_scalaris
Melampsora_laricipopulina
Meleagris_gallopavo
Melitaea_cinxia
Meloidogyne_floridensis
Meloidogyne_hapla
Meloidogyne_incognita
Mesocestoides_corti
Metarhizium_acridum
Metarhizium_album
Metarhizium_anisopliae
Metarhizium_brunneum
Metarhizium_guizhouense
Metarhizium_majus
Metarhizium_robertsii
Meyerozyma_guilliermondii
Microbotryum_violaceum
Microcebus_murinus
Micromonas_pusilla
Micromonas_sp_RCC299
Microsporidia_sp
Microsporum_canis
Microsporum_gypseum
Millerozyma_farinosa
Mimulus_guttatus
Mixia_osmundae
Mnemiopsis_leidyi
Moniliophthora_perniciosa
Moniliophthora_roreri
Monodelphis_domestica
Monosiga_brevicollis
Mucor_circinelloides
Mus_musculus
Musa_acuminata
Musca_domestica
Mustela_putorius_furo
Myceliophthora_thermophila
Mycosphaerella_fijiensis
Mycosphaerella_graminicola
Myotis_brandtii
Myotis_lucifugus
Naegleria_gruberi
Nasonia_vitripennis
Naumovozyma_castellii
Naumovozyma_dairenensis
Necator_americanus
Nectria_haematococca
Nematocida_parisii_ertm1
Nematocida_sp
Nematostella_vectensis
Neofusicoccum_parvum
Neosartorya_fischeri
Neospora_caninum
Neurospora_crassa
Neurospora_discreta
Neurospora_tetrasperma
Nicotiana_sp.
Nicotiana_tabacum
Nippostrongylus_brasiliensis
Nomascus_leucogenys
Nosema_apis
Nosema_bombycis
Nosema_ceranae
Ochotona_princeps
Octopus_bimaculoides
Oesophagostomum_dentatum
Ogataea_parapolymorpha
Oidiodendron_maius
Oikopleura_dioica
Onchocerca_flexuosa
Onchocerca_ochengi
Onchocerca_volvulus
Oncorhynchus_tshawytscha
Ophiocordyceps_sinensis
Ophiostoma_piceae_uamh
Ophisaurus_gracilis
Opisthorchis_viverrini
Ordospora_colligata
Oreochromis_niloticus
Ornithorhynchus_anatinus
Oryctolagus_cuniculus
Oryza_barthii
Oryza_brachyantha
Oryza_glaberrima
Oryza_glumaepatula
Oryza_indica
Oryza_longistaminata
Oryza_meridionalis
Oryza_nivara
Oryza_punctata
Oryza_rufipogon
Oryza_sativa
Oryzias_latipes
Ostreococcus_RCC809
Ostreococcus_lucimarinus
Ostreococcus_tauri
Otolemur_garnettii
Ovis_aries
Oxytricha_trifallax
PBM_CONSTRUCTS
Pan_paniscus
Pan_troglodytes
Panicum_virgatum
Papio_anubis
Paracoccidioides_brasiliensis
Paracoccidioides_sp_lutzii
Paramecium_tetraurelia
Parascaris_equorum
Parastrongyloides_trichosuri
Patiria_miniata
Paxillus_involutus
Paxillus_rubicundulus
Pediculus_humanus
Pelodiscus_sinensis
Penicillium_chrysogenum
Penicillium_digitatum
Penicillium_expansum
Penicillium_italicum
Penicillium_marneffei
Penicillium_oxalicum
Penicillium_rubens
Penicillium_solitum
Pestalotiopsis_fici
Petromyzon_marinus
Petroselinum_crispum
Petunia_x_hybrida
Phaeodactylum_tricornutum
Phaeosphaeria_nodorum
Phanerochaete_carnosa
Phanerochaete_chrysosporium
Phaseolus_vulgaris
Phlebiopsis_gigantea
Phlebotomus_papatasi
Phoenix_dactylifera
Phycomyces_blakesleeanus
Physcomitrella_patens
Phytophthora_capsici
Phytophthora_infestans
Phytophthora_kernoviae
Phytophthora_lateralis
Phytophthora_parasitica
Phytophthora_ramorum
Phytophthora_sojae
Pichia_angusta
Pichia_kudriavzevii
Pichia_pastoris
Pichia_stipitis
Piloderma_croceum
Piriformospora_indica
Pisolithus_microcarpus
Pisolithus_tinctorius
Pisum_sativum
Plasmodium_berghei
Plasmodium_chabaudi
Plasmodium_falciparum
Plasmodium_knowlesi
Plasmodium_vivax
Plasmodium_yoelii
Pleurobrachia_pileus
Pleurotus_djamor
Pleurotus_ostreatus
Plicaturopsis_crispa
Pneumocystis_carinii
Pneumocystis_murina
Podocoryne_carnea
Podospora_anserina
Poecilia_formosa
Pogona_vitticeps
Polysphondylium_pallidum
Pongo_abelii
Populus_trichocarpa
Postia_placenta
Pristionchus_exspectatus
Pristionchus_pacificus
Procavia_capensis
Protopolystoma_xenopodis
Prunus_mume
Prunus_persica
Pseudocercospora_fijiensis
Pseudogymnoascus_destructans
Pseudogymnoascus_pannorum
Pseudozyma_antarctica
Pseudozyma_aphidis
Pseudozyma_brasiliensis
Pseudozyma_flocculosa
Pseudozyma_hubeiensis
Pteropus_vampyrus
Puccinia_graminis
Puccinia_triticina
Punctularia_strigosozonata
Pyrenophora_teres
Pyrenophora_triticirepentis
Pythium_aphanidermatum
Pythium_arrhenomanes
Pythium_irregulare
Pythium_iwayamai
Pythium_ultimum
Pythium_vexans
Rattus_norvegicus
Rhabditophanes_kr3021
Rhinocladiella_mackenziei
Rhizoctonia_solani
Rhizophagus_irregularis
Rhizopus_oryzae
Rhodnius_prolixus
Rhodosporidium_toruloides
Rhodotorula_glutinis
Ricinus_communis
Romanomermis_culicivorax
Rozella_allomycis
Saccharomyces_arboricola
Saccharomyces_bayanus
Saccharomyces_castellii
Saccharomyces_cerevisiae
Saccharomyces_kudriavzevii
Saccharomyces_mikatae
Saccharomyces_paradoxus
Saccharomycetaceae_sp_ashbya_aceri
Saccoglossus_kowalevskii
Salix_purpurea
Salpingoeca_rosetta
Saprolegnia_parasitica
Sarcophilus_harrisii
Sarsia_sp._Long_Island_Sound
Scedosporium_apiospermum
Scheffersomyces_stipitis
Schistocephalus_solidus
Schistosoma_curassoni
Schistosoma_haematobium
Schistosoma_japonicum
Schistosoma_mansoni
Schistosoma_margrebowiei
Schistosoma_mattheei
Schistosoma_rodhaini
Schizophyllum_commune
Schizosaccharomyces_cryophilus
Schizosaccharomyces_japonicus
Schizosaccharomyces_octosporus
Schizosaccharomyces_pombe
Schmidtea_mediterranea
Scleroderma_citrinum
Sclerotinia_borealis
Sclerotinia_sclerotiorum
Selaginella_moellendorffii
Serendipita_vermifera
Serpula_lacrymans
Setaria_italica
Setosphaeria_turcica
Soboliphyme_baturini
Solanum_lycopersicum
Solanum_tuberosum
Solenopsis_invicta
Sordaria_macrospora
Sorex_araneus
Sorghum_bicolor
Spadella_cephaloptera
Spathaspora_passalidarum
Sphaerobolus_stellatus
Sphaerulina_musiva
Spirometra_erinaceieuropaei
Spizellomyces_punctatus
Sporisorium_reilianum
Sporobolomyces_roseus
Sporothrix_brasiliensis
Sporothrix_schenckii
Spraguea_lophii
Stachybotrys_chartarum
Stachybotrys_chlorohalonata
Stagonospora_nodorum
Steinernema_carpocapsae
Steinernema_feltiae
Steinernema_glaseri
Steinernema_monticolum
Steinernema_scapterisci
Strigamia_maritima
Strongylocentrotus_purpuratus
Strongyloides_papillosus
Strongyloides_stercoralis
Strongyloides_venezuelensis
Strongylus_vulgaris
Suillus_luteus
Sus_scrofa
Syphacia_muris
Taenia_asiatica
Taenia_solium
Taeniopygia_guttata
Takifugu_rubripes
Talaromyces_marneffei
Talaromyces_stipitatus
Tarsius_syrichta
Teladorsagia_circumcincta
Tetrahymena_thermophila
Tetranychus_urticae
Tetraodon_nigroviridis
Tetrapisispora_blattae
Tetrapisispora_phaffii
Thalassiosira_pseudonana
Thecamonas_trahens
Theileria_annulata
Theileria_parva
Thelazia_callipaeda
Theobroma_cacao
Thielavia_terrestris
Tilletiaria_anomala
Togninia_minima
Torrubiella_hemipterigena
Torulaspora_delbrueckii
Toxocara_canis
Toxoplasma_gondii
Trachipleistophora_hominis
Trametes_cinnabarina
Tremella_mesenterica
Tribolium_castaneum
Trichinella_nativa
Trichinella_spiralis
Trichobilharzia_regenti
Trichoderma_atroviride
Trichoderma_reesei
Trichoderma_virens
Trichomonas_vaginalis
Trichophyton_equinum
Trichophyton_interdigitale
Trichophyton_rubrum
Trichophyton_soudanense
Trichophyton_tonsurans
Trichophyton_verrucosum
Trichoplax_adhaerens
Trichosporon_asahii
Trichuris_muris
Trichuris_suis
Trichuris_trichiura
Trionyx_sinensis
Tripedalia_cystophora
Triticum_aestivum
Triticum_urartu
Trypanosoma_brucei
Trypanosoma_congolense
Trypanosoma_cruzi
Trypanosoma_vivax
Tuber_melanosporum
Tulasnella_calospora
Tupaia_belangeri
Tursiops_truncatus
Tyto_alba
Uncinocarpus_reesii
Ustilaginoidea_virens
Ustilago_hordei
Ustilago_maydis
Vanderwaltozyma_polyspora
Vavraia_culicis
Verruconis_gallopava
Verticillium_albo_atrum
Verticillium_alfalfae
Verticillium_dahliae
Vicugna_pacos
Vitis_vinifera
Vittaforma_corneae
Volvox_carteri
Wallemia_ichthyophaga
Wallemia_sebi
Wickerhamomyces_ciferrii
Wuchereria_bancrofti
Xenopus_laevis
Xenopus_tropicalis
Xiphophorus_maculatus
Yarrowia_lipolytica
Zea_mays
Zootermopsis_nevadensis
Zygosaccharomyces_bailii
Zygosaccharomyces_rouxii
Zymoseptoria_tritici
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Pick up motif information for one species and save as gimmemotif pfm file format.</span>

<span class="n">species</span> <span class="o">=</span> <span class="s2">&quot;Danio_rerio&quot;</span>
<span class="n">make_motif_file_from_cisbp_data</span><span class="p">(</span><span class="n">pwm_folder_path</span><span class="o">=</span><span class="s2">&quot;pwms&quot;</span><span class="p">,</span> <span class="n">tfinfo_df</span><span class="o">=</span><span class="n">df</span><span class="p">,</span> <span class="n">species</span><span class="o">=</span><span class="n">species</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(6133,)
(109560, 4) 5298
</pre></div></div>
</div>
</div>
<div class="section" id="4.-Check-results">
<h1>4. Check results<a class="headerlink" href="#4.-Check-results" title="Permalink to this headline">Â¶</a></h1>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Read motifs</span>

<span class="kn">from</span> <span class="nn">gimmemotifs.motif</span> <span class="kn">import</span> <span class="n">read_motifs</span>

<span class="n">path</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;CisDB_ver2_</span><span class="si">{</span><span class="n">species</span><span class="si">}</span><span class="s2">.pfm&quot;</span>
<span class="n">custom_motifs</span> <span class="o">=</span> <span class="n">read_motifs</span><span class="p">(</span><span class="n">path</span><span class="p">)</span>
<span class="n">custom_motifs</span><span class="p">[:</span><span class="mi">10</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[M00008_2.00_nnnAAww,
 M00045_2.00_GTAAACAA,
 M00056_2.00_TAATAAAT,
 M00066_2.00_nsGTTGCyAn,
 M00070_2.00_nrAACAATAnn,
 M00111_2.00_nGCCynnGGs,
 M00112_2.00_CCTsrGGCnA,
 M00113_2.00_nsCCnnAGGs,
 M00114_2.00_nnGCCynnGG,
 M00115_2.00_nnATnAAAn]
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Delete downloaded data</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="o">!</span> rm -r TF_Information_all_motifs*
<span class="o">!</span> rm PWMs.*
<span class="o">!</span> rm -r pwms

</pre></div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../../../tutorials/scrnaprocess.html" class="btn btn-neutral float-right" title="3. Single-cell RNA-seq data preprocessing" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="01_How_to_load_motif_data.html" class="btn btn-neutral float-left" title="0. Overview: How to use a different motifs for celloracle motif scan." accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, Samantha Morris Lab

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>