

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>0. Import libraries &mdash; celloracle 0.5.1 documentation</title>
  

  
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="0. Overview: How to use a different motifs for celloracle motif scan." href="motif_data_preparation/01_How_to_load_motif_data.html" />
    <link rel="prev" title="2. Transcription factor binding motif scan" href="../../tutorials/motifscan.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../index.html" class="icon icon-home" alt="Documentation Home"> celloracle
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../index.html">Welcome to celloracle’s documentation!</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../installation/index.html">Installation</a></li>
</ul>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="../../tutorials/index.html">Tutorial</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../../tutorials/atac.html">1. ATAC-seq data preprocessing</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../../tutorials/motifscan.html">2. Transcription factor binding motif scan</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="../../tutorials/motifscan.html#scan-dna-sequences-searching-for-tf-binding-motifs">Scan DNA sequences searching for TF binding motifs</a><ul class="current">
<li class="toctree-l4 current"><a class="current reference internal" href="#">0. Import libraries</a></li>
<li class="toctree-l4"><a class="reference internal" href="#1.-Rerefence-genome-data-preparation">1. Rerefence genome data preparation</a></li>
<li class="toctree-l4"><a class="reference internal" href="#2.-Load-data">2. Load data</a></li>
<li class="toctree-l4"><a class="reference internal" href="#3.-Instantiate-TFinfo-object-and-search-for-TF-binding-motifs">3. Instantiate TFinfo object and search for TF binding motifs</a></li>
<li class="toctree-l4"><a class="reference internal" href="#4.-Filtering-motifs">4. Filtering motifs</a></li>
<li class="toctree-l4"><a class="reference internal" href="#5.-Get-Final-results">5. Get Final results</a></li>
<li class="toctree-l4"><a class="reference internal" href="#6.-Save-TFinfo-as-dictionary-or-dataframe">6. Save TFinfo as dictionary or dataframe</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../tutorials/motifscan.html#optional-step-1-how-to-use-different-motif-data">[Optional step 1] How to use different motif data</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../tutorials/motifscan.html#optional-step-2-how-to-make-custom-motifs-for-celloracle-motif-analysis">[Optional step 2] How to Make custom motifs for celloracle motif analysis</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../tutorials/scrnaprocess.html">3. Single-cell RNA-seq data preprocessing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../tutorials/networkanalysis.html">4. Network analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../tutorials/simulation.html">5. Simulation with GRNs</a></li>
</ul>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../modules/index.html">API</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../changelog/index.html">Changelog</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../license/index.html">License</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../citation/index.html">Authors and citations</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">celloracle</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../../tutorials/index.html">Tutorial</a> &raquo;</li>
        
          <li><a href="../../tutorials/motifscan.html"><span class="section-number">2. </span>Transcription factor binding motif scan</a> &raquo;</li>
        
      <li>0. Import libraries</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../../_sources/notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_20200801.ipynb.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container,
div.nbinput.container div.prompt,
div.nbinput.container div.input_area,
div.nbinput.container div[class*=highlight],
div.nbinput.container div[class*=highlight] pre,
div.nboutput.container,
div.nboutput.container div.prompt,
div.nboutput.container div.output_area,
div.nboutput.container div[class*=highlight],
div.nboutput.container div[class*=highlight] pre {
    background: none;
    border: none;
    padding: 0 0;
    margin: 0;
    box-shadow: none;
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    background: #f5f5f5;
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt a.copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="0.-Import-libraries">
<h1>0. Import libraries<a class="headerlink" href="#0.-Import-libraries" title="Permalink to this headline">¶</a></h1>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>


<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>

<span class="kn">import</span> <span class="nn">os</span><span class="o">,</span> <span class="nn">sys</span><span class="o">,</span> <span class="nn">shutil</span><span class="o">,</span> <span class="nn">importlib</span><span class="o">,</span> <span class="nn">glob</span>
<span class="kn">from</span> <span class="nn">tqdm.notebook</span> <span class="kn">import</span> <span class="n">tqdm</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">celloracle</span> <span class="kn">import</span> <span class="n">motif_analysis</span> <span class="k">as</span> <span class="n">ma</span>
<span class="kn">from</span> <span class="nn">celloracle.utility</span> <span class="kn">import</span> <span class="n">save_as_pickled_object</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="o">%</span><span class="k">config</span> InlineBackend.figure_format = &#39;retina&#39;
<span class="o">%</span><span class="k">matplotlib</span> inline

<span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;figure.figsize&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="mi">15</span><span class="p">,</span><span class="mi">7</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s2">&quot;savefig.dpi&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">600</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="1.-Rerefence-genome-data-preparation">
<h1>1. Rerefence genome data preparation<a class="headerlink" href="#1.-Rerefence-genome-data-preparation" title="Permalink to this headline">¶</a></h1>
<div class="section" id="1.1.-Check-reference-genome-installation">
<h2>1.1. Check reference genome installation<a class="headerlink" href="#1.1.-Check-reference-genome-installation" title="Permalink to this headline">¶</a></h2>
<p>Celloracle uses genomepy to get DNA sequence data. Before starting celloracle analysis, we need to make sure that the reference genome is correctly installed in your computational environment. If not, please install reference genome.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># PLEASE make sure that you are setting correct ref genome.</span>
<span class="n">ref_genome</span> <span class="o">=</span> <span class="s2">&quot;mm10&quot;</span>

<span class="n">genome_installation</span> <span class="o">=</span> <span class="n">ma</span><span class="o">.</span><span class="n">is_genome_installed</span><span class="p">(</span><span class="n">ref_genome</span><span class="o">=</span><span class="n">ref_genome</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">ref_genome</span><span class="p">,</span> <span class="s2">&quot;installation: &quot;</span><span class="p">,</span> <span class="n">genome_installation</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
mm10 installation:  True
</pre></div></div>
</div>
</div>
<div class="section" id="1.2.-Install-reference-genome-(if-refgenome-is-not-installed)">
<h2>1.2. Install reference genome (if refgenome is not installed)<a class="headerlink" href="#1.2.-Install-reference-genome-(if-refgenome-is-not-installed)" title="Permalink to this headline">¶</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">if</span> <span class="ow">not</span> <span class="n">genome_installation</span><span class="p">:</span>
    <span class="kn">import</span> <span class="nn">genomepy</span>
    <span class="n">genomepy</span><span class="o">.</span><span class="n">install_genome</span><span class="p">(</span><span class="n">ref_genome</span><span class="p">,</span> <span class="s2">&quot;UCSC&quot;</span><span class="p">)</span>
<span class="k">else</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">ref_genome</span><span class="p">,</span> <span class="s2">&quot;is installed.&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
mm10 is installed.
</pre></div></div>
</div>
</div>
</div>
<div class="section" id="2.-Load-data">
<h1>2. Load data<a class="headerlink" href="#2.-Load-data" title="Permalink to this headline">¶</a></h1>
<div class="section" id="2.1.-Load-processed-peak-data">
<h2>2.1. Load processed peak data<a class="headerlink" href="#2.1.-Load-processed-peak-data" title="Permalink to this headline">¶</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Load annotated peak data.</span>
<span class="n">peaks</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_parquet</span><span class="p">(</span><span class="s2">&quot;../01_ATAC-seq_data_processing/option1_scATAC-seq_data_analysis_with_cicero/peak_file.parquet&quot;</span><span class="p">)</span>
<span class="n">peaks</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>peak_id</th>
      <th>gene_short_name</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr10_100015291_100017830</td>
      <td>Kitl</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr10_100486534_100488209</td>
      <td>Tmtc3</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr10_100588641_100589556</td>
      <td>4930430F08Rik</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr10_100741247_100742505</td>
      <td>Gm35722</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr10_101681379_101682124</td>
      <td>Mgat4c</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
</div>
<div class="section" id="2.1.-Check-data">
<h2>2.1. Check data<a class="headerlink" href="#2.1.-Check-data" title="Permalink to this headline">¶</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Define function for quality check</span>
<span class="k">def</span> <span class="nf">decompose_chrstr</span><span class="p">(</span><span class="n">peak_str</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Args:</span>
<span class="sd">        peak_str (str): peak_str. e.g. &#39;chr1_3094484_3095479&#39;</span>

<span class="sd">    Returns:</span>
<span class="sd">        tuple: chromosome name, start position, end position</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="o">*</span><span class="n">chr_</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="n">peak_str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span>
    <span class="n">chr_</span> <span class="o">=</span> <span class="s2">&quot;_&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">chr_</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">chr_</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span>

<span class="kn">from</span> <span class="nn">genomepy</span> <span class="kn">import</span> <span class="n">Genome</span>

<span class="k">def</span> <span class="nf">check_peak_foamat</span><span class="p">(</span><span class="n">peaks_df</span><span class="p">,</span> <span class="n">ref_genome</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Check peak fomat.</span>
<span class="sd">     (1) Check chromosome name.</span>
<span class="sd">     (2) Check peak size (length) and remove sort DNAs (&lt;5bp)</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">df</span> <span class="o">=</span> <span class="n">peaks_df</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

    <span class="n">n_peaks_before</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

    <span class="c1"># Decompose peaks and make df</span>
    <span class="n">decomposed</span> <span class="o">=</span> <span class="p">[</span><span class="n">decompose_chrstr</span><span class="p">(</span><span class="n">peak_str</span><span class="p">)</span> <span class="k">for</span> <span class="n">peak_str</span> <span class="ow">in</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;peak_id&quot;</span><span class="p">]]</span>
    <span class="n">df_decomposed</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">decomposed</span><span class="p">))</span>
    <span class="n">df_decomposed</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span> <span class="s2">&quot;start&quot;</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">]</span>
    <span class="n">df_decomposed</span><span class="p">[</span><span class="s2">&quot;start&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_decomposed</span><span class="p">[</span><span class="s2">&quot;start&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">int</span><span class="p">)</span>
    <span class="n">df_decomposed</span><span class="p">[</span><span class="s2">&quot;end&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_decomposed</span><span class="p">[</span><span class="s2">&quot;end&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">int</span><span class="p">)</span>

    <span class="c1"># Load genome data</span>
    <span class="n">genome_data</span> <span class="o">=</span> <span class="n">Genome</span><span class="p">(</span><span class="n">ref_genome</span><span class="p">)</span>
    <span class="n">all_chr_list</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">genome_data</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>


    <span class="c1"># DNA length check</span>
    <span class="n">lengths</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">df_decomposed</span><span class="p">[</span><span class="s2">&quot;end&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">df_decomposed</span><span class="p">[</span><span class="s2">&quot;start&quot;</span><span class="p">])</span>


    <span class="c1"># Filter peaks with invalid chromosome name</span>
    <span class="n">n_threshold</span> <span class="o">=</span> <span class="mi">5</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[(</span><span class="n">lengths</span> <span class="o">&gt;=</span> <span class="n">n_threshold</span><span class="p">)</span> <span class="o">&amp;</span> <span class="n">df_decomposed</span><span class="o">.</span><span class="n">chr</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">all_chr_list</span><span class="p">)]</span>

    <span class="c1"># DNA length check</span>
    <span class="n">lengths</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">df_decomposed</span><span class="p">[</span><span class="s2">&quot;end&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">df_decomposed</span><span class="p">[</span><span class="s2">&quot;start&quot;</span><span class="p">])</span>

    <span class="c1"># Data counting</span>
    <span class="n">n_invalid_length</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">lengths</span><span class="p">[</span><span class="n">lengths</span> <span class="o">&lt;</span> <span class="n">n_threshold</span><span class="p">])</span>
    <span class="n">n_peaks_invalid_chr</span> <span class="o">=</span> <span class="n">n_peaks_before</span> <span class="o">-</span> <span class="n">df_decomposed</span><span class="o">.</span><span class="n">chr</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">all_chr_list</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
    <span class="n">n_peaks_after</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

    <span class="c1">#</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Peaks before filtering: &quot;</span><span class="p">,</span> <span class="n">n_peaks_before</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Peaks with invalid chr_name: &quot;</span><span class="p">,</span> <span class="n">n_peaks_invalid_chr</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Peaks with invalid length: &quot;</span><span class="p">,</span> <span class="n">n_invalid_length</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Peaks after filtering: &quot;</span><span class="p">,</span> <span class="n">n_peaks_after</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">df</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">peaks</span> <span class="o">=</span> <span class="n">check_peak_foamat</span><span class="p">(</span><span class="n">peaks</span><span class="p">,</span> <span class="n">ref_genome</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Peaks before filtering:  15680
Peaks with invalid chr_name:  0
Peaks with invalid length:  0
Peaks after filtering:  15680
</pre></div></div>
</div>
</div>
<div class="section" id="2.2.-[Optional-step]-Load-motifs">
<h2>2.2. [Optional step] Load motifs<a class="headerlink" href="#2.2.-[Optional-step]-Load-motifs" title="Permalink to this headline">¶</a></h2>
<p>You can select TF binding motif data for Celloracle motif analysis. If you have no preference and just want to use a default motif, you can skip this step. If you want to use a non-default motif dataset, please prepare motif data as a list of motif class in gimmemotifs. We have several option for loading motif database as below.</p>
<div class="section" id="2.2.1.-[Optional-step]-Load-motif-data-from-gimmemotifs-dataset">
<h3>2.2.1. [Optional step] Load motif data from gimmemotifs dataset<a class="headerlink" href="#2.2.1.-[Optional-step]-Load-motif-data-from-gimmemotifs-dataset" title="Permalink to this headline">¶</a></h3>
<p>Many motif databases are included with GimmeMotifs. <a class="reference external" href="https://gimmemotifs.readthedocs.io/en/master/overview.html">https://gimmemotifs.readthedocs.io/en/master/overview.html</a> You can load them as follows.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># First, we need to pick up motifs for your dataset. We can get a list.</span>

<span class="c1"># Get folder path that stores motif data.</span>
<span class="kn">import</span> <span class="nn">os</span><span class="o">,</span> <span class="nn">glob</span>
<span class="kn">from</span> <span class="nn">gimmemotifs.motif</span> <span class="kn">import</span> <span class="n">MotifConfig</span>
<span class="n">config</span> <span class="o">=</span> <span class="n">MotifConfig</span><span class="p">()</span>
<span class="n">motif_dir</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">get_motif_dir</span><span class="p">()</span>

<span class="c1"># Get list for motif data name</span>
<span class="n">motifs_data_name</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">motif_dir</span><span class="p">)</span> <span class="k">if</span> <span class="n">i</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.pfm&quot;</span><span class="p">)]</span>
<span class="n">motifs_data_name</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
<span class="n">motifs_data_name</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[&#39;CIS-BP.pfm&#39;,
 &#39;ENCODE.pfm&#39;,
 &#39;HOCOMOCOv10_HUMAN.pfm&#39;,
 &#39;HOCOMOCOv10_MOUSE.pfm&#39;,
 &#39;HOCOMOCOv11_HUMAN.pfm&#39;,
 &#39;HOCOMOCOv11_MOUSE.pfm&#39;,
 &#39;HOMER.pfm&#39;,
 &#39;IMAGE.pfm&#39;,
 &#39;JASPAR2018.pfm&#39;,
 &#39;JASPAR2018_fungi.pfm&#39;,
 &#39;JASPAR2018_insects.pfm&#39;,
 &#39;JASPAR2018_nematodes.pfm&#39;,
 &#39;JASPAR2018_plants.pfm&#39;,
 &#39;JASPAR2018_urochordates.pfm&#39;,
 &#39;JASPAR2018_vertebrates.pfm&#39;,
 &#39;JASPAR2020.pfm&#39;,
 &#39;JASPAR2020_fungi.pfm&#39;,
 &#39;JASPAR2020_insects.pfm&#39;,
 &#39;JASPAR2020_nematodes.pfm&#39;,
 &#39;JASPAR2020_plants.pfm&#39;,
 &#39;JASPAR2020_urochordates.pfm&#39;,
 &#39;JASPAR2020_vertebrates.pfm&#39;,
 &#39;RSAT_insects.pfm&#39;,
 &#39;RSAT_plants.pfm&#39;,
 &#39;RSAT_vertebrates.pfm&#39;,
 &#39;SwissRegulon.pfm&#39;,
 &#39;factorbook.pfm&#39;,
 &#39;gimme.vertebrate.v5.0.pfm&#39;]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Once you picked up motifs, you can load the motif files with &quot;read_motifs&quot;</span>
<span class="kn">from</span> <span class="nn">gimmemotifs.motif</span> <span class="kn">import</span> <span class="n">read_motifs</span>

<span class="n">path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">motif_dir</span><span class="p">,</span> <span class="s2">&quot;JASPAR2018_plants.pfm&quot;</span><span class="p">)</span><span class="c1"># Please enter motifs name here</span>
<span class="n">motifs</span> <span class="o">=</span> <span class="n">read_motifs</span><span class="p">(</span><span class="n">path</span><span class="p">)</span>

<span class="c1"># Check first 10 motifs</span>
<span class="n">motifs</span><span class="p">[:</span><span class="mi">10</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[MA0020.1_Dof2_AAAGCn,
 MA0021.1_Dof3_AAAGyn,
 MA0034.1_Gam1_nnyAACCGmC,
 MA0044.1_HMG-1_sTTGTnyTy,
 MA0045.1_HMG-I/Y_nwAnAAAnrnmrAmAy,
 MA0053.1_MNB1A_AAAGC,
 MA0054.1_myb.Ph3_TAACnGTTw,
 MA0064.1_PBF_AAAGy,
 MA0082.1_squamosa_mCAwAwATrGwAAn,
 MA0096.1_bZIP910_mTGACGT]
</pre></div></div>
</div>
</div>
<div class="section" id="2.2.2-[Optional-step]-Load-motif-data-from-celloracle-motif-dataset">
<h3>2.2.2 [Optional step] Load motif data from celloracle motif dataset<a class="headerlink" href="#2.2.2-[Optional-step]-Load-motif-data-from-celloracle-motif-dataset" title="Permalink to this headline">¶</a></h3>
<p>Celloracle also provides many motif dataset that was generated from CisBP. <a class="reference external" href="http://cisbp.ccbr.utoronto.ca/">http://cisbp.ccbr.utoronto.ca/</a></p>
<p>These motifs were organized by each species. Please select motifs for your species.</p>
<p>If you have a request for motifs for a new species, you can ask us to add new motifs through github issue page.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Check available motifs</span>
<span class="n">ma</span><span class="o">.</span><span class="n">MOTIFS_LIST</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[&#39;CisBP_ver2_Arabidopsis_thaliana.pfm&#39;,
 &#39;CisBP_ver2_Caenorhabditis_elegans.pfm&#39;,
 &#39;CisBP_ver2_Danio_rerio.pfm&#39;,
 &#39;CisBP_ver2_Drosophila_ananassae.pfm&#39;,
 &#39;CisBP_ver2_Drosophila_erecta.pfm&#39;,
 &#39;CisBP_ver2_Drosophila_grimshawi.pfm&#39;,
 &#39;CisBP_ver2_Drosophila_melanogaster.pfm&#39;,
 &#39;CisBP_ver2_Drosophila_mix.pfm&#39;,
 &#39;CisBP_ver2_Drosophila_mojavensis.pfm&#39;,
 &#39;CisBP_ver2_Drosophila_persimilis.pfm&#39;,
 &#39;CisBP_ver2_Drosophila_pseudoobscura.pfm&#39;,
 &#39;CisBP_ver2_Drosophila_sechellia.pfm&#39;,
 &#39;CisBP_ver2_Drosophila_simulans.pfm&#39;,
 &#39;CisBP_ver2_Drosophila_virilis.pfm&#39;,
 &#39;CisBP_ver2_Drosophila_willistoni.pfm&#39;,
 &#39;CisBP_ver2_Drosophila_yakuba.pfm&#39;,
 &#39;CisBP_ver2_Homo_sapiens.pfm&#39;,
 &#39;CisBP_ver2_Mus_musculus.pfm&#39;,
 &#39;CisBP_ver2_Rattus_norvegicus.pfm&#39;,
 &#39;CisBP_ver2_Saccharomyces_cerevisiae.pfm&#39;,
 &#39;CisBP_ver2_Xenopus_laevis.pfm&#39;,
 &#39;CisBP_ver2_Xenopus_tropicalis.pfm&#39;,
 &#39;CisBP_ver2_Xenopus_tropicalis_and_Xenopus_laevis.pfm&#39;]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Load motifs from celloracle dataset.</span>
<span class="n">motifs</span> <span class="o">=</span> <span class="n">ma</span><span class="o">.</span><span class="n">load_motifs</span><span class="p">(</span><span class="s2">&quot;CisBP_ver2_Mus_musculus.pfm&quot;</span><span class="p">)</span>

<span class="c1"># Check first 10 motifs</span>
<span class="n">motifs</span><span class="p">[:</span><span class="mi">10</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[M00008_2.00_nnnAAww,
 M00044_2.00_nrTAAACAn,
 M00056_2.00_TAATAAAT,
 M00060_2.00_nnnTTCnnn,
 M00111_2.00_nGCCynnGGs,
 M00112_2.00_CCTsrGGCnA,
 M00113_2.00_nsCCnnAGGs,
 M00114_2.00_nnGCCynnGG,
 M00115_2.00_nnATnAAAn,
 M00116_2.00_nnAATATTAnn]
</pre></div></div>
</div>
</div>
</div>
</div>
<div class="section" id="3.-Instantiate-TFinfo-object-and-search-for-TF-binding-motifs">
<h1>3. Instantiate TFinfo object and search for TF binding motifs<a class="headerlink" href="#3.-Instantiate-TFinfo-object-and-search-for-TF-binding-motifs" title="Permalink to this headline">¶</a></h1>
<p>The motif analysis module has a custom class; TFinfo. The TFinfo object converts a peak data into a DNA sequences and scans the DNA sequences searching for TF binding motifs. Then, the results of motif scan will be filtered and converted into either a python dictionary or a depending on your preference. This TF information is necessary for GRN inference.</p>
<div class="section" id="3.1.-Instantiate-TFinfo-object">
<h2>3.1. Instantiate TFinfo object<a class="headerlink" href="#3.1.-Instantiate-TFinfo-object" title="Permalink to this headline">¶</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Instantiate TFinfo object</span>
<span class="n">tfi</span> <span class="o">=</span> <span class="n">ma</span><span class="o">.</span><span class="n">TFinfo</span><span class="p">(</span><span class="n">peak_data_frame</span><span class="o">=</span><span class="n">peaks</span><span class="p">,</span> <span class="c1"># peak info calculated from ATAC-seq data</span>
                       <span class="n">ref_genome</span><span class="o">=</span><span class="n">ref_genome</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="3.2.-Motif-scan">
<h2>3.2. Motif scan<a class="headerlink" href="#3.2.-Motif-scan" title="Permalink to this headline">¶</a></h2>
<p>!!You can set TF binding motif information as an argument:</p>
<p>tfi.scan(motifs=motifs)</p>
<p>If you don’t set motifs or set None, default motifs will be loaded automatically.</p>
<ul class="simple">
<li><p>For mouse and human, “gimme.vertebrate.v5.0.” will be used as a default motifs.</p></li>
<li><p>For another species, a species specific TF binding motif data extracted from CisBP ver2.0 will be used.</p></li>
</ul>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="o">%%time</span>
<span class="c1"># Scan motifs. !!CAUTION!! This step may take several hours if you have many peaks!</span>
<span class="n">tfi</span><span class="o">.</span><span class="n">scan</span><span class="p">(</span><span class="n">fpr</span><span class="o">=</span><span class="mf">0.02</span><span class="p">,</span>
            <span class="n">motifs</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>  <span class="c1"># If you enter None, default motifs will be loaded.</span>
            <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># Save tfinfo object</span>
<span class="n">tfi</span><span class="o">.</span><span class="n">to_hdf5</span><span class="p">(</span><span class="n">file_path</span><span class="o">=</span><span class="s2">&quot;test1.celloracle.tfinfo&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Check motif scan results</span>
<span class="n">tfi</span><span class="o">.</span><span class="n">scanned_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>seqname</th>
      <th>motif_id</th>
      <th>factors_direct</th>
      <th>factors_indirect</th>
      <th>score</th>
      <th>pos</th>
      <th>strand</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>0</td>
      <td>chr10_100015291_100017830</td>
      <td>GM.5.0.Homeodomain.0001</td>
      <td>TGIF1</td>
      <td>ENSG00000234254, TGIF1</td>
      <td>10.311002</td>
      <td>1003</td>
      <td>1</td>
    </tr>
    <tr>
      <td>1</td>
      <td>chr10_100015291_100017830</td>
      <td>GM.5.0.Mixed.0001</td>
      <td></td>
      <td>SRF, EGR1</td>
      <td>7.925873</td>
      <td>481</td>
      <td>1</td>
    </tr>
    <tr>
      <td>2</td>
      <td>chr10_100015291_100017830</td>
      <td>GM.5.0.Mixed.0001</td>
      <td></td>
      <td>SRF, EGR1</td>
      <td>7.321375</td>
      <td>911</td>
      <td>-1</td>
    </tr>
    <tr>
      <td>3</td>
      <td>chr10_100015291_100017830</td>
      <td>GM.5.0.Mixed.0001</td>
      <td></td>
      <td>SRF, EGR1</td>
      <td>7.276585</td>
      <td>811</td>
      <td>-1</td>
    </tr>
    <tr>
      <td>4</td>
      <td>chr10_100015291_100017830</td>
      <td>GM.5.0.Nuclear_receptor.0002</td>
      <td>NR2C2</td>
      <td>NR2C2, Nr2c2</td>
      <td>9.067331</td>
      <td>449</td>
      <td>-1</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>We have the score for each sequence and motif_id pair. In the next step we will filter the motifs with low score.</p>
</div>
</div>
<div class="section" id="4.-Filtering-motifs">
<h1>4. Filtering motifs<a class="headerlink" href="#4.-Filtering-motifs" title="Permalink to this headline">¶</a></h1>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Reset filtering</span>
<span class="n">tfi</span><span class="o">.</span><span class="n">reset_filtering</span><span class="p">()</span>

<span class="c1"># Do filtering</span>
<span class="n">tfi</span><span class="o">.</span><span class="n">filter_motifs_by_score</span><span class="p">(</span><span class="n">threshold</span><span class="o">=</span><span class="mf">10.5</span><span class="p">)</span>

<span class="c1"># Do post filtering process. Convert results into several file format.</span>
<span class="n">tfi</span><span class="o">.</span><span class="n">make_TFinfo_dataframe_and_dictionary</span><span class="p">(</span><span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>


</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
peaks were filtered: 12952283 -&gt; 2288874
1. converting scanned results into one-hot encoded dataframe.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "d4f5cfb59ff04daabcd23a5bf12e557c", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

2. converting results into dictionaries.
converting scan results into dictionaries...
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "81d283cff9864e8da048554a7d8afdfd", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "b496e9f51545423c9f72100f22ba2ab7", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

</pre></div></div>
</div>
</div>
<div class="section" id="5.-Get-Final-results">
<h1>5. Get Final results<a class="headerlink" href="#5.-Get-Final-results" title="Permalink to this headline">¶</a></h1>
<div class="section" id="5.1.-Get-resutls-as-a-dictionary">
<h2>5.1. Get resutls as a dictionary<a class="headerlink" href="#5.1.-Get-resutls-as-a-dictionary" title="Permalink to this headline">¶</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">td</span> <span class="o">=</span> <span class="n">tfi</span><span class="o">.</span><span class="n">to_dictionary</span><span class="p">(</span><span class="n">dictionary_type</span><span class="o">=</span><span class="s2">&quot;targetgene2TFs&quot;</span><span class="p">)</span>

</pre></div>
</div>
</div>
</div>
<div class="section" id="5.2.-Get-results-as-a-dataframe">
<h2>5.2. Get results as a dataframe<a class="headerlink" href="#5.2.-Get-results-as-a-dataframe" title="Permalink to this headline">¶</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">df</span> <span class="o">=</span> <span class="n">tfi</span><span class="o">.</span><span class="n">to_dataframe</span><span class="p">()</span>
<span class="n">df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>peak_id</th>
      <th>gene_short_name</th>
      <th>9430076c15rik</th>
      <th>Ac002126.6</th>
      <th>Ac012531.1</th>
      <th>Ac226150.2</th>
      <th>Afp</th>
      <th>Ahr</th>
      <th>Ahrr</th>
      <th>Aire</th>
      <th>...</th>
      <th>Znf784</th>
      <th>Znf8</th>
      <th>Znf816</th>
      <th>Znf85</th>
      <th>Zscan10</th>
      <th>Zscan16</th>
      <th>Zscan22</th>
      <th>Zscan26</th>
      <th>Zscan31</th>
      <th>Zscan4</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr10_100015291_100017830</td>
      <td>Kitl</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr10_100486534_100488209</td>
      <td>Tmtc3</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>...</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr10_100588641_100589556</td>
      <td>4930430F08Rik</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr10_100741247_100742505</td>
      <td>Gm35722</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr10_101681379_101682124</td>
      <td>Mgat4c</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 1092 columns</p>
</div></div>
</div>
</div>
</div>
<div class="section" id="6.-Save-TFinfo-as-dictionary-or-dataframe">
<h1>6. Save TFinfo as dictionary or dataframe<a class="headerlink" href="#6.-Save-TFinfo-as-dictionary-or-dataframe" title="Permalink to this headline">¶</a></h1>
<p>We’ll use this information when making the GRNs. Save the results.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">folder</span> <span class="o">=</span> <span class="s2">&quot;TFinfo_outputs&quot;</span>
<span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">folder</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># save TFinfo as a dictionary</span>
<span class="n">td</span> <span class="o">=</span> <span class="n">tfi</span><span class="o">.</span><span class="n">to_dictionary</span><span class="p">(</span><span class="n">dictionary_type</span><span class="o">=</span><span class="s2">&quot;targetgene2TFs&quot;</span><span class="p">)</span>
<span class="n">save_as_pickled_object</span><span class="p">(</span><span class="n">td</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">folder</span><span class="p">,</span> <span class="s2">&quot;TFinfo_targetgene2TFs.pickled&quot;</span><span class="p">))</span>

<span class="c1"># save TFinfo as a dataframe</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">tfi</span><span class="o">.</span><span class="n">to_dataframe</span><span class="p">()</span>
<span class="n">df</span><span class="o">.</span><span class="n">to_parquet</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">folder</span><span class="p">,</span> <span class="s2">&quot;TFinfo_dataframe.parquet&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="motif_data_preparation/01_How_to_load_motif_data.html" class="btn btn-neutral float-right" title="0. Overview: How to use a different motifs for celloracle motif scan." accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="../../tutorials/motifscan.html" class="btn btn-neutral float-left" title="2. Transcription factor binding motif scan" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, Samantha Morris Lab

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>